{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Pipeline to Assemble the genome of bacteria\n",
    "\n",
    "Combination of two bacteria genome assembling pipelines, Tychus and Mypro.  \n",
    "Realization of Tychus with Python scripts. Add SOAPdenovo, improve the genus database for prokka with data from Mypro https://sourceforge.net/projects/sb2nhri/files/MyPro/ "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## setup the files and parameters you are going to work with."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### two reads files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "read1 = '/mnt/d/linux/W/MoonNGS/raw/20180813Novogene16species_Human_Methyl/1.rawdata/MNH08365/MNH08365_1.fq.gz'\n",
    "read2 = '/mnt/d/linux/W/MoonNGS/raw/20180813Novogene16species_Human_Methyl/1.rawdata/MNH08365/MNH08365_2.fq.gz'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### parameters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [],
   "source": [
    "thread = 24\n",
    "outfolder = '/mnt/d/linux/W/MoonNGS/raw/20180813Novogene16species_Human_Methyl/1.rawdata/MNH08365/Tychus'\n",
    "bacteriaStrain = 'MNH08365' # change as needed\n",
    "cleanOrNot = True # whether or not to remove intermediate files"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### usually fixed parameters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "adapters = '/home/p/Trimmomatic-0.38/adapters/TruSeq3-PE.fa'\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## import required packages"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [],
   "source": [
    "import subprocess\n",
    "import os\n",
    "import shutil\n",
    "from Bio import SeqIO\n",
    "import time"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## set up some constants\n",
    "Below are some programs that will be used"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "## program fastqc\n",
    "FASTQC = '/home/p/FastQC/fastqc'\n",
    "## program, Trim reads\n",
    "TRIMMOMATIC = '/home/p/Trimmomatic-0.38/trimmomatic-0.38.jar'\n",
    "## program, reads quality report and Trim\n",
    "FASTP = '/home/p/fastp/fastp'\n",
    "## program, Identify the best kmer for genome assembling\n",
    "KMERGENIE = '/home/p/KmerGenie/kmergenie-1.7048/kmergenie'\n",
    "## program, assemble the genome\n",
    "ABYSS = '/home/p/anaconda/anaconda3_5.2.0/bin/abyss-pe'\n",
    "## program, assemble the genome\n",
    "VELVETH = '/home/p/anaconda/anaconda3_5.2.0/bin/velveth'\n",
    "VELVETG = '/home/p/anaconda/anaconda3_5.2.0/bin/velvetg'\n",
    "## program, assemble the genome\n",
    "SPADES = '/home/p/anaconda/anaconda3_5.2.0/bin/spades.py'\n",
    "## program, assemble the genome\n",
    "FQ2FA = '/home/p/anaconda/anaconda3_5.2.0/bin/fq2fa'\n",
    "IDBA_UD = '/home/p/anaconda/anaconda3_5.2.0/bin/idba_ud'\n",
    "## program, assemble the genome\n",
    "SOAPDENOVO = '/home/p/anaconda/anaconda3_5.2.0/bin/SOAPdenovo-127mer'\n",
    "## program, merge assembled results\n",
    "MERGE = '/home/p/CISA/CISA1.3/Merge.py'\n",
    "CISAPATH = '/home/p/CISA/CISA1.3/'\n",
    "CISA = '/home/p/CISA/CISA1.3/CISA.py'\n",
    "## program, merge assembled results\n",
    "NOVO_STITCH = '/home/p/Novo_Stitch/scripts/main.py'\n",
    "REFALIGNER = '/home/p/RefalignerAssembler/latest/RefAligner'\n",
    "GLPK = '/usr/local/bin/glpsol'\n",
    "## program, annotate merged genome\n",
    "PROKKA = '/home/p/prokka/prokka-master/bin/prokka'\n",
    "## program, evaluate result\n",
    "QUAST = '/home/p/quast/quast-4.6.3/quast.py'\n",
    "\n",
    "## used programs\n",
    "NCUMER = '/home/p/quast/quast-4.6.3/quast_libs/MUMmer/nucmer'\n",
    "MAKEBLASTDB = '/home/p/blast/bin/makeblastdb'\n",
    "BLASTN = '/home/p/blast/bin/blastn'\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## check quality of input reads with fastqc"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "creat the folder to store fastqc result"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "if not os.path.exists(outfolder):\n",
    "    os.mkdir(outfolder)\n",
    "os.chdir(outfolder)\n",
    "if not os.path.exists('fastqc'):\n",
    "    os.mkdir('fastqc')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "run the command"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/home/p/FastQC/fastqc /mnt/d/linux/W/MoonNGS/raw/20180813Novogene16species_Human_Methyl/1.rawdata/MNH08365/MNH08365_1.fq.gz /mnt/d/linux/W/MoonNGS/raw/20180813Novogene16species_Human_Methyl/1.rawdata/MNH08365/MNH08365_2.fq.gz -w 24 -o /mnt/d/linux/W/MoonNGS/raw/20180813Novogene16species_Human_Methyl/1.rawdata/MNH08365/Tychus/fastqc\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "CompletedProcess(args='/home/p/FastQC/fastqc /mnt/d/linux/W/MoonNGS/raw/20180813Novogene16species_Human_Methyl/1.rawdata/MNH08365/MNH08365_1.fq.gz /mnt/d/linux/W/MoonNGS/raw/20180813Novogene16species_Human_Methyl/1.rawdata/MNH08365/MNH08365_2.fq.gz -w 24 -o /mnt/d/linux/W/MoonNGS/raw/20180813Novogene16species_Human_Methyl/1.rawdata/MNH08365/Tychus/fastqc', returncode=0)"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "command_line = '{FASTQC} {read1} {read2} -t {thread} -o {outfolder}/fastqc'.format( \\\n",
    "               FASTQC = FASTQC, read1 = read1, read2 = read2, thread = thread, outfolder = outfolder)\n",
    "print(command_line)\n",
    "subprocess.run(command_line,shell=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## trim reads with TRIMMOMATIC"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "creat the folder to store the result"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "if not os.path.exists('Trimmomatic'):\n",
    "    os.mkdir('Trimmomatic')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "java -jar /home/p/Trimmomatic-0.38/trimmomatic-0.38.jar PE -threads 24 -phred33 /mnt/d/linux/W/MoonNGS/raw/20180813Novogene16species_Human_Methyl/1.rawdata/MNH08365/MNH08365_1.fq.gz /mnt/d/linux/W/MoonNGS/raw/20180813Novogene16species_Human_Methyl/1.rawdata/MNH08365/MNH08365_2.fq.gz                 ./Trimmomatic/read1.pair.fq ./Trimmomatic/read1.single.fq ./Trimmomatic/read2.pair.fq ./Trimmomatic/read2.single.fq                 ILLUMINACLIP:/home/p/Trimmomatic-0.38/adapters/TruSeq3-PE.fa:2:30:10 SLIDINGWINDOW:4:20 MINLEN:50 \n",
      "CompletedProcess(args='java -jar /home/p/Trimmomatic-0.38/trimmomatic-0.38.jar PE -threads 24 -phred33 /mnt/d/linux/W/MoonNGS/raw/20180813Novogene16species_Human_Methyl/1.rawdata/MNH08365/MNH08365_1.fq.gz /mnt/d/linux/W/MoonNGS/raw/20180813Novogene16species_Human_Methyl/1.rawdata/MNH08365/MNH08365_2.fq.gz                 ./Trimmomatic/read1.pair.fq ./Trimmomatic/read1.single.fq ./Trimmomatic/read2.pair.fq ./Trimmomatic/read2.single.fq                 ILLUMINACLIP:/home/p/Trimmomatic-0.38/adapters/TruSeq3-PE.fa:2:30:10 SLIDINGWINDOW:4:20 MINLEN:50 ', returncode=0)\n"
     ]
    }
   ],
   "source": [
    "command_line = '''java -jar {TRIMMOMATIC} PE -threads {thread} -phred33 {read1} {read2} \\\n",
    "                ./Trimmomatic/read1.pair.fq ./Trimmomatic/read1.single.fq ./Trimmomatic/read2.pair.fq ./Trimmomatic/read2.single.fq \\\n",
    "                ILLUMINACLIP:{adapters}:2:30:10 SLIDINGWINDOW:4:20 MINLEN:50 '''.format( \\\n",
    "               TRIMMOMATIC = TRIMMOMATIC, read1 = read1, read2 = read2, thread = thread, adapters = adapters)\n",
    "print(command_line)\n",
    "result = subprocess.run(command_line,shell=True, check=True)\n",
    "print(result)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Identify best kmer with kmergenie"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "if not os.path.exists('kmergenie'):\n",
    "    os.mkdir('kmergenie')\n",
    "fout = open('readlist.txt','w')\n",
    "fout.write(os.path.join(outfolder,'Trimmomatic','read1.pair.fq') + '\\n')\n",
    "fout.write(os.path.join(outfolder,'Trimmomatic','read2.pair.fq') + '\\n')\n",
    "fout.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/home/p/KmerGenie/kmergenie-1.7048/kmergenie /mnt/d/linux/W/MoonNGS/raw/20180813Novogene16species_Human_Methyl/1.rawdata/MNH08365/Tychus/readlist.txt -t 24 -o /mnt/d/linux/W/MoonNGS/raw/20180813Novogene16species_Human_Methyl/1.rawdata/MNH08365/Tychus/kmergenie/kmergenie.histogram \n"
     ]
    }
   ],
   "source": [
    "command_line = '''{KMERGENIE} {outfolder}/readlist.txt -t {thread} -o {outfolder}/kmergenie/kmergenie.histogram '''.format( \\\n",
    "               KMERGENIE = KMERGENIE, outfolder = outfolder, thread = thread, adapters = adapters)\n",
    "print(command_line)\n",
    "result = subprocess.check_output(command_line, shell=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "running histogram estimation\n",
      "list of reads:\n",
      "/mnt/d/linux/W/MoonNGS/raw/20180813Novogene16species_Human_Methyl/1.rawdata/MNH08365/Tychus/Trimmomatic/read1.pair.fq\n",
      "/mnt/d/linux/W/MoonNGS/raw/20180813Novogene16species_Human_Methyl/1.rawdata/MNH08365/Tychus/Trimmomatic/read2.pair.fq\n",
      "Setting maximum kmer length to: 150 bp\n",
      "computing histograms (from k=21 to k=121): 121 111 101 91 81 71 51 61 41 31 21 estimation of the best k so far: 61\n",
      "\n",
      "ntCard wall-clock time over all k values: 34 seconds \n",
      "fitting model to histograms to estimate best k\n",
      "refining estimation around [55; 67], with a step of 2\n",
      "running histogram estimation\n",
      "list of reads:\n",
      "/mnt/d/linux/W/MoonNGS/raw/20180813Novogene16species_Human_Methyl/1.rawdata/MNH08365/Tychus/Trimmomatic/read1.pair.fq\n",
      "/mnt/d/linux/W/MoonNGS/raw/20180813Novogene16species_Human_Methyl/1.rawdata/MNH08365/Tychus/Trimmomatic/read2.pair.fq\n",
      "Setting maximum kmer length to: 150 bp\n",
      "computing histograms (from k=57 to k=67): 63 65 61 59 67 57 table of predicted num. of genomic k-mers: /mnt/d/linux/W/MoonNGS/raw/20180813Novogene16species_Human_Methyl/1.rawdata/MNH08365/Tychus/kmergenie/kmergenie.histogram.dat\n",
      "recommended coverage cut-off for best k: 123\n",
      "best k: 61\n",
      "\n",
      "ntCard wall-clock time over all k values: 30 seconds \n",
      "fitting model to histograms to estimate best k\n",
      "\n"
     ]
    }
   ],
   "source": [
    "result = result.decode('utf-8')\n",
    "print(result)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "best kmer is  61\n"
     ]
    }
   ],
   "source": [
    "result.split('\\n')[-5]\n",
    "bestkmer = result.split('\\n')[-5].split()[-1]\n",
    "print('best kmer is ', bestkmer)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## run abyss"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "if not os.path.exists('abyss'):\n",
    "    os.mkdir('abyss')\n",
    "os.chdir('abyss')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/home/p/anaconda/anaconda3_5.2.0/bin/abyss-pe k=61 np=24  in=\"/mnt/d/linux/W/MoonNGS/raw/20180813Novogene16species_Human_Methyl/1.rawdata/MNH08365/Tychus/Trimmomatic/read1.pair.fq /mnt/d/linux/W/MoonNGS/raw/20180813Novogene16species_Human_Methyl/1.rawdata/MNH08365/Tychus/Trimmomatic/read2.pair.fq\" name=abyss\n"
     ]
    }
   ],
   "source": [
    "command_line = '''{ABYSS} k={bestkmer} np={thread}  in=\"{read1} {read2}\" name=abyss'''.format( \\\n",
    "               ABYSS = ABYSS, bestkmer=bestkmer, outfolder = outfolder, thread = thread, \\\n",
    "               read1 = os.path.join(outfolder,'Trimmomatic','read1.pair.fq'), read2 = os.path.join(outfolder,'Trimmomatic','read2.pair.fq'))\n",
    "print(command_line)\n",
    "result = subprocess.check_output(command_line, shell=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/home/p/anaconda/anaconda3_5.2.0/bin/mpirun -np 24 ABYSS-P -k61 -q3    --coverage-hist=coverage.hist -s abyss-bubbles.fa  -o abyss-1.fa /mnt/d/linux/W/MoonNGS/raw/20180813Novogene16species_Human_Methyl/1.rawdata/MNH08365/Tychus/Trimmomatic/read1.pair.fq /mnt/d/linux/W/MoonNGS/raw/20180813Novogene16species_Human_Methyl/1.rawdata/MNH08365/Tychus/Trimmomatic/read2.pair.fq \n",
      "ABySS 2.1.0\n",
      "ABYSS-P -k61 -q3 --coverage-hist=coverage.hist -s abyss-bubbles.fa -o abyss-1.fa /mnt/d/linux/W/MoonNGS/raw/20180813Novogene16species_Human_Methyl/1.rawdata/MNH08365/Tychus/Trimmomatic/read1.pair.fq /mnt/d/linux/W/MoonNGS/raw/20180813Novogene16species_Human_Methyl/1.rawdata/MNH08365/Tychus/Trimmomatic/read2.pair.fq\n",
      "Running on 24 processors\n",
      "1: Running on host localhost.localdomain\n",
      "2: Running on host localhost.localdomain\n",
      "4: Running on host localhost.localdomain\n",
      "6: Running on host localhost.localdomain\n",
      "9: Running on host localhost.localdomain\n",
      "10: Running on host localhost.localdomain\n",
      "11: Running on host localhost.localdomain\n",
      "12: Running on host localhost.localdomain\n",
      "13: Running on host localhost.localdomain\n",
      "14: Running on host localhost.localdomain\n",
      "16: Running on host localhost.localdomain\n",
      "17: Running on host localhost.localdomain\n",
      "18: Running on host localhost.localdomain\n",
      "19: Running on host localhost.localdomain\n",
      "20: Running on host localhost.localdomain\n",
      "21: Running on host localhost.localdomain\n",
      "22: Running on host localhost.localdomain\n",
      "0: Running on host localhost.localdomain\n",
      "3: Running on host localhost.localdomain\n",
      "5: Running on host localhost.localdomain\n",
      "7: Running on host localhost.localdomain\n",
      "8: Running on host localhost.localdomain\n",
      "15: Running on host localhost.localdomain\n",
      "23: Running on host localhost.localdomain\n",
      "0: Reading `/mnt/d/linux/W/MoonNGS/raw/20180813Novogene16species_Human_Methyl/1.rawdata/MNH08365/Tychus/Trimmomatic/read1.pair.fq'...\n",
      "1: Reading `/mnt/d/linux/W/MoonNGS/raw/20180813Novogene16species_Human_Methyl/1.rawdata/MNH08365/Tychus/Trimmomatic/read2.pair.fq'...\n",
      "0: Loaded 4411541 k-mer.\n",
      "22: Loaded 4426190 k-mer.\n",
      "1: Loaded 4405095 k-mer.\n",
      "23: Loaded 4414719 k-mer.\n",
      "2: Loaded 4410692 k-mer.\n",
      "3: Loaded 4412940 k-mer.\n",
      "4: Loaded 4414846 k-mer.\n",
      "5: Loaded 4417937 k-mer.\n",
      "7: Loaded 4421887 k-mer.\n",
      "6: Loaded 4412197 k-mer.\n",
      "9: Loaded 4406039 k-mer.\n",
      "10: Loaded 4407070 k-mer.\n",
      "11: Loaded 4414708 k-mer.\n",
      "12: Loaded 4420957 k-mer.\n",
      "13: Loaded 4410235 k-mer.\n",
      "8: Loaded 4419871 k-mer.\n",
      "14: Loaded 4418390 k-mer.\n",
      "16: Loaded 4417713 k-mer.\n",
      "17: Loaded 4404158 k-mer.\n",
      "18: Loaded 4410786 k-mer.\n",
      "19: Loaded 4408002 k-mer.\n",
      "20: Loaded 4416974 k-mer.\n",
      "21: Loaded 4422534 k-mer.\n",
      "15: Loaded 4407990 k-mer.\n",
      "Loaded 105933471 k-mer. At least 4.24 GB of RAM is required.\n",
      "Minimum k-mer coverage is 54\n",
      "Using a coverage threshold of 22...\n",
      "The median k-mer coverage is 490\n",
      "The reconstruction is 3038549\n",
      "The k-mer coverage threshold is 22.1359\n",
      "Setting parameter e (erode) to 22\n",
      "Setting parameter E (erodeStrand) to 1\n",
      "Setting parameter c (coverage) to 22.1359\n",
      "Finding adjacent k-mer...\n",
      "15: Added 8907028 edges.\n",
      "22: Added 8942759 edges.\n",
      "1: Added 8899688 edges.\n",
      "23: Added 8919642 edges.\n",
      "2: Added 8911163 edges.\n",
      "3: Added 8916224 edges.\n",
      "4: Added 8920085 edges.\n",
      "5: Added 8925728 edges.\n",
      "7: Added 8934754 edges.\n",
      "6: Added 8914696 edges.\n",
      "9: Added 8901870 edges.\n",
      "10: Added 8903937 edges.\n",
      "11: Added 8919793 edges.\n",
      "12: Added 8930858 edges.\n",
      "13: Added 8911076 edges.\n",
      "8: Added 8930577 edges.\n",
      "14: Added 8927137 edges.\n",
      "16: Added 8925336 edges.\n",
      "17: Added 8897479 edges.\n",
      "18: Added 8911182 edges.\n",
      "19: Added 8905936 edges.\n",
      "20: Added 8924361 edges.\n",
      "21: Added 8936409 edges.\n",
      "0: Added 8913498 edges.\n",
      "Added 214031216 edges.\n",
      "Eroding tips...\n",
      "18: Eroded 2455582 tips.\n",
      "21: Eroded 2560201 tips.\n",
      "5: Eroded 2585392 tips.\n",
      "0: Eroded 2601391 tips.\n",
      "4: Eroded 2622314 tips.\n",
      "13: Eroded 2625968 tips.\n",
      "12: Eroded 2681442 tips.\n",
      "14: Eroded 2680742 tips.\n",
      "23: Eroded 2682742 tips.\n",
      "22: Eroded 2697478 tips.\n",
      "9: Eroded 2692702 tips.\n",
      "10: Eroded 2700046 tips.\n",
      "6: Eroded 2707606 tips.\n",
      "8: Eroded 2717887 tips.\n",
      "20: Eroded 2724011 tips.\n",
      "19: Eroded 2749859 tips.\n",
      "7: Eroded 2764654 tips.\n",
      "2: Eroded 2763610 tips.\n",
      "1: Eroded 2775742 tips.\n",
      "3: Eroded 2781632 tips.\n",
      "17: Eroded 2781637 tips.\n",
      "15: Eroded 2792821 tips.\n",
      "11: Eroded 2799821 tips.\n",
      "16: Eroded 2814066 tips.\n",
      "18: Eroded 376690 tips.\n",
      "17: Eroded 47096 tips.\n",
      "23: Eroded 152582 tips.\n",
      "2: Eroded 71994 tips.\n",
      "14: Eroded 157190 tips.\n",
      "7: Eroded 75034 tips.\n",
      "1: Eroded 57280 tips.\n",
      "0: Eroded 232901 tips.\n",
      "21: Eroded 280135 tips.\n",
      "19: Eroded 82922 tips.\n",
      "10: Eroded 132299 tips.\n",
      "22: Eroded 145393 tips.\n",
      "9: Eroded 138154 tips.\n",
      "15: Eroded 41155 tips.\n",
      "13: Eroded 207683 tips.\n",
      "4: Eroded 214489 tips.\n",
      "6: Eroded 128591 tips.\n",
      "12: Eroded 159630 tips.\n",
      "11: Eroded 36588 tips.\n",
      "3: Eroded 52667 tips.\n",
      "5: Eroded 250822 tips.\n",
      "20: Eroded 112611 tips.\n",
      "8: Eroded 122194 tips.\n",
      "16: Eroded 25383 tips.\n",
      "Eroded 68060829 tips.\n",
      "Pruning tips shorter than 1 bp...\n",
      "1: Pruned 0 tips.\n",
      "4: Pruned 0 tips.\n",
      "10: Pruned 0 tips.\n",
      "16: Pruned 0 tips.\n",
      "18: Pruned 0 tips.\n",
      "6: Pruned 0 tips.\n",
      "12: Pruned 0 tips.\n",
      "15: Pruned 0 tips.\n",
      "7: Pruned 0 tips.\n",
      "14: Pruned 0 tips.\n",
      "20: Pruned 0 tips.\n",
      "0: Pruned 0 tips.\n",
      "19: Pruned 0 tips.\n",
      "22: Pruned 0 tips.\n",
      "13: Pruned 0 tips.\n",
      "9: Pruned 0 tips.\n",
      "3: Pruned 0 tips.\n",
      "8: Pruned 1 tips.\n",
      "5: Pruned 0 tips.\n",
      "11: Pruned 0 tips.\n",
      "21: Pruned 0 tips.\n",
      "2: Pruned 0 tips.\n",
      "23: Pruned 0 tips.\n",
      "17: Pruned 0 tips.\n",
      "Pruned 1 k-mer in 1 tips.\n",
      "Pruning tips shorter than 2 bp...\n",
      "16: Pruned 0 tips.\n",
      "10: Pruned 0 tips.\n",
      "4: Pruned 0 tips.\n",
      "1: Pruned 0 tips.\n",
      "18: Pruned 0 tips.\n",
      "6: Pruned 0 tips.\n",
      "12: Pruned 0 tips.\n",
      "15: Pruned 0 tips.\n",
      "8: Pruned 0 tips.\n",
      "22: Pruned 0 tips.\n",
      "20: Pruned 0 tips.\n",
      "19: Pruned 0 tips.\n",
      "11: Pruned 0 tips.\n",
      "3: Pruned 0 tips.\n",
      "5: Pruned 0 tips.\n",
      "9: Pruned 0 tips.\n",
      "23: Pruned 0 tips.\n",
      "13: Pruned 0 tips.\n",
      "17: Pruned 0 tips.\n",
      "7: Pruned 0 tips.\n",
      "21: Pruned 0 tips.\n",
      "0: Pruned 0 tips.\n",
      "14: Pruned 0 tips.\n",
      "2: Pruned 0 tips.\n",
      "Pruning tips shorter than 4 bp...\n",
      "4: Pruned 0 tips.\n",
      "10: Pruned 0 tips.\n",
      "16: Pruned 0 tips.\n",
      "18: Pruned 0 tips.\n",
      "1: Pruned 0 tips.\n",
      "6: Pruned 0 tips.\n",
      "12: Pruned 0 tips.\n",
      "15: Pruned 0 tips.\n",
      "8: Pruned 0 tips.\n",
      "20: Pruned 0 tips.\n",
      "19: Pruned 0 tips.\n",
      "22: Pruned 0 tips.\n",
      "9: Pruned 0 tips.\n",
      "14: Pruned 0 tips.\n",
      "3: Pruned 0 tips.\n",
      "21: Pruned 0 tips.\n",
      "0: Pruned 0 tips.\n",
      "11: Pruned 0 tips.\n",
      "5: Pruned 0 tips.\n",
      "13: Pruned 0 tips.\n",
      "7: Pruned 0 tips.\n",
      "23: Pruned 0 tips.\n",
      "17: Pruned 0 tips.\n",
      "2: Pruned 0 tips.\n",
      "Pruning tips shorter than 8 bp...\n",
      "10: Pruned 0 tips.\n",
      "4: Pruned 0 tips.\n",
      "16: Pruned 0 tips.\n",
      "1: Pruned 0 tips.\n",
      "18: Pruned 0 tips.\n",
      "6: Pruned 0 tips.\n",
      "12: Pruned 0 tips.\n",
      "8: Pruned 0 tips.\n",
      "15: Pruned 0 tips.\n",
      "22: Pruned 0 tips.\n",
      "11: Pruned 0 tips.\n",
      "14: Pruned 0 tips.\n",
      "20: Pruned 0 tips.\n",
      "19: Pruned 0 tips.\n",
      "9: Pruned 0 tips.\n",
      "3: Pruned 0 tips.\n",
      "7: Pruned 0 tips.\n",
      "5: Pruned 0 tips.\n",
      "13: Pruned 1 tips.\n",
      "23: Pruned 0 tips.\n",
      "17: Pruned 0 tips.\n",
      "21: Pruned 0 tips.\n",
      "0: Pruned 0 tips.\n",
      "2: Pruned 0 tips.\n",
      "Pruned 8 k-mer in 1 tips.\n",
      "Pruning tips shorter than 16 bp...\n",
      "16: Pruned 0 tips.\n",
      "1: Pruned 0 tips.\n",
      "18: Pruned 0 tips.\n",
      "4: Pruned 0 tips.\n",
      "6: Pruned 0 tips.\n",
      "12: Pruned 0 tips.\n",
      "8: Pruned 0 tips.\n",
      "15: Pruned 0 tips.\n",
      "22: Pruned 0 tips.\n",
      "20: Pruned 0 tips.\n",
      "19: Pruned 0 tips.\n",
      "14: Pruned 0 tips.\n",
      "3: Pruned 0 tips.\n",
      "7: Pruned 0 tips.\n",
      "5: Pruned 0 tips.\n",
      "21: Pruned 0 tips.\n",
      "13: Pruned 0 tips.\n",
      "23: Pruned 0 tips.\n",
      "0: Pruned 0 tips.\n",
      "9: Pruned 0 tips.\n",
      "17: Pruned 0 tips.\n",
      "10: Pruned 0 tips.\n",
      "2: Pruned 0 tips.\n",
      "11: Pruned 0 tips.\n",
      "Pruning tips shorter than 32 bp...\n",
      "10: Pruned 0 tips.\n",
      "4: Pruned 0 tips.\n",
      "16: Pruned 0 tips.\n",
      "18: Pruned 0 tips.\n",
      "6: Pruned 0 tips.\n",
      "1: Pruned 0 tips.\n",
      "12: Pruned 0 tips.\n",
      "8: Pruned 0 tips.\n",
      "15: Pruned 0 tips.\n",
      "22: Pruned 0 tips.\n",
      "20: Pruned 0 tips.\n",
      "14: Pruned 0 tips.\n",
      "19: Pruned 0 tips.\n",
      "17: Pruned 0 tips.\n",
      "9: Pruned 0 tips.\n",
      "3: Pruned 0 tips.\n",
      "21: Pruned 0 tips.\n",
      "5: Pruned 0 tips.\n",
      "0: Pruned 0 tips.\n",
      "11: Pruned 0 tips.\n",
      "7: Pruned 0 tips.\n",
      "23: Pruned 0 tips.\n",
      "2: Pruned 0 tips.\n",
      "13: Pruned 0 tips.\n",
      "Pruning tips shorter than 61 bp...\n",
      "4: Pruned 0 tips.\n",
      "10: Pruned 0 tips.\n",
      "16: Pruned 0 tips.\n",
      "6: Pruned 0 tips.\n",
      "8: Pruned 0 tips.\n",
      "18: Pruned 0 tips.\n",
      "12: Pruned 0 tips.\n",
      "22: Pruned 0 tips.\n",
      "19: Pruned 0 tips.\n",
      "20: Pruned 0 tips.\n",
      "15: Pruned 0 tips.\n",
      "14: Pruned 0 tips.\n",
      "1: Pruned 0 tips.\n",
      "11: Pruned 0 tips.\n",
      "3: Pruned 0 tips.\n",
      "5: Pruned 0 tips.\n",
      "23: Pruned 0 tips.\n",
      "17: Pruned 0 tips.\n",
      "13: Pruned 0 tips.\n",
      "7: Pruned 0 tips.\n",
      "9: Pruned 0 tips.\n",
      "21: Pruned 0 tips.\n",
      "2: Pruned 0 tips.\n",
      "0: Pruned 0 tips.\n",
      "Pruned 2 tips in 6 rounds.\n",
      "Marking ambiguous branches...\n",
      "12: Marked 112083 edges of 54679 ambiguous vertices.\n",
      "19: Marked 111885 edges of 54607 ambiguous vertices.\n",
      "4: Marked 113004 edges of 55126 ambiguous vertices.\n",
      "1: Marked 111989 edges of 54691 ambiguous vertices.\n",
      "11: Marked 112369 edges of 54833 ambiguous vertices.\n",
      "16: Marked 112534 edges of 54861 ambiguous vertices.\n",
      "23: Marked 112164 edges of 54721 ambiguous vertices.\n",
      "15: Marked 112522 edges of 54863 ambiguous vertices.\n",
      "8: Marked 112092 edges of 54675 ambiguous vertices.\n",
      "13: Marked 111863 edges of 54564 ambiguous vertices.\n",
      "5: Marked 112552 edges of 54906 ambiguous vertices.\n",
      "2: Marked 111931 edges of 54627 ambiguous vertices.\n",
      "18: Marked 111338 edges of 54347 ambiguous vertices.\n",
      "17: Marked 112230 edges of 54783 ambiguous vertices.\n",
      "7: Marked 112594 edges of 54895 ambiguous vertices.\n",
      "14: Marked 112903 edges of 55060 ambiguous vertices.\n",
      "6: Marked 111923 edges of 54604 ambiguous vertices.\n",
      "9: Marked 111977 edges of 54642 ambiguous vertices.\n",
      "10: Marked 111421 edges of 54364 ambiguous vertices.\n",
      "20: Marked 113251 edges of 55217 ambiguous vertices.\n",
      "21: Marked 112898 edges of 55053 ambiguous vertices.\n",
      "3: Marked 113105 edges of 55171 ambiguous vertices.\n",
      "0: Marked 112268 edges of 54771 ambiguous vertices.\n",
      "22: Marked 112991 edges of 55138 ambiguous vertices.\n",
      "Marked 1315198 ambiguous branches.\n",
      "Removing low-coverage contigs (mean k-mer coverage < 22.1359)...\n",
      "12: Found 1572799 k-mer in 67105 contigs before removing low-coverage contigs.\n",
      "Removed 1446922 k-mer in 28880 low-coverage contigs.\n",
      "10: Found 1571336 k-mer in 66851 contigs before removing low-coverage contigs.\n",
      "Removed 1446357 k-mer in 28692 low-coverage contigs.\n",
      "3: Found 1562677 k-mer in 67001 contigs before removing low-coverage contigs.\n",
      "Removed 1435657 k-mer in 28667 low-coverage contigs.\n",
      "0: Found 1577130 k-mer in 67123 contigs before removing low-coverage contigs.\n",
      "Removed 1450163 k-mer in 28883 low-coverage contigs.\n",
      "4: Found 1588017 k-mer in 67364 contigs before removing low-coverage contigs.\n",
      "Removed 1461467 k-mer in 28958 low-coverage contigs.\n",
      "21: Found 1560368 k-mer in 66866 contigs before removing low-coverage contigs.\n",
      "Removed 1432839 k-mer in 28335 low-coverage contigs.\n",
      "9: Found 1576059 k-mer in 66937 contigs before removing low-coverage contigs.\n",
      "Removed 1449124 k-mer in 28692 low-coverage contigs.\n",
      "16: Found 1574410 k-mer in 66995 contigs before removing low-coverage contigs.\n",
      "Removed 1447964 k-mer in 28718 low-coverage contigs.\n",
      "15: Found 1578890 k-mer in 67449 contigs before removing low-coverage contigs.\n",
      "Removed 1451504 k-mer in 28897 low-coverage contigs.\n",
      "1: Found 1578155 k-mer in 67213 contigs before removing low-coverage contigs.\n",
      "Removed 1451529 k-mer in 28827 low-coverage contigs.\n",
      "11: Found 1577608 k-mer in 66897 contigs before removing low-coverage contigs.\n",
      "Removed 1451382 k-mer in 28751 low-coverage contigs.\n",
      "6: Found 1570280 k-mer in 66920 contigs before removing low-coverage contigs.\n",
      "Removed 1444386 k-mer in 28739 low-coverage contigs.\n",
      "2: Found 1580908 k-mer in 67070 contigs before removing low-coverage contigs.\n",
      "Removed 1454273 k-mer in 28797 low-coverage contigs.\n",
      "17: Found 1593575 k-mer in 67146 contigs before removing low-coverage contigs.\n",
      "Removed 1467411 k-mer in 29215 low-coverage contigs.\n",
      "22: Found 1588544 k-mer in 67790 contigs before removing low-coverage contigs.\n",
      "Removed 1460978 k-mer in 29132 low-coverage contigs.\n",
      "18: Found 1592922 k-mer in 67296 contigs before removing low-coverage contigs.\n",
      "Removed 1466082 k-mer in 29022 low-coverage contigs.\n",
      "23: Found 1580522 k-mer in 67052 contigs before removing low-coverage contigs.\n",
      "Removed 1454833 k-mer in 29046 low-coverage contigs.\n",
      "14: Found 1574472 k-mer in 67309 contigs before removing low-coverage contigs.\n",
      "Removed 1447748 k-mer in 28813 low-coverage contigs.\n",
      "19: Found 1582881 k-mer in 66895 contigs before removing low-coverage contigs.\n",
      "Removed 1457134 k-mer in 28849 low-coverage contigs.\n",
      "8: Found 1558206 k-mer in 66601 contigs before removing low-coverage contigs.\n",
      "Removed 1431648 k-mer in 28430 low-coverage contigs.\n",
      "7: Found 1584274 k-mer in 67229 contigs before removing low-coverage contigs.\n",
      "Removed 1457897 k-mer in 28927 low-coverage contigs.\n",
      "20: Found 1584229 k-mer in 67278 contigs before removing low-coverage contigs.\n",
      "Removed 1456346 k-mer in 28888 low-coverage contigs.\n",
      "13: Found 1581761 k-mer in 66703 contigs before removing low-coverage contigs.\n",
      "Removed 1456942 k-mer in 28761 low-coverage contigs.\n",
      "5: Found 1581443 k-mer in 67200 contigs before removing low-coverage contigs.\n",
      "Removed 1454435 k-mer in 28858 low-coverage contigs.\n",
      "Found 37871466 k-mer in 1610290 contigs before removing low-coverage contigs.\n",
      "Removed 34835021 k-mer in 691777 low-coverage contigs.\n",
      "Splitting ambiguous branches...\n",
      "18: Split 57643 ambigiuous branches.\n",
      "0: Split 57977 ambigiuous branches.\n",
      "2: Split 57713 ambigiuous branches.\n",
      "9: Split 57275 ambigiuous branches.\n",
      "1: Split 57415 ambigiuous branches.\n",
      "16: Split 57450 ambigiuous branches.\n",
      "3: Split 57446 ambigiuous branches.\n",
      "21: Split 57293 ambigiuous branches.\n",
      "22: Split 58089 ambigiuous branches.\n",
      "11: Split 57300 ambigiuous branches.\n",
      "10: Split 57053 ambigiuous branches.\n",
      "12: Split 57755 ambigiuous branches.\n",
      "17: Split 58008 ambigiuous branches.\n",
      "15: Split 57701 ambigiuous branches.\n",
      "23: Split 57842 ambigiuous branches.\n",
      "6: Split 57971 ambigiuous branches.\n",
      "4: Split 57932 ambigiuous branches.\n",
      "14: Split 57642 ambigiuous branches.\n",
      "13: Split 57938 ambigiuous branches.\n",
      "20: Split 57779 ambigiuous branches.\n",
      "19: Split 57396 ambigiuous branches.\n",
      "7: Split 57794 ambigiuous branches.\n",
      "8: Split 57318 ambigiuous branches.\n",
      "5: Split 57824 ambigiuous branches.\n",
      "Split 1383554 ambiguous branches.\n",
      "Removed 34835021 marked k-mer.\n",
      "Eroding tips...\n",
      "5: Eroded 6 tips.\n",
      "20: Eroded 12 tips.\n",
      "10: Eroded 4 tips.\n",
      "15: Eroded 4 tips.\n",
      "21: Eroded 9 tips.\n",
      "1: Eroded 9 tips.\n",
      "6: Eroded 12 tips.\n",
      "3: Eroded 6 tips.\n",
      "23: Eroded 6 tips.\n",
      "19: Eroded 5 tips.\n",
      "12: Eroded 18 tips.\n",
      "2: Eroded 10 tips.\n",
      "8: Eroded 14 tips.\n",
      "18: Eroded 19 tips.\n",
      "11: Eroded 14 tips.\n",
      "14: Eroded 14 tips.\n",
      "13: Eroded 16 tips.\n",
      "7: Eroded 16 tips.\n",
      "22: Eroded 15 tips.\n",
      "0: Eroded 16 tips.\n",
      "9: Eroded 12 tips.\n",
      "17: Eroded 16 tips.\n",
      "4: Eroded 24 tips.\n",
      "16: Eroded 16 tips.\n",
      "1: Eroded 91 tips.\n",
      "15: Eroded 82 tips.\n",
      "3: Eroded 80 tips.\n",
      "18: Eroded 73 tips.\n",
      "2: Eroded 83 tips.\n",
      "9: Eroded 87 tips.\n",
      "0: Eroded 79 tips.\n",
      "20: Eroded 79 tips.\n",
      "7: Eroded 59 tips.\n",
      "8: Eroded 76 tips.\n",
      "22: Eroded 70 tips.\n",
      "19: Eroded 79 tips.\n",
      "21: Eroded 81 tips.\n",
      "4: Eroded 77 tips.\n",
      "5: Eroded 92 tips.\n",
      "13: Eroded 95 tips.\n",
      "12: Eroded 79 tips.\n",
      "11: Eroded 89 tips.\n",
      "6: Eroded 88 tips.\n",
      "14: Eroded 76 tips.\n",
      "10: Eroded 66 tips.\n",
      "17: Eroded 92 tips.\n",
      "23: Eroded 75 tips.\n",
      "16: Eroded 89 tips.\n",
      "Eroded 2230 tips.\n",
      "Pruning tips shorter than 1 bp...\n",
      "5: Pruned 3 tips.\n",
      "13: Pruned 0 tips.\n",
      "10: Pruned 2 tips.\n",
      "4: Pruned 0 tips.\n",
      "12: Pruned 0 tips.\n",
      "16: Pruned 2 tips.\n",
      "21: Pruned 0 tips.\n",
      "14: Pruned 0 tips.\n",
      "23: Pruned 0 tips.\n",
      "7: Pruned 0 tips.\n",
      "22: Pruned 0 tips.\n",
      "0: Pruned 0 tips.\n",
      "2: Pruned 0 tips.\n",
      "19: Pruned 1 tips.\n",
      "11: Pruned 0 tips.\n",
      "6: Pruned 0 tips.\n",
      "18: Pruned 0 tips.\n",
      "1: Pruned 0 tips.\n",
      "9: Pruned 0 tips.\n",
      "20: Pruned 0 tips.\n",
      "3: Pruned 0 tips.\n",
      "8: Pruned 1 tips.\n",
      "17: Pruned 1 tips.\n",
      "15: Pruned 0 tips.\n",
      "Pruned 10 k-mer in 10 tips.\n",
      "Pruning tips shorter than 2 bp...\n",
      "12: Pruned 0 tips.\n",
      "23: Pruned 2 tips.\n",
      "3: Pruned 0 tips.\n",
      "10: Pruned 0 tips.\n",
      "17: Pruned 1 tips.\n",
      "0: Pruned 0 tips.\n",
      "5: Pruned 0 tips.\n",
      "7: Pruned 0 tips.\n",
      "19: Pruned 1 tips.\n",
      "14: Pruned 0 tips.\n",
      "21: Pruned 0 tips.\n",
      "13: Pruned 1 tips.\n",
      "4: Pruned 1 tips.\n",
      "6: Pruned 0 tips.\n",
      "9: Pruned 0 tips.\n",
      "11: Pruned 2 tips.\n",
      "20: Pruned 0 tips.\n",
      "15: Pruned 2 tips.\n",
      "16: Pruned 0 tips.\n",
      "1: Pruned 1 tips.\n",
      "2: Pruned 0 tips.\n",
      "22: Pruned 2 tips.\n",
      "8: Pruned 0 tips.\n",
      "18: Pruned 0 tips.\n",
      "Pruned 18 k-mer in 13 tips.\n",
      "Pruning tips shorter than 4 bp...\n",
      "10: Pruned 0 tips.\n",
      "23: Pruned 0 tips.\n",
      "13: Pruned 0 tips.\n",
      "21: Pruned 2 tips.\n",
      "11: Pruned 0 tips.\n",
      "15: Pruned 1 tips.\n",
      "22: Pruned 1 tips.\n",
      "5: Pruned 0 tips.\n",
      "18: Pruned 1 tips.\n",
      "7: Pruned 1 tips.\n",
      "8: Pruned 0 tips.\n",
      "3: Pruned 1 tips.\n",
      "16: Pruned 0 tips.\n",
      "12: Pruned 0 tips.\n",
      "14: Pruned 0 tips.\n",
      "19: Pruned 0 tips.\n",
      "17: Pruned 0 tips.\n",
      "6: Pruned 0 tips.\n",
      "2: Pruned 1 tips.\n",
      "0: Pruned 2 tips.\n",
      "4: Pruned 0 tips.\n",
      "20: Pruned 1 tips.\n",
      "9: Pruned 0 tips.\n",
      "1: Pruned 1 tips.\n",
      "Pruned 36 k-mer in 12 tips.\n",
      "Pruning tips shorter than 8 bp...\n",
      "10: Pruned 0 tips.\n",
      "18: Pruned 0 tips.\n",
      "15: Pruned 2 tips.\n",
      "2: Pruned 2 tips.\n",
      "23: Pruned 0 tips.\n",
      "21: Pruned 1 tips.\n",
      "1: Pruned 2 tips.\n",
      "22: Pruned 3 tips.\n",
      "8: Pruned 1 tips.\n",
      "12: Pruned 2 tips.\n",
      "6: Pruned 1 tips.\n",
      "16: Pruned 3 tips.\n",
      "0: Pruned 1 tips.\n",
      "17: Pruned 0 tips.\n",
      "5: Pruned 1 tips.\n",
      "13: Pruned 0 tips.\n",
      "14: Pruned 1 tips.\n",
      "4: Pruned 0 tips.\n",
      "19: Pruned 0 tips.\n",
      "20: Pruned 1 tips.\n",
      "3: Pruned 3 tips.\n",
      "9: Pruned 1 tips.\n",
      "7: Pruned 0 tips.\n",
      "11: Pruned 2 tips.\n",
      "Pruned 141 k-mer in 27 tips.\n",
      "Pruning tips shorter than 16 bp...\n",
      "7: Pruned 1 tips.\n",
      "8: Pruned 2 tips.\n",
      "2: Pruned 1 tips.\n",
      "4: Pruned 3 tips.\n",
      "11: Pruned 0 tips.\n",
      "1: Pruned 2 tips.\n",
      "0: Pruned 0 tips.\n",
      "12: Pruned 1 tips.\n",
      "13: Pruned 1 tips.\n",
      "6: Pruned 1 tips.\n",
      "22: Pruned 1 tips.\n",
      "21: Pruned 1 tips.\n",
      "19: Pruned 1 tips.\n",
      "5: Pruned 1 tips.\n",
      "23: Pruned 5 tips.\n",
      "16: Pruned 3 tips.\n",
      "17: Pruned 2 tips.\n",
      "3: Pruned 1 tips.\n",
      "15: Pruned 1 tips.\n",
      "14: Pruned 1 tips.\n",
      "10: Pruned 1 tips.\n",
      "18: Pruned 0 tips.\n",
      "20: Pruned 3 tips.\n",
      "9: Pruned 1 tips.\n",
      "Pruned 346 k-mer in 34 tips.\n",
      "Pruning tips shorter than 32 bp...\n",
      "10: Pruned 0 tips.\n",
      "1: Pruned 1 tips.\n",
      "23: Pruned 2 tips.\n",
      "8: Pruned 3 tips.\n",
      "22: Pruned 2 tips.\n",
      "15: Pruned 1 tips.\n",
      "13: Pruned 1 tips.\n",
      "5: Pruned 2 tips.\n",
      "11: Pruned 3 tips.\n",
      "14: Pruned 1 tips.\n",
      "20: Pruned 2 tips.\n",
      "6: Pruned 2 tips.\n",
      "2: Pruned 2 tips.\n",
      "21: Pruned 2 tips.\n",
      "7: Pruned 2 tips.\n",
      "16: Pruned 2 tips.\n",
      "12: Pruned 2 tips.\n",
      "17: Pruned 2 tips.\n",
      "4: Pruned 3 tips.\n",
      "0: Pruned 4 tips.\n",
      "3: Pruned 8 tips.\n",
      "9: Pruned 2 tips.\n",
      "19: Pruned 4 tips.\n",
      "18: Pruned 2 tips.\n",
      "Pruned 1192 k-mer in 55 tips.\n",
      "Pruning tips shorter than 61 bp...\n",
      "13: Pruned 0 tips.\n",
      "22: Pruned 0 tips.\n",
      "23: Pruned 0 tips.\n",
      "5: Pruned 1 tips.\n",
      "8: Pruned 1 tips.\n",
      "11: Pruned 1 tips.\n",
      "7: Pruned 1 tips.\n",
      "10: Pruned 1 tips.\n",
      "4: Pruned 2 tips.\n",
      "2: Pruned 1 tips.\n",
      "18: Pruned 2 tips.\n",
      "15: Pruned 1 tips.\n",
      "21: Pruned 1 tips.\n",
      "17: Pruned 2 tips.\n",
      "3: Pruned 2 tips.\n",
      "20: Pruned 1 tips.\n",
      "14: Pruned 0 tips.\n",
      "0: Pruned 1 tips.\n",
      "1: Pruned 2 tips.\n",
      "6: Pruned 3 tips.\n",
      "16: Pruned 2 tips.\n",
      "19: Pruned 2 tips.\n",
      "12: Pruned 5 tips.\n",
      "9: Pruned 4 tips.\n",
      "Pruned 1598 k-mer in 36 tips.\n",
      "Pruning tips shorter than 61 bp...\n",
      "6: Pruned 0 tips.\n",
      "17: Pruned 0 tips.\n",
      "20: Pruned 0 tips.\n",
      "13: Pruned 0 tips.\n",
      "19: Pruned 0 tips.\n",
      "3: Pruned 0 tips.\n",
      "16: Pruned 0 tips.\n",
      "1: Pruned 0 tips.\n",
      "4: Pruned 0 tips.\n",
      "12: Pruned 0 tips.\n",
      "10: Pruned 0 tips.\n",
      "23: Pruned 0 tips.\n",
      "22: Pruned 0 tips.\n",
      "8: Pruned 0 tips.\n",
      "21: Pruned 0 tips.\n",
      "11: Pruned 0 tips.\n",
      "5: Pruned 0 tips.\n",
      "7: Pruned 0 tips.\n",
      "18: Pruned 0 tips.\n",
      "2: Pruned 0 tips.\n",
      "9: Pruned 0 tips.\n",
      "15: Pruned 0 tips.\n",
      "14: Pruned 0 tips.\n",
      "0: Pruned 0 tips.\n",
      "Pruned 187 tips in 7 rounds.\n",
      "Popping bubbles...\n",
      "0: Removed 18 bubbles.\n",
      "1: Removed 18 bubbles.\n",
      "2: Removed 16 bubbles.\n",
      "3: Removed 18 bubbles.\n",
      "4: Removed 12 bubbles.\n",
      "5: Removed 16 bubbles.\n",
      "6: Removed 16 bubbles.\n",
      "7: Removed 13 bubbles.\n",
      "8: Removed 8 bubbles.\n",
      "9: Removed 9 bubbles.\n",
      "10: Removed 3 bubbles.\n",
      "11: Removed 13 bubbles.\n",
      "12: Removed 3 bubbles.\n",
      "13: Removed 6 bubbles.\n",
      "14: Removed 7 bubbles.\n",
      "15: Removed 10 bubbles.\n",
      "16: Removed 4 bubbles.\n",
      "17: Removed 4 bubbles.\n",
      "18: Removed 6 bubbles.\n",
      "19: Removed 3 bubbles.\n",
      "20: Removed 4 bubbles.\n",
      "21: Removed 0 bubbles.\n",
      "22: Removed 0 bubbles.\n",
      "23: Removed 1 bubbles.\n",
      "Removed 208 bubbles.\n",
      "Removed 208 bubbles.\n",
      "Marking ambiguous branches...\n",
      "6: Marked 34 edges of 16 ambiguous vertices.\n",
      "1: Marked 42 edges of 20 ambiguous vertices.\n",
      "23: Marked 41 edges of 20 ambiguous vertices.\n",
      "13: Marked 29 edges of 14 ambiguous vertices.\n",
      "18: Marked 37 edges of 18 ambiguous vertices.\n",
      "20: Marked 29 edges of 14 ambiguous vertices.\n",
      "4: Marked 48 edges of 24 ambiguous vertices.\n",
      "2: Marked 29 edges of 14 ambiguous vertices.\n",
      "10: Marked 40 edges of 20 ambiguous vertices.\n",
      "12: Marked 43 edges of 21 ambiguous vertices.\n",
      "14: Marked 36 edges of 18 ambiguous vertices.\n",
      "11: Marked 24 edges of 12 ambiguous vertices.\n",
      "22: Marked 35 edges of 17 ambiguous vertices.\n",
      "3: Marked 48 edges of 23 ambiguous vertices.\n",
      "19: Marked 35 edges of 17 ambiguous vertices.\n",
      "5: Marked 44 edges of 22 ambiguous vertices.\n",
      "8: Marked 39 edges of 19 ambiguous vertices.\n",
      "0: Marked 34 edges of 17 ambiguous vertices.\n",
      "7: Marked 28 edges of 14 ambiguous vertices.\n",
      "16: Marked 51 edges of 25 ambiguous vertices.\n",
      "17: Marked 28 edges of 14 ambiguous vertices.\n",
      "15: Marked 37 edges of 18 ambiguous vertices.\n",
      "21: Marked 40 edges of 20 ambiguous vertices.\n",
      "9: Marked 66 edges of 33 ambiguous vertices.\n",
      "Marked 450 ambiguous branches.\n",
      "Assembling...\n",
      "11: Assembled 39430 k-mer in 24 contigs.\n",
      "17: Assembled 21489 k-mer in 24 contigs.\n",
      "18: Assembled 54712 k-mer in 20 contigs.\n",
      "10: Assembled 63967 k-mer in 29 contigs.\n",
      "14: Assembled 71131 k-mer in 28 contigs.\n",
      "12: Assembled 20992 k-mer in 29 contigs.\n",
      "6: Assembled 93083 k-mer in 25 contigs.\n",
      "13: Assembled 59811 k-mer in 21 contigs.\n",
      "8: Assembled 162380 k-mer in 21 contigs.\n",
      "5: Assembled 60725 k-mer in 23 contigs.\n",
      "9: Assembled 183153 k-mer in 34 contigs.\n",
      "7: Assembled 50420 k-mer in 20 contigs.\n",
      "0: Assembled 178459 k-mer in 24 contigs.\n",
      "16: Assembled 141722 k-mer in 36 contigs.\n",
      "19: Assembled 126161 k-mer in 21 contigs.\n",
      "23: Assembled 136459 k-mer in 34 contigs.\n",
      "1: Assembled 224337 k-mer in 34 contigs.\n",
      "2: Assembled 119261 k-mer in 22 contigs.\n",
      "22: Assembled 195119 k-mer in 28 contigs.\n",
      "3: Assembled 106296 k-mer in 28 contigs.\n",
      "15: Assembled 270788 k-mer in 31 contigs.\n",
      "21: Assembled 113306 k-mer in 27 contigs.\n",
      "20: Assembled 381667 k-mer in 31 contigs.\n",
      "4: Assembled 141770 k-mer in 34 contigs.\n",
      "Assembled 3016638 k-mer in 648 contigs.\n",
      "Concatenating fasta files to abyss-1.fa\n",
      "Concatenating fasta files to abyss-bubbles.fa\n",
      "Done.\n",
      "AdjList    -k61 -m50 --dot abyss-1.fa >abyss-1.dot\n",
      "abyss-filtergraph  --dot   -k61 -g abyss-2.dot1 abyss-1.dot abyss-1.fa >abyss-1.path\n",
      "MergeContigs --dot   -k61 -g abyss-2.dot -o abyss-2.fa abyss-1.fa abyss-2.dot1 abyss-1.path\n",
      "PopBubbles  --dot -j24 -k61  -p0.9  -g abyss-3.dot abyss-2.fa abyss-2.dot >abyss-2.path\n",
      "MergeContigs   -k61 -o abyss-3.fa abyss-2.fa abyss-2.dot abyss-2.path\n",
      "awk '!/^>/ {x[\">\" $1]=1; next} {getline s} $1 in x {print $0 \"\\n\" s}' \\\n",
      "\tabyss-2.path abyss-1.fa >abyss-indel.fa\n",
      "ln -sf abyss-3.fa abyss-unitigs.fa\n",
      "abyss-map   -j24 -l40    /mnt/d/linux/W/MoonNGS/raw/20180813Novogene16species_Human_Methyl/1.rawdata/MNH08365/Tychus/Trimmomatic/read1.pair.fq /mnt/d/linux/W/MoonNGS/raw/20180813Novogene16species_Human_Methyl/1.rawdata/MNH08365/Tychus/Trimmomatic/read2.pair.fq abyss-3.fa \\\n",
      "\t|abyss-fixmate   -l40  -h abyss-3.hist \\\n",
      "\t|sort -snk3 -k4 \\\n",
      "\t|DistanceEst   -j24 -k61  -l40 -s1000 -n10  -o abyss-3.dist abyss-3.hist\n",
      "Overlap  --dot   -k61 -g abyss-4.dot -o abyss-4.fa abyss-3.fa abyss-3.dot abyss-3.dist\n",
      "Overlap: 1\n",
      "Scaffold: 0\n",
      "No overlap: 0\n",
      "Insignificant (<5bp): 0\n",
      "Homopolymer: 0\n",
      "Motif: 0\n",
      "Ambiguous: 0\n",
      "SimpleGraph   -s1000 -n10  -j24 -k61 -o abyss-4.path1 abyss-4.dot abyss-3.dist\n",
      "Seed too short: 0\n",
      "Seeds with no edges: 0\n",
      "Edges removed: 0\n",
      "Total paths attempted: 274\n",
      "Unique path: 162\n",
      "No possible paths: 109\n",
      "No valid paths: 2\n",
      "Repetitive: 0\n",
      "Multiple valid paths: 1\n",
      "Too many solutions: 0\n",
      "Too complex: 0\n",
      "\n",
      "The minimum number of pairs in a distance estimate is 10.\n",
      "The minimum number of pairs used in a path is 10.\n",
      "abyss-index  --fai abyss-3.fa\n",
      "abyss-index  --fai abyss-4.fa\n",
      "cat abyss-3.fa.fai abyss-4.fa.fai \\\n",
      "\t|MergePaths   -j24 -k61 -s1000  -o abyss-4.path2 - abyss-4.path1\n",
      "Ignored 0 paths whose seeds are shorter than 1000 bp.\n",
      "PathOverlap --assemble   -k61  abyss-4.dot abyss-4.path2 >abyss-4.path3\n",
      "cat abyss-3.fa abyss-4.fa \\\n",
      "\t|PathConsensus  --dot -k61  -p0.9  -o abyss-5.path -s abyss-5.fa -g abyss-5.dot - abyss-4.dot abyss-4.path3\n",
      "cat abyss-3.fa abyss-4.fa abyss-5.fa |MergeContigs   -k61 -o abyss-6.fa - abyss-5.dot abyss-5.path\n",
      "ln -sf abyss-6.fa abyss-contigs.fa\n",
      "PathOverlap --overlap   -k61 --dot abyss-5.dot abyss-5.path >abyss-6.dot\n",
      "ln -sf abyss-6.dot abyss-contigs.dot\n",
      "abyss-map   -j24 -l40    /mnt/d/linux/W/MoonNGS/raw/20180813Novogene16species_Human_Methyl/1.rawdata/MNH08365/Tychus/Trimmomatic/read1.pair.fq /mnt/d/linux/W/MoonNGS/raw/20180813Novogene16species_Human_Methyl/1.rawdata/MNH08365/Tychus/Trimmomatic/read2.pair.fq abyss-6.fa \\\n",
      "\t|abyss-fixmate   -l40  -h abyss-6.hist \\\n",
      "\t|sort -snk3 -k4 \\\n",
      "\t|DistanceEst   --dot --mean -j24 -k61  -l40 -s1000 -n10  -o abyss-6.dist.dot abyss-6.hist\n",
      "abyss-scaffold    -k61 -s1000-10000 -n10 -g abyss-6.path.dot  abyss-6.dot abyss-6.dist.dot >abyss-6.path\n",
      "PathConsensus  --dot -k61  -p0.9  -s abyss-7.fa -g abyss-7.dot -o abyss-7.path abyss-6.fa abyss-6.dot abyss-6.path\n",
      "cat abyss-6.fa abyss-7.fa \\\n",
      "\t|MergeContigs   -k61 -o abyss-8.fa - abyss-7.dot abyss-7.path\n",
      "ln -sf abyss-8.fa abyss-scaffolds.fa\n",
      "PathOverlap --overlap   -k61 --dot abyss-7.dot abyss-7.path >abyss-8.dot\n",
      "ln -sf abyss-8.dot abyss-scaffolds.dot\n",
      "abyss-fac   abyss-unitigs.fa abyss-contigs.fa abyss-scaffolds.fa |tee abyss-stats.tab\n",
      "n\tn:500\tL50\tmin\tN80\tN50\tN20\tE-size\tmax\tsum\tname\n",
      "386\t151\t17\t501\t18302\t47884\t127859\t65489\t157763\t3001362\tabyss-unitigs.fa\n",
      "282\t91\t11\t501\t32193\t93105\t164133\t101286\t278946\t3107797\tabyss-contigs.fa\n",
      "273\t83\t10\t501\t39815\t114121\t164905\t111962\t278946\t3107719\tabyss-scaffolds.fa\n",
      "ln -sf abyss-stats.tab abyss-stats\n",
      "tr '\\t' , <abyss-stats.tab >abyss-stats.csv\n",
      "abyss-tabtomd abyss-stats.tab >abyss-stats.md\n",
      "pandoc abyss-stats.md >abyss-stats.html\n",
      "\n"
     ]
    }
   ],
   "source": [
    "result = result.decode('utf-8')\n",
    "print(result)\n",
    "os.chdir('..')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## run velvet"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "if not os.path.exists('velvet'):\n",
    "    os.mkdir('velvet')\n",
    "os.chdir('velvet')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/mnt/d/linux/W/MoonNGS/raw/20180813Novogene16species_Human_Methyl/1.rawdata/MNH08365/Tychus/velvet'"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "os.getcwd()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/home/p/anaconda/anaconda3_5.2.0/bin/velveth . 61 -separate -fastq -shortPaired /mnt/d/linux/W/MoonNGS/raw/20180813Novogene16species_Human_Methyl/1.rawdata/MNH08365/Tychus/Trimmomatic/read1.pair.fq /mnt/d/linux/W/MoonNGS/raw/20180813Novogene16species_Human_Methyl/1.rawdata/MNH08365/Tychus/Trimmomatic/read2.pair.fq \n"
     ]
    }
   ],
   "source": [
    "command_line = '''{VELVETH} . {bestkmer} -separate -fastq -shortPaired {read1} {read2} '''.format( \\\n",
    "               VELVETH = VELVETH, bestkmer=bestkmer, outfolder = outfolder, thread = thread, \\\n",
    "               read1 = os.path.join(outfolder,'Trimmomatic','read1.pair.fq'), read2 = os.path.join(outfolder,'Trimmomatic','read2.pair.fq'))\n",
    "print(command_line)\n",
    "result = subprocess.check_output(command_line, shell=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[0.000000] Reading FastQ file /mnt/d/linux/W/MoonNGS/raw/20180813Novogene16species_Human_Methyl/1.rawdata/MNH08365/Tychus/Trimmomatic/read1.pair.fq;\n",
      "[0.000053] Reading FastQ file /mnt/d/linux/W/MoonNGS/raw/20180813Novogene16species_Human_Methyl/1.rawdata/MNH08365/Tychus/Trimmomatic/read2.pair.fq;\n",
      "[73.855966] 19695418 sequences found in total in the paired sequence files\n",
      "[73.855972] Done\n",
      "[73.874804] Reading read set file ./Sequences;\n",
      "[78.061765] 19695418 sequences found\n",
      "[96.199030] Done\n",
      "[96.199033] 19695418 sequences in total.\n",
      "[96.199078] Writing into roadmap file ./Roadmaps...\n",
      "[125.001654] Inputting sequences...\n",
      "[125.012275] Inputting sequence 0 / 19695418\n",
      "[127.018804] Inputting sequence 14000000 / 19695418\n",
      "[128.189542] Inputting sequence 5000000 / 19695418\n",
      "[130.099874] Inputting sequence 19000000 / 19695418\n",
      "[131.392979] Inputting sequence 10000000 / 19695418\n",
      "[132.537936] Inputting sequence 1000000 / 19695418\n",
      "[135.368511] Inputting sequence 15000000 / 19695418\n",
      "[136.250908] Inputting sequence 6000000 / 19695418\n",
      "[139.221170] Inputting sequence 11000000 / 19695418\n",
      "[140.185788] Inputting sequence 2000000 / 19695418\n",
      "[142.850568] Inputting sequence 16000000 / 19695418\n",
      "[143.629966] Inputting sequence 7000000 / 19695418\n",
      "[146.861881] Inputting sequence 12000000 / 19695418\n",
      "[148.775898] Inputting sequence 3000000 / 19695418\n",
      "[151.156600] Inputting sequence 17000000 / 19695418\n",
      "[151.852703] Inputting sequence 8000000 / 19695418\n",
      "[154.606351] Inputting sequence 13000000 / 19695418\n",
      "[156.805598] Inputting sequence 4000000 / 19695418\n",
      "[158.142798] Inputting sequence 18000000 / 19695418\n",
      "[161.211133] Inputting sequence 9000000 / 19695418\n",
      "[162.445865]  === Sequences loaded in 37.444340 s\n",
      "[162.446827] Done inputting sequences\n",
      "[162.446893] Destroying splay table\n",
      "[163.471140] Splay table destroyed\n",
      "\n"
     ]
    }
   ],
   "source": [
    "result = result.decode('utf-8')\n",
    "print(result)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/home/p/anaconda/anaconda3_5.2.0/bin/velvetg .  -exp_cov auto -cov_cutoff auto \n"
     ]
    }
   ],
   "source": [
    "command_line = '''{VELVETG} .  -exp_cov auto -cov_cutoff auto '''.format( VELVETG = VELVETG)\n",
    "print(command_line)\n",
    "result = subprocess.check_output(command_line, shell=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[0.000000] Reading roadmap file ./Roadmaps\n",
      "[27.580867] 19695418 roadmaps read\n",
      "[27.596294] Creating insertion markers\n",
      "[34.005227] Ordering insertion markers\n",
      "[35.643671] Counting preNodes\n",
      "[39.000879] 5973749 preNodes counted, creating them now\n",
      "[43.344941] Sequence 1000000 / 19695418\n",
      "[47.595644] Sequence 2000000 / 19695418\n",
      "[51.771762] Sequence 3000000 / 19695418\n",
      "[55.746266] Sequence 4000000 / 19695418\n",
      "[60.079233] Sequence 5000000 / 19695418\n",
      "[63.994490] Sequence 6000000 / 19695418\n",
      "[67.916661] Sequence 7000000 / 19695418\n",
      "[71.940916] Sequence 8000000 / 19695418\n",
      "[75.886416] Sequence 9000000 / 19695418\n",
      "[79.999447] Sequence 10000000 / 19695418\n",
      "[83.882210] Sequence 11000000 / 19695418\n",
      "[87.755178] Sequence 12000000 / 19695418\n",
      "[91.761211] Sequence 13000000 / 19695418\n",
      "[96.171603] Sequence 14000000 / 19695418\n",
      "[100.460566] Sequence 15000000 / 19695418\n",
      "[104.717268] Sequence 16000000 / 19695418\n",
      "[108.804571] Sequence 17000000 / 19695418\n",
      "[112.947955] Sequence 18000000 / 19695418\n",
      "[117.227555] Sequence 19000000 / 19695418\n",
      "[119.854399] Adjusting marker info...\n",
      "[123.987706] Connecting preNodes\n",
      "[124.828216] Connecting 5000000 / 19695418\n",
      "[125.493482] Connecting 10000000 / 19695418\n",
      "[125.976913] Connecting 15000000 / 19695418\n",
      "[128.310854] Connecting 3000000 / 19695418\n",
      "[128.897890] Connecting 8000000 / 19695418\n",
      "[129.412827] Connecting 13000000 / 19695418\n",
      "[129.885050] Connecting 18000000 / 19695418\n",
      "[132.018264] Connecting 1000000 / 19695418\n",
      "[132.266004] Connecting 6000000 / 19695418\n",
      "[133.594483] Connecting 11000000 / 19695418\n",
      "[133.897803] Connecting 16000000 / 19695418\n",
      "[136.634692] Connecting 4000000 / 19695418\n",
      "[137.362946] Connecting 9000000 / 19695418\n",
      "[137.948132] Connecting 14000000 / 19695418\n",
      "[138.640469] Connecting 19000000 / 19695418\n",
      "[140.950410] Connecting 2000000 / 19695418\n",
      "[141.336544] Connecting 7000000 / 19695418\n",
      "[142.686674] Connecting 12000000 / 19695418\n",
      "[143.446775] Connecting 17000000 / 19695418\n",
      "[145.916455] Cleaning up memory\n",
      "[146.086971] Done creating preGraph\n",
      "[146.086992] Concatenation...\n",
      "[149.102069] Renumbering preNodes\n",
      "[149.102077] Initial preNode count 5973749\n",
      "[149.437812] Destroyed 2913185 preNodes\n",
      "[149.437815] Concatenation over!\n",
      "[149.437816] Clipping short tips off preGraph\n",
      "[150.274774] Concatenation...\n",
      "[151.118251] Renumbering preNodes\n",
      "[151.118255] Initial preNode count 3060564\n",
      "[151.330721] Destroyed 1369741 preNodes\n",
      "[151.330724] Concatenation over!\n",
      "[151.330725] 821896 tips cut off\n",
      "[151.330725] 1690823 nodes left\n",
      "[151.330778] Writing into pregraph file ./PreGraph...\n",
      "[157.926449] Reading read set file ./Sequences;\n",
      "[161.876149] 19695418 sequences found\n",
      "[180.203999] Done\n",
      "[210.542656] Reading pre-graph file ./PreGraph\n",
      "[210.547680] Graph has 1690823 nodes and 19695418 sequences\n",
      "[214.331161] Scanning pre-graph file ./PreGraph for k-mers\n",
      "[216.903523] 38982239 kmers found\n",
      "[223.205011] Sorting kmer occurence table ... \n",
      "[250.897259] Sorting done.\n",
      "[250.897267] Computing acceleration table... \n",
      "[251.220139] Computing offsets... \n",
      "[251.655569] Ghost Threading through reads 0 / 19695418\n",
      "[255.573451] Ghost Threading through reads 14000000 / 19695418\n",
      "[256.649540] Ghost Threading through reads 5000000 / 19695418\n",
      "[261.101903] Ghost Threading through reads 19000000 / 19695418\n",
      "[263.076510] Ghost Threading through reads 10000000 / 19695418\n",
      "[263.229482] Ghost Threading through reads 1000000 / 19695418\n",
      "[267.085543] Ghost Threading through reads 15000000 / 19695418\n",
      "[267.903953] Ghost Threading through reads 6000000 / 19695418\n",
      "[272.601032] Ghost Threading through reads 11000000 / 19695418\n",
      "[273.768627] Ghost Threading through reads 2000000 / 19695418\n",
      "[278.364547] Ghost Threading through reads 16000000 / 19695418\n",
      "[279.367549] Ghost Threading through reads 7000000 / 19695418\n",
      "[282.920414] Ghost Threading through reads 12000000 / 19695418\n",
      "[283.744935] Ghost Threading through reads 3000000 / 19695418\n",
      "[289.398866] Ghost Threading through reads 17000000 / 19695418\n",
      "[290.564832] Ghost Threading through reads 8000000 / 19695418\n",
      "[296.190894] Ghost Threading through reads 13000000 / 19695418\n",
      "[296.252517] Ghost Threading through reads 4000000 / 19695418\n",
      "[299.175319] Ghost Threading through reads 18000000 / 19695418\n",
      "[300.371045] Ghost Threading through reads 9000000 / 19695418\n",
      "[306.950201]  === Ghost-Threaded in 55.298062 s\n",
      "[306.950323] Threading through reads 0 / 19695418\n",
      "[313.291055] Threading through reads 14000000 / 19695418\n",
      "[315.196540] Threading through reads 5000000 / 19695418\n",
      "[319.140353] Threading through reads 19000000 / 19695418\n",
      "[321.137710] Threading through reads 10000000 / 19695418\n",
      "[323.219635] Threading through reads 1000000 / 19695418\n",
      "[326.736698] Threading through reads 15000000 / 19695418\n",
      "[329.182195] Threading through reads 6000000 / 19695418\n",
      "[335.835499] Threading through reads 11000000 / 19695418\n",
      "[337.656661] Threading through reads 2000000 / 19695418\n",
      "[341.647527] Threading through reads 16000000 / 19695418\n",
      "[342.692551] Threading through reads 7000000 / 19695418\n",
      "[349.018607] Threading through reads 12000000 / 19695418\n",
      "[351.513643] Threading through reads 3000000 / 19695418\n",
      "[355.297664] Threading through reads 17000000 / 19695418\n",
      "[357.299616] Threading through reads 8000000 / 19695418\n",
      "[363.900064] Threading through reads 13000000 / 19695418\n",
      "[364.092295] Threading through reads 4000000 / 19695418\n",
      "[370.096463] Threading through reads 18000000 / 19695418\n",
      "[370.098303] Threading through reads 9000000 / 19695418\n",
      "[374.473369]  === Threaded in 67.523156 s\n",
      "[421.126286] Correcting graph with cutoff 0.200000\n",
      "[421.354338] Determining eligible starting points\n",
      "[424.167845] Done listing starting nodes\n",
      "[424.167850] Initializing todo lists\n",
      "[424.849330] Done with initilization\n",
      "[424.849333] Activating arc lookup table\n",
      "[426.450473] Done activating arc lookup table\n",
      "[426.840379] 10000 / 1690823 nodes visited\n",
      "[427.386106] 20000 / 1690823 nodes visited\n",
      "[427.903083] 30000 / 1690823 nodes visited\n",
      "[428.429216] 40000 / 1690823 nodes visited\n",
      "[428.966965] 50000 / 1690823 nodes visited\n",
      "[429.530333] 60000 / 1690823 nodes visited\n",
      "[430.185899] 70000 / 1690823 nodes visited\n",
      "[430.787305] 80000 / 1690823 nodes visited\n",
      "[431.424606] 90000 / 1690823 nodes visited\n",
      "[431.956872] 100000 / 1690823 nodes visited\n",
      "[432.473826] 110000 / 1690823 nodes visited\n",
      "[432.970712] 120000 / 1690823 nodes visited\n",
      "[433.352679] 130000 / 1690823 nodes visited\n",
      "[433.699025] 140000 / 1690823 nodes visited\n",
      "[434.092772] 150000 / 1690823 nodes visited\n",
      "[434.449488] 160000 / 1690823 nodes visited\n",
      "[434.932847] 170000 / 1690823 nodes visited\n",
      "[435.423484] 180000 / 1690823 nodes visited\n",
      "[435.957694] 190000 / 1690823 nodes visited\n",
      "[436.455587] 200000 / 1690823 nodes visited\n",
      "[436.963538] 210000 / 1690823 nodes visited\n",
      "[437.436750] 220000 / 1690823 nodes visited\n",
      "[437.894380] 230000 / 1690823 nodes visited\n",
      "[438.302467] 240000 / 1690823 nodes visited\n",
      "[438.725989] 250000 / 1690823 nodes visited\n",
      "[439.153151] 260000 / 1690823 nodes visited\n",
      "[439.578455] 270000 / 1690823 nodes visited\n",
      "[440.013908] 280000 / 1690823 nodes visited\n",
      "[440.456285] 290000 / 1690823 nodes visited\n",
      "[440.914423] 300000 / 1690823 nodes visited\n",
      "[441.369811] 310000 / 1690823 nodes visited\n",
      "[441.833085] 320000 / 1690823 nodes visited\n",
      "[442.273340] 330000 / 1690823 nodes visited\n",
      "[442.733505] 340000 / 1690823 nodes visited\n",
      "[443.172153] 350000 / 1690823 nodes visited\n",
      "[443.610244] 360000 / 1690823 nodes visited\n",
      "[444.079275] 370000 / 1690823 nodes visited\n",
      "[444.543316] 380000 / 1690823 nodes visited\n",
      "[445.011759] 390000 / 1690823 nodes visited\n",
      "[445.494385] 400000 / 1690823 nodes visited\n",
      "[445.989732] 410000 / 1690823 nodes visited\n",
      "[446.454779] 420000 / 1690823 nodes visited\n",
      "[446.948821] 430000 / 1690823 nodes visited\n",
      "[447.426976] 440000 / 1690823 nodes visited\n",
      "[447.918461] 450000 / 1690823 nodes visited\n",
      "[448.393040] 460000 / 1690823 nodes visited\n",
      "[448.886462] 470000 / 1690823 nodes visited\n",
      "[449.391026] 480000 / 1690823 nodes visited\n",
      "[449.882970] 490000 / 1690823 nodes visited\n",
      "[450.368717] 500000 / 1690823 nodes visited\n",
      "[450.892375] 510000 / 1690823 nodes visited\n",
      "[451.396857] 520000 / 1690823 nodes visited\n",
      "[451.880343] 530000 / 1690823 nodes visited\n",
      "[452.373969] 540000 / 1690823 nodes visited\n",
      "[452.850144] 550000 / 1690823 nodes visited\n",
      "[453.337992] 560000 / 1690823 nodes visited\n",
      "[453.823058] 570000 / 1690823 nodes visited\n",
      "[454.322662] 580000 / 1690823 nodes visited\n",
      "[454.815704] 590000 / 1690823 nodes visited\n",
      "[455.287327] 600000 / 1690823 nodes visited\n",
      "[455.740237] 610000 / 1690823 nodes visited\n",
      "[456.191879] 620000 / 1690823 nodes visited\n",
      "[456.662927] 630000 / 1690823 nodes visited\n",
      "[457.151787] 640000 / 1690823 nodes visited\n",
      "[457.610856] 650000 / 1690823 nodes visited\n",
      "[458.051506] 660000 / 1690823 nodes visited\n",
      "[458.481964] 670000 / 1690823 nodes visited\n",
      "[458.927007] 680000 / 1690823 nodes visited\n",
      "[459.378664] 690000 / 1690823 nodes visited\n",
      "[459.833493] 700000 / 1690823 nodes visited\n",
      "[460.359183] 710000 / 1690823 nodes visited\n",
      "[460.853018] 720000 / 1690823 nodes visited\n",
      "[461.305105] 730000 / 1690823 nodes visited\n",
      "[461.800045] 740000 / 1690823 nodes visited\n",
      "[462.260087] 750000 / 1690823 nodes visited\n",
      "[462.721074] 760000 / 1690823 nodes visited\n",
      "[463.224969] 770000 / 1690823 nodes visited\n",
      "[463.722306] 780000 / 1690823 nodes visited\n",
      "[464.197066] 790000 / 1690823 nodes visited\n",
      "[464.697498] 800000 / 1690823 nodes visited\n",
      "[465.204005] 810000 / 1690823 nodes visited\n",
      "[465.722395] 820000 / 1690823 nodes visited\n",
      "[466.231405] 830000 / 1690823 nodes visited\n",
      "[466.733954] 840000 / 1690823 nodes visited\n",
      "[467.226764] 850000 / 1690823 nodes visited\n",
      "[467.736952] 860000 / 1690823 nodes visited\n",
      "[468.240873] 870000 / 1690823 nodes visited\n",
      "[468.734583] 880000 / 1690823 nodes visited\n",
      "[469.235012] 890000 / 1690823 nodes visited\n",
      "[469.718146] 900000 / 1690823 nodes visited\n",
      "[470.218134] 910000 / 1690823 nodes visited\n",
      "[470.721531] 920000 / 1690823 nodes visited\n",
      "[471.213506] 930000 / 1690823 nodes visited\n",
      "[471.732167] 940000 / 1690823 nodes visited\n",
      "[472.245261] 950000 / 1690823 nodes visited\n",
      "[472.758610] 960000 / 1690823 nodes visited\n",
      "[473.281921] 970000 / 1690823 nodes visited\n",
      "[473.801056] 980000 / 1690823 nodes visited\n",
      "[474.319150] 990000 / 1690823 nodes visited\n",
      "[474.854599] 1000000 / 1690823 nodes visited\n",
      "[475.396433] 1010000 / 1690823 nodes visited\n",
      "[475.921876] 1020000 / 1690823 nodes visited\n",
      "[476.424550] 1030000 / 1690823 nodes visited\n",
      "[476.970947] 1040000 / 1690823 nodes visited\n",
      "[477.490951] 1050000 / 1690823 nodes visited\n",
      "[478.038211] 1060000 / 1690823 nodes visited\n",
      "[478.577795] 1070000 / 1690823 nodes visited\n",
      "[479.110847] 1080000 / 1690823 nodes visited\n",
      "[479.641178] 1090000 / 1690823 nodes visited\n",
      "[480.211390] 1100000 / 1690823 nodes visited\n",
      "[480.765182] 1110000 / 1690823 nodes visited\n",
      "[481.332920] 1120000 / 1690823 nodes visited\n",
      "[481.877403] 1130000 / 1690823 nodes visited\n",
      "[482.410578] 1140000 / 1690823 nodes visited\n",
      "[482.951434] 1150000 / 1690823 nodes visited\n",
      "[483.493172] 1160000 / 1690823 nodes visited\n",
      "[484.035249] 1170000 / 1690823 nodes visited\n",
      "[484.567690] 1180000 / 1690823 nodes visited\n",
      "[485.074593] 1190000 / 1690823 nodes visited\n",
      "[485.613418] 1200000 / 1690823 nodes visited\n",
      "[486.170953] 1210000 / 1690823 nodes visited\n",
      "[486.746449] 1220000 / 1690823 nodes visited\n",
      "[487.246689] 1230000 / 1690823 nodes visited\n",
      "[487.783840] 1240000 / 1690823 nodes visited\n",
      "[488.312897] 1250000 / 1690823 nodes visited\n",
      "[488.852060] 1260000 / 1690823 nodes visited\n",
      "[489.399105] 1270000 / 1690823 nodes visited\n",
      "[489.924091] 1280000 / 1690823 nodes visited\n",
      "[490.461704] 1290000 / 1690823 nodes visited\n",
      "[490.973969] 1300000 / 1690823 nodes visited\n",
      "[491.491506] 1310000 / 1690823 nodes visited\n",
      "[492.009515] 1320000 / 1690823 nodes visited\n",
      "[492.557323] 1330000 / 1690823 nodes visited\n",
      "[493.103386] 1340000 / 1690823 nodes visited\n",
      "[493.632234] 1350000 / 1690823 nodes visited\n",
      "[494.134523] 1360000 / 1690823 nodes visited\n",
      "[494.677778] 1370000 / 1690823 nodes visited\n",
      "[495.188361] 1380000 / 1690823 nodes visited\n",
      "[495.733040] 1390000 / 1690823 nodes visited\n",
      "[496.280793] 1400000 / 1690823 nodes visited\n",
      "[496.834147] 1410000 / 1690823 nodes visited\n",
      "[497.400502] 1420000 / 1690823 nodes visited\n",
      "[497.970464] 1430000 / 1690823 nodes visited\n",
      "[498.508693] 1440000 / 1690823 nodes visited\n",
      "[499.102706] 1450000 / 1690823 nodes visited\n",
      "[499.709250] 1460000 / 1690823 nodes visited\n",
      "[500.286210] 1470000 / 1690823 nodes visited\n",
      "[500.913237] 1480000 / 1690823 nodes visited\n",
      "[501.540581] 1490000 / 1690823 nodes visited\n",
      "[502.136721] 1500000 / 1690823 nodes visited\n",
      "[502.697595] 1510000 / 1690823 nodes visited\n",
      "[503.266712] 1520000 / 1690823 nodes visited\n",
      "[503.824615] 1530000 / 1690823 nodes visited\n",
      "[504.385105] 1540000 / 1690823 nodes visited\n",
      "[504.925898] 1550000 / 1690823 nodes visited\n",
      "[505.526476] 1560000 / 1690823 nodes visited\n",
      "[506.098489] 1570000 / 1690823 nodes visited\n",
      "[506.682585] 1580000 / 1690823 nodes visited\n",
      "[507.202281] 1590000 / 1690823 nodes visited\n",
      "[507.692319] 1600000 / 1690823 nodes visited\n",
      "[508.149193] 1610000 / 1690823 nodes visited\n",
      "[508.554327] 1620000 / 1690823 nodes visited\n",
      "[508.931824] 1630000 / 1690823 nodes visited\n",
      "[509.328165] 1640000 / 1690823 nodes visited\n",
      "[509.655629] 1650000 / 1690823 nodes visited\n",
      "[509.906291] 1660000 / 1690823 nodes visited\n",
      "[509.992776] 1670000 / 1690823 nodes visited\n",
      "[510.021991] 1680000 / 1690823 nodes visited\n",
      "[510.051073] 1690000 / 1690823 nodes visited\n",
      "[510.083679] 1700000 / 1690823 nodes visited\n",
      "[510.116094] 1710000 / 1690823 nodes visited\n",
      "[510.144361] 1720000 / 1690823 nodes visited\n",
      "[510.145872] Concatenation...\n",
      "[510.227979] Renumbering nodes\n",
      "[510.227982] Initial node count 1690823\n",
      "[510.252529] Removed 1408896 null nodes\n",
      "[510.252532] Concatenation over!\n",
      "[510.252533] Clipping short tips off graph, drastic\n",
      "[510.458286] Concatenation...\n",
      "[513.084725] Renumbering nodes\n",
      "[513.084730] Initial node count 281927\n",
      "[513.112268] Removed 120696 null nodes\n",
      "[513.112270] Concatenation over!\n",
      "[513.112271] 161231 nodes left\n",
      "[513.112454] Writing into graph file ./Graph2...\n",
      "[528.521190] Measuring median coverage depth...\n",
      "[528.669977] Median coverage depth = 388.227723\n",
      "[528.670044] Removing contigs with coverage < 194.113861...\n",
      "[528.848363] Concatenation...\n",
      "[615.104137] Renumbering nodes\n",
      "[615.104145] Initial node count 161231\n",
      "[615.104509] Removed 160666 null nodes\n",
      "[615.104511] Concatenation over!\n",
      "[615.104540] Concatenation...\n",
      "[615.104653] Renumbering nodes\n",
      "[615.104654] Initial node count 565\n",
      "[615.104655] Removed 0 null nodes\n",
      "[615.104655] Concatenation over!\n",
      "[615.104661] Clipping short tips off graph, drastic\n",
      "[615.104677] Concatenation...\n",
      "[615.107712] Renumbering nodes\n",
      "[615.107713] Initial node count 565\n",
      "[615.107720] Removed 4 null nodes\n",
      "[615.107720] Concatenation over!\n",
      "[615.107721] 561 nodes left\n",
      "[615.107724] Read coherency...\n",
      "[615.107731] Identifying unique nodes\n",
      "[615.107741] Done, 163 unique nodes counted\n",
      "[615.107741] Trimming read tips\n",
      "[615.107762] Renumbering nodes\n",
      "[615.107762] Initial node count 561\n",
      "[615.107763] Removed 0 null nodes\n",
      "[615.107764] Confronted to 0 multiple hits and 0 null over 0\n",
      "[615.107764] Read coherency over!\n",
      "[615.223088] Starting pebble resolution...\n",
      "[615.498099] Computing read to node mapping array sizes\n",
      "[616.238381] Computing read to node mappings\n",
      "[617.328016] Estimating library insert lengths...\n",
      "[618.900733] Paired-end library 1 has length: 326, sample standard deviation: 63\n",
      "[618.900743] Done\n",
      "[618.900943] Computing direct node to node mappings\n",
      "[619.849108] Scaffolding node 0\n",
      "[619.967202]  === Nodes Scaffolded in 1.066258 s\n",
      "[620.261066] Preparing to correct graph with cutoff 0.200000\n",
      "[620.633037] Cleaning memory\n",
      "[620.633050] Deactivating local correction settings\n",
      "[620.633093] Pebble done.\n",
      "[620.633094] Starting pebble resolution...\n",
      "[620.751781] Computing read to node mapping array sizes\n",
      "[621.194398] Computing read to node mappings\n",
      "[622.212133] Estimating library insert lengths...\n",
      "[623.663897] Paired-end library 1 has length: 326, sample standard deviation: 63\n",
      "[623.663904] Done\n",
      "[623.664032] Computing direct node to node mappings\n",
      "[624.424624] Scaffolding node 0\n",
      "[624.487887]  === Nodes Scaffolded in 0.823853 s\n",
      "[624.636280] Preparing to correct graph with cutoff 0.200000\n",
      "[624.637392] Cleaning memory\n",
      "[624.637394] Deactivating local correction settings\n",
      "[624.637423] Pebble done.\n",
      "[624.637429] Concatenation...\n",
      "[624.703633] Renumbering nodes\n",
      "[624.703636] Initial node count 561\n",
      "[624.703682] Removed 186 null nodes\n",
      "[624.703685] Concatenation over!\n",
      "[624.703686] Removing reference contigs with coverage < 194.113861...\n",
      "[624.703705] Concatenation...\n",
      "[624.703751] Renumbering nodes\n",
      "[624.703751] Initial node count 375\n",
      "[624.703752] Removed 0 null nodes\n",
      "[624.703754] Concatenation over!\n",
      "[624.880129] Writing contigs into ./contigs.fa...\n",
      "[630.612550] Writing into stats file ./stats.txt...\n",
      "[630.738173] Writing into graph file ./LastGraph...\n",
      "[634.276858] Estimated Coverage = 388.227723\n",
      "[634.276863] Estimated Coverage cutoff = 194.113861\n",
      "Final graph has 375 nodes and n50 of 90990, max 434471, total 3018656, using 19253629/19695418 reads\n",
      "\n"
     ]
    }
   ],
   "source": [
    "result = result.decode('utf-8')\n",
    "print(result)\n",
    "os.chdir('..')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## run SPAdes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "if not os.path.exists('spades'):\n",
    "    os.mkdir('spades')\n",
    "os.chdir('spades')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/mnt/d/linux/W/MoonNGS/raw/20180813Novogene16species_Human_Methyl/1.rawdata/MNH08365/Tychus/spades'"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "os.getcwd()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/home/p/anaconda/anaconda3_5.2.0/bin/spades.py --pe1-1 /mnt/d/linux/W/MoonNGS/raw/20180813Novogene16species_Human_Methyl/1.rawdata/MNH08365/Tychus/Trimmomatic/read1.pair.fq --pe1-2  /mnt/d/linux/W/MoonNGS/raw/20180813Novogene16species_Human_Methyl/1.rawdata/MNH08365/Tychus/Trimmomatic/read2.pair.fq -t 24 -o .  \n"
     ]
    }
   ],
   "source": [
    "command_line = '''{SPADES} --pe1-1 {read1} --pe1-2  {read2} -t {thread} -o .  '''.format( \\\n",
    "               SPADES = SPADES, bestkmer=bestkmer, outfolder = outfolder, thread = thread, \\\n",
    "               read1 = os.path.join(outfolder,'Trimmomatic','read1.pair.fq'), read2 = os.path.join(outfolder,'Trimmomatic','read2.pair.fq'))\n",
    "print(command_line)\n",
    "result = subprocess.check_output(command_line, shell=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Command line: /home/p/anaconda/anaconda3_5.2.0/bin/spades.py\t--pe1-1\t/mnt/d/linux/W/MoonNGS/raw/20180813Novogene16species_Human_Methyl/1.rawdata/MNH08365/Tychus/Trimmomatic/read1.pair.fq\t--pe1-2\t/mnt/d/linux/W/MoonNGS/raw/20180813Novogene16species_Human_Methyl/1.rawdata/MNH08365/Tychus/Trimmomatic/read2.pair.fq\t-t\t24\t-o\t/mnt/d/linux/W/MoonNGS/raw/20180813Novogene16species_Human_Methyl/1.rawdata/MNH08365/Tychus/spades\t\n",
      "\n",
      "System information:\n",
      "  SPAdes version: 3.12.0\n",
      "  Python version: 3.6.4\n",
      "  OS: Linux-3.10.0-862.6.3.el7.x86_64-x86_64-with-centos-7.5.1804-Core\n",
      "\n",
      "Output dir: /mnt/d/linux/W/MoonNGS/raw/20180813Novogene16species_Human_Methyl/1.rawdata/MNH08365/Tychus/spades\n",
      "Mode: read error correction and assembling\n",
      "Debug mode is turned OFF\n",
      "\n",
      "Dataset parameters:\n",
      "  Multi-cell mode (you should set '--sc' flag if input data was obtained with MDA (single-cell) technology or --meta flag if processing metagenomic dataset)\n",
      "  Reads:\n",
      "    Library number: 1, library type: paired-end\n",
      "      orientation: fr\n",
      "      left reads: ['/mnt/d/linux/W/MoonNGS/raw/20180813Novogene16species_Human_Methyl/1.rawdata/MNH08365/Tychus/Trimmomatic/read1.pair.fq']\n",
      "      right reads: ['/mnt/d/linux/W/MoonNGS/raw/20180813Novogene16species_Human_Methyl/1.rawdata/MNH08365/Tychus/Trimmomatic/read2.pair.fq']\n",
      "      interlaced reads: not specified\n",
      "      single reads: not specified\n",
      "      merged reads: not specified\n",
      "Read error correction parameters:\n",
      "  Iterations: 1\n",
      "  PHRED offset will be auto-detected\n",
      "  Corrected reads will be compressed\n",
      "Assembly parameters:\n",
      "  k: automatic selection based on read length\n",
      "  Repeat resolution is enabled\n",
      "  Mismatch careful mode is turned OFF\n",
      "  MismatchCorrector will be SKIPPED\n",
      "  Coverage cutoff is turned OFF\n",
      "Other parameters:\n",
      "  Dir for temp files: /mnt/d/linux/W/MoonNGS/raw/20180813Novogene16species_Human_Methyl/1.rawdata/MNH08365/Tychus/spades/tmp\n",
      "  Threads: 24\n",
      "  Memory limit (in Gb): 62\n",
      "\n",
      "\n",
      "======= SPAdes pipeline started. Log can be found here: /mnt/d/linux/W/MoonNGS/raw/20180813Novogene16species_Human_Methyl/1.rawdata/MNH08365/Tychus/spades/spades.log\n",
      "\n",
      "\n",
      "===== Read error correction started. \n",
      "\n",
      "\n",
      "== Running read error correction tool: /home/p/anaconda/anaconda3_5.2.0/share/spades-3.12.0-1/bin/spades-hammer /mnt/d/linux/W/MoonNGS/raw/20180813Novogene16species_Human_Methyl/1.rawdata/MNH08365/Tychus/spades/corrected/configs/config.info\n",
      "\n",
      "  0:00:00.000     4M / 4M    INFO    General                 (main.cpp                  :  75)   Starting BayesHammer, built from N/A, git revision N/A\n",
      "  0:00:00.001     4M / 4M    INFO    General                 (main.cpp                  :  76)   Loading config from /mnt/d/linux/W/MoonNGS/raw/20180813Novogene16species_Human_Methyl/1.rawdata/MNH08365/Tychus/spades/corrected/configs/config.info\n",
      "  0:00:00.004     4M / 4M    INFO    General                 (main.cpp                  :  78)   Maximum # of threads to use (adjusted due to OMP capabilities): 24\n",
      "  0:00:00.004     4M / 4M    INFO    General                 (memory_limit.cpp          :  49)   Memory limit set to 62 Gb\n",
      "  0:00:00.004     4M / 4M    INFO    General                 (main.cpp                  :  86)   Trying to determine PHRED offset\n",
      "  0:00:00.004     4M / 4M    INFO    General                 (main.cpp                  :  92)   Determined value is 33\n",
      "  0:00:00.004     4M / 4M    INFO    General                 (hammer_tools.cpp          :  36)   Hamming graph threshold tau=1, k=21, subkmer positions = [ 0 10 ]\n",
      "  0:00:00.004     4M / 4M    INFO    General                 (main.cpp                  : 113)   Size of aux. kmer data 24 bytes\n",
      "     === ITERATION 0 begins ===\n",
      "  0:00:00.005     4M / 4M    INFO   K-mer Index Building     (kmer_index_builder.hpp    : 301)   Building kmer index\n",
      "  0:00:00.005     4M / 4M    INFO    General                 (kmer_index_builder.hpp    : 117)   Splitting kmer instances into 384 files using 24 threads. This might take a while.\n",
      "  0:00:00.006     4M / 4M    INFO    General                 (file_limit.hpp            :  32)   Open file limit set to 1024\n",
      "  0:00:00.006     4M / 4M    INFO    General                 (kmer_splitters.hpp        :  89)   Memory available for splitting buffers: 0.861057 Gb\n",
      "  0:00:00.006     4M / 4M    INFO    General                 (kmer_splitters.hpp        :  97)   Using cell size of 174762\n",
      "  0:00:00.147    13G / 13G   INFO   K-mer Splitting          (kmer_data.cpp             :  97)   Processing /mnt/d/linux/W/MoonNGS/raw/20180813Novogene16species_Human_Methyl/1.rawdata/MNH08365/Tychus/Trimmomatic/read1.pair.fq\n",
      "  0:00:24.896    13G / 13G   INFO   K-mer Splitting          (kmer_data.cpp             : 107)   Processed 5135636 reads\n",
      "  0:00:47.573    13G / 13G   INFO   K-mer Splitting          (kmer_data.cpp             : 107)   Processed 9847709 reads\n",
      "  0:00:47.574    13G / 13G   INFO   K-mer Splitting          (kmer_data.cpp             :  97)   Processing /mnt/d/linux/W/MoonNGS/raw/20180813Novogene16species_Human_Methyl/1.rawdata/MNH08365/Tychus/Trimmomatic/read2.pair.fq\n",
      "  0:01:14.368    13G / 13G   INFO   K-mer Splitting          (kmer_data.cpp             : 107)   Processed 15410538 reads\n",
      "  0:01:34.696    13G / 13G   INFO   K-mer Splitting          (kmer_data.cpp             : 107)   Processed 19695418 reads\n",
      "  0:01:34.696    13G / 13G   INFO   K-mer Splitting          (kmer_data.cpp             : 112)   Total 19695418 reads processed\n",
      "  0:01:36.527    96M / 13G   INFO    General                 (kmer_index_builder.hpp    : 120)   Starting k-mer counting.\n",
      "  0:01:37.517    96M / 13G   INFO    General                 (kmer_index_builder.hpp    : 127)   K-mer counting done. There are 116316938 kmers in total.\n",
      "  0:01:37.517    96M / 13G   INFO    General                 (kmer_index_builder.hpp    : 133)   Merging temporary buckets.\n",
      "  0:01:38.792    96M / 13G   INFO   K-mer Index Building     (kmer_index_builder.hpp    : 314)   Building perfect hash indices\n",
      "  0:01:41.077   100M / 13G   INFO    General                 (kmer_index_builder.hpp    : 150)   Merging final buckets.\n",
      "  0:01:42.796   100M / 13G   INFO   K-mer Index Building     (kmer_index_builder.hpp    : 336)   Index built. Total 53944440 bytes occupied (3.71017 bits per kmer).\n",
      "  0:01:42.797   100M / 13G   INFO   K-mer Counting           (kmer_data.cpp             : 356)   Arranging kmers in hash map order\n",
      "  0:01:46.626     1G / 13G   INFO    General                 (main.cpp                  : 148)   Clustering Hamming graph.\n",
      "  0:03:16.623     1G / 13G   INFO    General                 (main.cpp                  : 155)   Extracting clusters\n",
      "  0:04:27.165     1G / 13G   INFO    General                 (main.cpp                  : 167)   Clustering done. Total clusters: 20625902\n",
      "  0:04:27.288   984M / 13G   INFO   K-mer Counting           (kmer_data.cpp             : 376)   Collecting K-mer information, this takes a while.\n",
      "  0:04:28.883     3G / 13G   INFO   K-mer Counting           (kmer_data.cpp             : 382)   Processing /mnt/d/linux/W/MoonNGS/raw/20180813Novogene16species_Human_Methyl/1.rawdata/MNH08365/Tychus/Trimmomatic/read1.pair.fq\n",
      "  0:05:53.822     3G / 13G   INFO   K-mer Counting           (kmer_data.cpp             : 382)   Processing /mnt/d/linux/W/MoonNGS/raw/20180813Novogene16species_Human_Methyl/1.rawdata/MNH08365/Tychus/Trimmomatic/read2.pair.fq\n",
      "  0:07:14.289     3G / 13G   INFO   K-mer Counting           (kmer_data.cpp             : 389)   Collection done, postprocessing.\n",
      "  0:07:14.745     3G / 13G   INFO   K-mer Counting           (kmer_data.cpp             : 403)   There are 116316938 kmers in total. Among them 78944844 (67.8705%) are singletons.\n",
      "  0:07:14.746     3G / 13G   INFO    General                 (main.cpp                  : 173)   Subclustering Hamming graph\n",
      "  0:07:28.942     3G / 13G   INFO   Hamming Subclustering    (kmer_cluster.cpp          : 649)   Subclustering done. Total 19 non-read kmers were generated.\n",
      "  0:07:28.942     3G / 13G   INFO   Hamming Subclustering    (kmer_cluster.cpp          : 650)   Subclustering statistics:\n",
      "  0:07:28.942     3G / 13G   INFO   Hamming Subclustering    (kmer_cluster.cpp          : 651)     Total singleton hamming clusters: 14680360. Among them 3737586 (25.4598%) are good\n",
      "  0:07:28.942     3G / 13G   INFO   Hamming Subclustering    (kmer_cluster.cpp          : 652)     Total singleton subclusters: 28843. Among them 28823 (99.9307%) are good\n",
      "  0:07:28.942     3G / 13G   INFO   Hamming Subclustering    (kmer_cluster.cpp          : 653)     Total non-singleton subcluster centers: 6249577. Among them 6110130 (97.7687%) are good\n",
      "  0:07:28.942     3G / 13G   INFO   Hamming Subclustering    (kmer_cluster.cpp          : 654)     Average size of non-trivial subcluster: 16.263 kmers\n",
      "  0:07:28.942     3G / 13G   INFO   Hamming Subclustering    (kmer_cluster.cpp          : 655)     Average number of sub-clusters per non-singleton cluster: 1.05599\n",
      "  0:07:28.942     3G / 13G   INFO   Hamming Subclustering    (kmer_cluster.cpp          : 656)     Total solid k-mers: 9876539\n",
      "  0:07:28.942     3G / 13G   INFO   Hamming Subclustering    (kmer_cluster.cpp          : 657)     Substitution probabilities: [4,4]((0.931926,0.044178,0.0047499,0.019146),(0.0257645,0.96702,0.00295413,0.00426175),(0.00422791,0.00293985,0.967003,0.0258295),(0.0191177,0.00472561,0.0441311,0.932026))\n",
      "  0:07:28.976     3G / 13G   INFO    General                 (main.cpp                  : 178)   Finished clustering.\n",
      "  0:07:28.976     3G / 13G   INFO    General                 (main.cpp                  : 197)   Starting solid k-mers expansion in 24 threads.\n",
      "  0:08:28.524     3G / 13G   INFO    General                 (main.cpp                  : 218)   Solid k-mers iteration 0 produced 405497 new k-mers.\n",
      "  0:09:28.276     3G / 13G   INFO    General                 (main.cpp                  : 218)   Solid k-mers iteration 1 produced 29799 new k-mers.\n",
      "  0:10:28.327     3G / 13G   INFO    General                 (main.cpp                  : 218)   Solid k-mers iteration 2 produced 1022 new k-mers.\n",
      "  0:11:28.760     3G / 13G   INFO    General                 (main.cpp                  : 218)   Solid k-mers iteration 3 produced 104 new k-mers.\n",
      "  0:12:28.994     3G / 13G   INFO    General                 (main.cpp                  : 218)   Solid k-mers iteration 4 produced 12 new k-mers.\n",
      "  0:13:29.196     3G / 13G   INFO    General                 (main.cpp                  : 218)   Solid k-mers iteration 5 produced 0 new k-mers.\n",
      "  0:13:29.196     3G / 13G   INFO    General                 (main.cpp                  : 222)   Solid k-mers finalized\n",
      "  0:13:29.196     3G / 13G   INFO    General                 (hammer_tools.cpp          : 220)   Starting read correction in 24 threads.\n",
      "  0:13:29.196     3G / 13G   INFO    General                 (hammer_tools.cpp          : 233)   Correcting pair of reads: /mnt/d/linux/W/MoonNGS/raw/20180813Novogene16species_Human_Methyl/1.rawdata/MNH08365/Tychus/Trimmomatic/read1.pair.fq and /mnt/d/linux/W/MoonNGS/raw/20180813Novogene16species_Human_Methyl/1.rawdata/MNH08365/Tychus/Trimmomatic/read2.pair.fq\n",
      "  0:13:40.702     5G / 13G   INFO    General                 (hammer_tools.cpp          : 168)   Prepared batch 0 of 2400000 reads.\n",
      "  0:13:55.859     6G / 13G   INFO    General                 (hammer_tools.cpp          : 175)   Processed batch 0\n",
      "  0:14:00.512     6G / 13G   INFO    General                 (hammer_tools.cpp          : 185)   Written batch 0\n",
      "  0:14:10.671     6G / 13G   INFO    General                 (hammer_tools.cpp          : 168)   Prepared batch 1 of 2400000 reads.\n",
      "  0:14:26.981     6G / 13G   INFO    General                 (hammer_tools.cpp          : 175)   Processed batch 1\n",
      "  0:14:31.674     6G / 13G   INFO    General                 (hammer_tools.cpp          : 185)   Written batch 1\n",
      "  0:14:41.446     6G / 13G   INFO    General                 (hammer_tools.cpp          : 168)   Prepared batch 2 of 2400000 reads.\n",
      "  0:14:56.901     6G / 13G   INFO    General                 (hammer_tools.cpp          : 175)   Processed batch 2\n",
      "  0:15:02.032     6G / 13G   INFO    General                 (hammer_tools.cpp          : 185)   Written batch 2\n",
      "  0:15:12.957     6G / 13G   INFO    General                 (hammer_tools.cpp          : 168)   Prepared batch 3 of 2400000 reads.\n",
      "  0:15:28.704     6G / 13G   INFO    General                 (hammer_tools.cpp          : 175)   Processed batch 3\n",
      "  0:15:34.592     6G / 13G   INFO    General                 (hammer_tools.cpp          : 185)   Written batch 3\n",
      "  0:15:35.619     6G / 13G   INFO    General                 (hammer_tools.cpp          : 168)   Prepared batch 4 of 247709 reads.\n",
      "  0:15:37.564     6G / 13G   INFO    General                 (hammer_tools.cpp          : 175)   Processed batch 4\n",
      "  0:15:38.099     6G / 13G   INFO    General                 (hammer_tools.cpp          : 185)   Written batch 4\n",
      "  0:15:40.354     3G / 13G   INFO    General                 (hammer_tools.cpp          : 274)   Correction done. Changed 4834273 bases in 3366769 reads.\n",
      "  0:15:40.354     3G / 13G   INFO    General                 (hammer_tools.cpp          : 275)   Failed to correct 2402 bases out of 2887288194.\n",
      "  0:15:40.998    96M / 13G   INFO    General                 (main.cpp                  : 255)   Saving corrected dataset description to /mnt/d/linux/W/MoonNGS/raw/20180813Novogene16species_Human_Methyl/1.rawdata/MNH08365/Tychus/spades/corrected/corrected.yaml\n",
      "  0:15:40.998    96M / 13G   INFO    General                 (main.cpp                  : 262)   All done. Exiting.\n",
      "\n",
      "== Compressing corrected reads (with pigz)\n",
      "\n",
      "== Dataset description file was created: /mnt/d/linux/W/MoonNGS/raw/20180813Novogene16species_Human_Methyl/1.rawdata/MNH08365/Tychus/spades/corrected/corrected.yaml\n",
      "\n",
      "\n",
      "===== Read error correction finished. \n",
      "\n",
      "\n",
      "===== Assembling started.\n",
      "\n",
      "\n",
      "== Running assembler: K21\n",
      "\n",
      "  0:00:00.000     4M / 4M    INFO    General                 (main.cpp                  :  74)   Loaded config from /mnt/d/linux/W/MoonNGS/raw/20180813Novogene16species_Human_Methyl/1.rawdata/MNH08365/Tychus/spades/K21/configs/config.info\n",
      "  0:00:00.000     4M / 4M    INFO    General                 (memory_limit.cpp          :  49)   Memory limit set to 62 Gb\n",
      "  0:00:00.000     4M / 4M    INFO    General                 (main.cpp                  :  87)   Starting SPAdes, built from N/A, git revision N/A\n",
      "  0:00:00.000     4M / 4M    INFO    General                 (main.cpp                  :  88)   Maximum k-mer length: 128\n",
      "  0:00:00.000     4M / 4M    INFO    General                 (main.cpp                  :  89)   Assembling dataset (/mnt/d/linux/W/MoonNGS/raw/20180813Novogene16species_Human_Methyl/1.rawdata/MNH08365/Tychus/spades/dataset.info) with K=21\n",
      "  0:00:00.000     4M / 4M    INFO    General                 (main.cpp                  :  90)   Maximum # of threads to use (adjusted due to OMP capabilities): 24\n",
      "  0:00:00.000     4M / 4M    INFO    General                 (launch.hpp                :  51)   SPAdes started\n",
      "  0:00:00.000     4M / 4M    INFO    General                 (launch.hpp                :  58)   Starting from stage: construction\n",
      "  0:00:00.000     4M / 4M    INFO    General                 (launch.hpp                :  65)   Two-step RR enabled: 0\n",
      "  0:00:00.001     4M / 4M    INFO   StageManager             (stage.cpp                 : 132)   STAGE == de Bruijn graph construction\n",
      "  0:00:00.002     4M / 4M    INFO    General                 (read_converter.hpp        :  77)   Converting reads to binary format for library #0 (takes a while)\n",
      "  0:00:00.002     4M / 4M    INFO    General                 (read_converter.hpp        :  78)   Converting paired reads\n",
      "  0:00:00.342    80M / 80M   INFO    General                 (binary_converter.hpp      :  93)   16384 reads processed\n",
      "  0:00:00.583    96M / 96M   INFO    General                 (binary_converter.hpp      :  93)   32768 reads processed\n",
      "  0:00:01.006   128M / 128M  INFO    General                 (binary_converter.hpp      :  93)   65536 reads processed\n",
      "  0:00:01.711   188M / 188M  INFO    General                 (binary_converter.hpp      :  93)   131072 reads processed\n",
      "  0:00:03.204   308M / 308M  INFO    General                 (binary_converter.hpp      :  93)   262144 reads processed\n",
      "  0:00:06.185   548M / 548M  INFO    General                 (binary_converter.hpp      :  93)   524288 reads processed\n",
      "  0:00:15.289   920M / 920M  INFO    General                 (binary_converter.hpp      :  93)   1048576 reads processed\n",
      "  0:00:29.531   920M / 920M  INFO    General                 (binary_converter.hpp      :  93)   2097152 reads processed\n",
      "  0:00:57.933   924M / 924M  INFO    General                 (binary_converter.hpp      :  93)   4194304 reads processed\n",
      "  0:01:57.702   932M / 932M  INFO    General                 (binary_converter.hpp      :  93)   8388608 reads processed\n",
      "  0:02:18.168   720M / 932M  INFO    General                 (binary_converter.hpp      : 117)   9842709 reads written\n",
      "  0:02:18.793     4M / 932M  INFO    General                 (read_converter.hpp        :  87)   Converting single reads\n",
      "  0:02:18.931   200M / 932M  INFO    General                 (binary_converter.hpp      : 117)   4800 reads written\n",
      "  0:02:18.944     4M / 932M  INFO    General                 (read_converter.hpp        :  95)   Converting merged reads\n",
      "  0:02:19.050   196M / 932M  INFO    General                 (binary_converter.hpp      : 117)   0 reads written\n",
      "  0:02:19.062     4M / 932M  INFO    General                 (construction.cpp          : 111)   Max read length 150\n",
      "  0:02:19.062     4M / 932M  INFO    General                 (construction.cpp          : 117)   Average read length 146.615\n",
      "  0:02:19.062     4M / 932M  INFO    General                 (stage.cpp                 : 101)   PROCEDURE == k+1-mer counting\n",
      "  0:02:19.064     4M / 932M  INFO    General                 (kmer_index_builder.hpp    : 117)   Splitting kmer instances into 576 files using 24 threads. This might take a while.\n",
      "  0:02:19.065     4M / 932M  INFO    General                 (file_limit.hpp            :  32)   Open file limit set to 1024\n",
      "  0:02:19.065     4M / 932M  INFO    General                 (kmer_splitters.hpp        :  89)   Memory available for splitting buffers: 0.861057 Gb\n",
      "  0:02:19.065     4M / 932M  INFO    General                 (kmer_splitters.hpp        :  97)   Using cell size of 116508\n",
      "  0:02:33.482    13G / 13G   INFO    General                 (kmer_splitters.hpp        : 289)   Processed 24309926 reads\n",
      "  0:02:41.882    13G / 13G   INFO    General                 (kmer_splitters.hpp        : 289)   Processed 39380436 reads\n",
      "  0:02:41.882    13G / 13G   INFO    General                 (kmer_splitters.hpp        : 295)   Adding contigs from previous K\n",
      "  0:02:43.436    96M / 13G   INFO    General                 (kmer_splitters.hpp        : 308)   Used 39380436 reads\n",
      "  0:02:43.436    96M / 13G   INFO    General                 (kmer_index_builder.hpp    : 120)   Starting k-mer counting.\n",
      "  0:02:43.726    96M / 13G   INFO    General                 (kmer_index_builder.hpp    : 127)   K-mer counting done. There are 9972070 kmers in total.\n",
      "  0:02:43.727    96M / 13G   INFO    General                 (kmer_index_builder.hpp    : 133)   Merging temporary buckets.\n",
      "  0:02:43.868    96M / 13G   INFO    General                 (stage.cpp                 : 101)   PROCEDURE == Extension index construction\n",
      "  0:02:43.868    96M / 13G   INFO   K-mer Index Building     (kmer_index_builder.hpp    : 301)   Building kmer index\n",
      "  0:02:43.868    96M / 13G   INFO    General                 (kmer_index_builder.hpp    : 117)   Splitting kmer instances into 384 files using 24 threads. This might take a while.\n",
      "  0:02:43.868    96M / 13G   INFO    General                 (file_limit.hpp            :  32)   Open file limit set to 1024\n",
      "  0:02:43.868    96M / 13G   INFO    General                 (kmer_splitters.hpp        :  89)   Memory available for splitting buffers: 0.859809 Gb\n",
      "  0:02:43.868    96M / 13G   INFO    General                 (kmer_splitters.hpp        :  97)   Using cell size of 174762\n",
      "  0:02:44.433    13G / 13G   INFO    General                 (kmer_splitters.hpp        : 380)   Processed 9972070 kmers\n",
      "  0:02:44.433    13G / 13G   INFO    General                 (kmer_splitters.hpp        : 385)   Used 9972070 kmers.\n",
      "  0:02:44.580    96M / 13G   INFO    General                 (kmer_index_builder.hpp    : 120)   Starting k-mer counting.\n",
      "  0:02:44.694    96M / 13G   INFO    General                 (kmer_index_builder.hpp    : 127)   K-mer counting done. There are 9957107 kmers in total.\n",
      "  0:02:44.695    96M / 13G   INFO    General                 (kmer_index_builder.hpp    : 133)   Merging temporary buckets.\n",
      "  0:02:44.801    96M / 13G   INFO   K-mer Index Building     (kmer_index_builder.hpp    : 314)   Building perfect hash indices\n",
      "  0:02:45.075    96M / 13G   INFO    General                 (kmer_index_builder.hpp    : 150)   Merging final buckets.\n",
      "  0:02:45.179    96M / 13G   INFO   K-mer Index Building     (kmer_index_builder.hpp    : 336)   Index built. Total 4625512 bytes occupied (3.71635 bits per kmer).\n",
      "  0:02:45.187   108M / 13G   INFO   DeBruijnExtensionIndexBu (kmer_extension_index_build:  99)   Building k-mer extensions from k+1-mers\n",
      "  0:02:45.520   108M / 13G   INFO   DeBruijnExtensionIndexBu (kmer_extension_index_build: 103)   Building k-mer extensions from k+1-mers finished.\n",
      "  0:02:45.521   108M / 13G   INFO    General                 (stage.cpp                 : 101)   PROCEDURE == Early tip clipping\n",
      "  0:02:45.521   108M / 13G   INFO    General                 (construction.cpp          : 253)   Early tip clipper length bound set as (RL - K)\n",
      "  0:02:45.521   108M / 13G   INFO   Early tip clipping       (early_simplification.hpp  : 181)   Early tip clipping\n",
      "  0:02:45.930   108M / 13G   INFO   Early tip clipping       (early_simplification.hpp  : 184)   990930 22-mers were removed by early tip clipper\n",
      "  0:02:45.930   108M / 13G   INFO    General                 (stage.cpp                 : 101)   PROCEDURE == Condensing graph\n",
      "  0:02:45.932   108M / 13G   INFO   UnbranchingPathExtractor (debruijn_graph_constructor: 355)   Extracting unbranching paths\n",
      "  0:02:46.183   116M / 13G   INFO   UnbranchingPathExtractor (debruijn_graph_constructor: 374)   Extracting unbranching paths finished. 173120 sequences extracted\n",
      "  0:02:46.358   116M / 13G   INFO   UnbranchingPathExtractor (debruijn_graph_constructor: 310)   Collecting perfect loops\n",
      "  0:02:46.459   116M / 13G   INFO   UnbranchingPathExtractor (debruijn_graph_constructor: 343)   Collecting perfect loops finished. 272 loops collected\n",
      "  0:02:46.549   116M / 13G   INFO    General                 (stage.cpp                 : 101)   PROCEDURE == Filling coverage indices (PHM)\n",
      "  0:02:46.549   116M / 13G   INFO   K-mer Index Building     (kmer_index_builder.hpp    : 301)   Building kmer index\n",
      "  0:02:46.549   116M / 13G   INFO   K-mer Index Building     (kmer_index_builder.hpp    : 314)   Building perfect hash indices\n",
      "  0:02:46.679   116M / 13G   INFO   K-mer Index Building     (kmer_index_builder.hpp    : 336)   Index built. Total 4636616 bytes occupied (3.71968 bits per kmer).\n",
      "  0:02:46.707   156M / 13G   INFO    General                 (construction.cpp          : 388)   Collecting k-mer coverage information from reads, this takes a while.\n",
      "  0:03:15.566   156M / 13G   INFO    General                 (construction.cpp          : 508)   Filling coverage and flanking coverage from PHM\n",
      "  0:03:15.817   156M / 13G   INFO    General                 (construction.cpp          : 464)   Processed 346631 edges\n",
      "  0:03:15.863   104M / 13G   INFO   StageManager             (stage.cpp                 : 132)   STAGE == EC Threshold Finding\n",
      "  0:03:15.865   104M / 13G   INFO    General                 (kmer_coverage_model.cpp   : 181)   Kmer coverage valley at: 8\n",
      "  0:03:15.865   104M / 13G   INFO    General                 (kmer_coverage_model.cpp   : 201)   K-mer histogram maximum: 785\n",
      "  0:03:15.865   104M / 13G   INFO    General                 (kmer_coverage_model.cpp   : 237)   Estimated median coverage: 785. Coverage mad: 106.747\n",
      "  0:03:15.865   104M / 13G   INFO    General                 (kmer_coverage_model.cpp   : 259)   Fitting coverage model\n",
      "  0:03:16.515   104M / 13G   INFO    General                 (kmer_coverage_model.cpp   : 295)   ... iteration 2\n",
      "  0:03:18.167   104M / 13G   INFO    General                 (kmer_coverage_model.cpp   : 295)   ... iteration 4\n",
      "  0:03:23.969   104M / 13G   INFO    General                 (kmer_coverage_model.cpp   : 295)   ... iteration 8\n",
      "  0:03:28.921   104M / 13G   INFO    General                 (kmer_coverage_model.cpp   : 309)   Fitted mean coverage: 766.355. Fitted coverage std. dev: 96.658\n",
      "  0:03:28.949   104M / 13G   INFO    General                 (kmer_coverage_model.cpp   : 334)   Probability of erroneous kmer at valley: 1\n",
      "  0:03:28.949   104M / 13G   INFO    General                 (kmer_coverage_model.cpp   : 358)   Preliminary threshold calculated as: 463\n",
      "  0:03:28.949   104M / 13G   INFO    General                 (kmer_coverage_model.cpp   : 362)   Threshold adjusted to: 387\n",
      "  0:03:28.949   104M / 13G   INFO    General                 (kmer_coverage_model.cpp   : 375)   Estimated genome size (ignoring repeats): 3000542\n",
      "  0:03:28.949   104M / 13G   INFO    General                 (genomic_info_filler.cpp   : 112)   Mean coverage was calculated as 766.355\n",
      "  0:03:28.949   104M / 13G   INFO    General                 (genomic_info_filler.cpp   : 127)   EC coverage threshold value was calculated as 387\n",
      "  0:03:28.949   104M / 13G   INFO    General                 (genomic_info_filler.cpp   : 128)   Trusted kmer low bound: 0\n",
      "  0:03:28.949   104M / 13G   INFO   StageManager             (stage.cpp                 : 132)   STAGE == Raw Simplification\n",
      "  0:03:28.952   104M / 13G   INFO    General                 (simplification.cpp        : 128)   PROCEDURE == InitialCleaning\n",
      "  0:03:28.954   104M / 13G   INFO    General                 (graph_simplification.hpp  : 662)   Flanking coverage based disconnection disabled\n",
      "  0:03:28.954   104M / 13G   INFO   Simplification           (parallel_processing.hpp   : 165)   Running Self conjugate edge remover\n",
      "  0:03:28.979   104M / 13G   INFO   Simplification           (parallel_processing.hpp   : 167)   Self conjugate edge remover triggered 0 times\n",
      "  0:03:28.979   104M / 13G   INFO   Simplification           (parallel_processing.hpp   : 165)   Running Initial tip clipper\n",
      "  0:03:29.020   104M / 13G   INFO   Simplification           (parallel_processing.hpp   : 167)   Initial tip clipper triggered 992 times\n",
      "  0:03:29.020   104M / 13G   INFO   Simplification           (parallel_processing.hpp   : 165)   Running Initial ec remover\n",
      "  0:03:29.604   104M / 13G   INFO   Simplification           (parallel_processing.hpp   : 167)   Initial ec remover triggered 30972 times\n",
      "  0:03:29.604   104M / 13G   INFO   Simplification           (parallel_processing.hpp   : 165)   Running Initial isolated edge remover\n",
      "  0:03:29.699   104M / 13G   INFO   Simplification           (parallel_processing.hpp   : 167)   Initial isolated edge remover triggered 25673 times\n",
      "  0:03:29.703   104M / 13G   INFO   StageManager             (stage.cpp                 : 132)   STAGE == Simplification\n",
      "  0:03:29.704   104M / 13G   INFO    General                 (simplification.cpp        : 357)   Graph simplification started\n",
      "  0:03:29.704   104M / 13G   INFO    General                 (graph_simplification.hpp  : 634)   Creating parallel br instance\n",
      "  0:03:29.704   104M / 13G   INFO    General                 (simplification.cpp        : 362)   PROCEDURE == Simplification cycle, iteration 1\n",
      "  0:03:29.704   104M / 13G   INFO   Simplification           (parallel_processing.hpp   : 165)   Running Tip clipper\n",
      "  0:03:29.710   104M / 13G   INFO   Simplification           (parallel_processing.hpp   : 167)   Tip clipper triggered 179 times\n",
      "  0:03:29.710   104M / 13G   INFO   Simplification           (parallel_processing.hpp   : 165)   Running Bulge remover\n",
      "  0:03:30.156   104M / 13G   INFO   Simplification           (parallel_processing.hpp   : 167)   Bulge remover triggered 10237 times\n",
      "  0:03:30.156   104M / 13G   INFO   Simplification           (parallel_processing.hpp   : 165)   Running Low coverage edge remover\n",
      "  0:03:30.292   104M / 13G   INFO   Simplification           (parallel_processing.hpp   : 167)   Low coverage edge remover triggered 5828 times\n",
      "  0:03:30.292   104M / 13G   INFO    General                 (simplification.cpp        : 362)   PROCEDURE == Simplification cycle, iteration 2\n",
      "  0:03:30.292   104M / 13G   INFO   Simplification           (parallel_processing.hpp   : 165)   Running Tip clipper\n",
      "  0:03:30.294   104M / 13G   INFO   Simplification           (parallel_processing.hpp   : 167)   Tip clipper triggered 16 times\n",
      "  0:03:30.294   104M / 13G   INFO   Simplification           (parallel_processing.hpp   : 165)   Running Bulge remover\n",
      "  0:03:30.321   104M / 13G   INFO   Simplification           (parallel_processing.hpp   : 167)   Bulge remover triggered 338 times\n",
      "  0:03:30.321   104M / 13G   INFO   Simplification           (parallel_processing.hpp   : 165)   Running Low coverage edge remover\n",
      "  0:03:30.321   104M / 13G   INFO   Simplification           (parallel_processing.hpp   : 167)   Low coverage edge remover triggered 6 times\n",
      "  0:03:30.321   104M / 13G   INFO    General                 (simplification.cpp        : 362)   PROCEDURE == Simplification cycle, iteration 3\n",
      "  0:03:30.321   104M / 13G   INFO   Simplification           (parallel_processing.hpp   : 165)   Running Tip clipper\n",
      "  0:03:30.321   104M / 13G   INFO   Simplification           (parallel_processing.hpp   : 167)   Tip clipper triggered 0 times\n",
      "  0:03:30.321   104M / 13G   INFO   Simplification           (parallel_processing.hpp   : 165)   Running Bulge remover\n",
      "  0:03:30.321   104M / 13G   INFO   Simplification           (parallel_processing.hpp   : 167)   Bulge remover triggered 0 times\n",
      "  0:03:30.321   104M / 13G   INFO   Simplification           (parallel_processing.hpp   : 165)   Running Low coverage edge remover\n",
      "  0:03:30.321   104M / 13G   INFO   Simplification           (parallel_processing.hpp   : 167)   Low coverage edge remover triggered 1 times\n",
      "  0:03:30.321   104M / 13G   INFO    General                 (simplification.cpp        : 362)   PROCEDURE == Simplification cycle, iteration 4\n",
      "  0:03:30.321   104M / 13G   INFO   Simplification           (parallel_processing.hpp   : 165)   Running Tip clipper\n",
      "  0:03:30.321   104M / 13G   INFO   Simplification           (parallel_processing.hpp   : 167)   Tip clipper triggered 0 times\n",
      "  0:03:30.321   104M / 13G   INFO   Simplification           (parallel_processing.hpp   : 165)   Running Bulge remover\n",
      "  0:03:30.321   104M / 13G   INFO   Simplification           (parallel_processing.hpp   : 167)   Bulge remover triggered 0 times\n",
      "  0:03:30.321   104M / 13G   INFO   Simplification           (parallel_processing.hpp   : 165)   Running Low coverage edge remover\n",
      "  0:03:30.321   104M / 13G   INFO   Simplification           (parallel_processing.hpp   : 167)   Low coverage edge remover triggered 0 times\n",
      "  0:03:30.321   104M / 13G   INFO    General                 (simplification.cpp        : 362)   PROCEDURE == Simplification cycle, iteration 5\n",
      "  0:03:30.321   104M / 13G   INFO   Simplification           (parallel_processing.hpp   : 165)   Running Tip clipper\n",
      "  0:03:30.321   104M / 13G   INFO   Simplification           (parallel_processing.hpp   : 167)   Tip clipper triggered 0 times\n",
      "  0:03:30.321   104M / 13G   INFO   Simplification           (parallel_processing.hpp   : 165)   Running Bulge remover\n",
      "  0:03:30.321   104M / 13G   INFO   Simplification           (parallel_processing.hpp   : 167)   Bulge remover triggered 0 times\n",
      "  0:03:30.321   104M / 13G   INFO   Simplification           (parallel_processing.hpp   : 165)   Running Low coverage edge remover\n",
      "  0:03:30.321   104M / 13G   INFO   Simplification           (parallel_processing.hpp   : 167)   Low coverage edge remover triggered 0 times\n",
      "  0:03:30.321   104M / 13G   INFO    General                 (simplification.cpp        : 362)   PROCEDURE == Simplification cycle, iteration 6\n",
      "  0:03:30.321   104M / 13G   INFO   Simplification           (parallel_processing.hpp   : 165)   Running Tip clipper\n",
      "  0:03:30.321   104M / 13G   INFO   Simplification           (parallel_processing.hpp   : 167)   Tip clipper triggered 0 times\n",
      "  0:03:30.321   104M / 13G   INFO   Simplification           (parallel_processing.hpp   : 165)   Running Bulge remover\n",
      "  0:03:30.321   104M / 13G   INFO   Simplification           (parallel_processing.hpp   : 167)   Bulge remover triggered 0 times\n",
      "  0:03:30.321   104M / 13G   INFO   Simplification           (parallel_processing.hpp   : 165)   Running Low coverage edge remover\n",
      "  0:03:30.321   104M / 13G   INFO   Simplification           (parallel_processing.hpp   : 167)   Low coverage edge remover triggered 0 times\n",
      "  0:03:30.321   104M / 13G   INFO    General                 (simplification.cpp        : 362)   PROCEDURE == Simplification cycle, iteration 7\n",
      "  0:03:30.321   104M / 13G   INFO   Simplification           (parallel_processing.hpp   : 165)   Running Tip clipper\n",
      "  0:03:30.321   104M / 13G   INFO   Simplification           (parallel_processing.hpp   : 167)   Tip clipper triggered 0 times\n",
      "  0:03:30.321   104M / 13G   INFO   Simplification           (parallel_processing.hpp   : 165)   Running Bulge remover\n",
      "  0:03:30.321   104M / 13G   INFO   Simplification           (parallel_processing.hpp   : 167)   Bulge remover triggered 0 times\n",
      "  0:03:30.321   104M / 13G   INFO   Simplification           (parallel_processing.hpp   : 165)   Running Low coverage edge remover\n",
      "  0:03:30.321   104M / 13G   INFO   Simplification           (parallel_processing.hpp   : 167)   Low coverage edge remover triggered 0 times\n",
      "  0:03:30.321   104M / 13G   INFO    General                 (simplification.cpp        : 362)   PROCEDURE == Simplification cycle, iteration 8\n",
      "  0:03:30.321   104M / 13G   INFO   Simplification           (parallel_processing.hpp   : 165)   Running Tip clipper\n",
      "  0:03:30.321   104M / 13G   INFO   Simplification           (parallel_processing.hpp   : 167)   Tip clipper triggered 0 times\n",
      "  0:03:30.321   104M / 13G   INFO   Simplification           (parallel_processing.hpp   : 165)   Running Bulge remover\n",
      "  0:03:30.321   104M / 13G   INFO   Simplification           (parallel_processing.hpp   : 167)   Bulge remover triggered 0 times\n",
      "  0:03:30.321   104M / 13G   INFO   Simplification           (parallel_processing.hpp   : 165)   Running Low coverage edge remover\n",
      "  0:03:30.321   104M / 13G   INFO   Simplification           (parallel_processing.hpp   : 167)   Low coverage edge remover triggered 0 times\n",
      "  0:03:30.321   104M / 13G   INFO    General                 (simplification.cpp        : 362)   PROCEDURE == Simplification cycle, iteration 9\n",
      "  0:03:30.321   104M / 13G   INFO   Simplification           (parallel_processing.hpp   : 165)   Running Tip clipper\n",
      "  0:03:30.321   104M / 13G   INFO   Simplification           (parallel_processing.hpp   : 167)   Tip clipper triggered 0 times\n",
      "  0:03:30.321   104M / 13G   INFO   Simplification           (parallel_processing.hpp   : 165)   Running Bulge remover\n",
      "  0:03:30.321   104M / 13G   INFO   Simplification           (parallel_processing.hpp   : 167)   Bulge remover triggered 0 times\n",
      "  0:03:30.321   104M / 13G   INFO   Simplification           (parallel_processing.hpp   : 165)   Running Low coverage edge remover\n",
      "  0:03:30.321   104M / 13G   INFO   Simplification           (parallel_processing.hpp   : 167)   Low coverage edge remover triggered 1 times\n",
      "  0:03:30.321   104M / 13G   INFO    General                 (simplification.cpp        : 362)   PROCEDURE == Simplification cycle, iteration 10\n",
      "  0:03:30.321   104M / 13G   INFO   Simplification           (parallel_processing.hpp   : 165)   Running Tip clipper\n",
      "  0:03:30.321   104M / 13G   INFO   Simplification           (parallel_processing.hpp   : 167)   Tip clipper triggered 0 times\n",
      "  0:03:30.321   104M / 13G   INFO   Simplification           (parallel_processing.hpp   : 165)   Running Bulge remover\n",
      "  0:03:30.321   104M / 13G   INFO   Simplification           (parallel_processing.hpp   : 167)   Bulge remover triggered 0 times\n",
      "  0:03:30.321   104M / 13G   INFO   Simplification           (parallel_processing.hpp   : 165)   Running Low coverage edge remover\n",
      "  0:03:30.321   104M / 13G   INFO   Simplification           (parallel_processing.hpp   : 167)   Low coverage edge remover triggered 5 times\n",
      "  0:03:30.321   104M / 13G   INFO    General                 (simplification.cpp        : 362)   PROCEDURE == Simplification cycle, iteration 11\n",
      "  0:03:30.321   104M / 13G   INFO   Simplification           (parallel_processing.hpp   : 165)   Running Tip clipper\n",
      "  0:03:30.323   104M / 13G   INFO   Simplification           (parallel_processing.hpp   : 167)   Tip clipper triggered 0 times\n",
      "  0:03:30.323   104M / 13G   INFO   Simplification           (parallel_processing.hpp   : 165)   Running Bulge remover\n",
      "  0:03:30.333   104M / 13G   INFO   Simplification           (parallel_processing.hpp   : 167)   Bulge remover triggered 2 times\n",
      "  0:03:30.333   104M / 13G   INFO   Simplification           (parallel_processing.hpp   : 165)   Running Low coverage edge remover\n",
      "  0:03:30.337   104M / 13G   INFO   Simplification           (parallel_processing.hpp   : 167)   Low coverage edge remover triggered 0 times\n",
      "  0:03:30.337   104M / 13G   INFO    General                 (simplification.cpp        : 362)   PROCEDURE == Simplification cycle, iteration 12\n",
      "  0:03:30.337   104M / 13G   INFO   Simplification           (parallel_processing.hpp   : 165)   Running Tip clipper\n",
      "  0:03:30.337   104M / 13G   INFO   Simplification           (parallel_processing.hpp   : 167)   Tip clipper triggered 0 times\n",
      "  0:03:30.337   104M / 13G   INFO   Simplification           (parallel_processing.hpp   : 165)   Running Bulge remover\n",
      "  0:03:30.337   104M / 13G   INFO   Simplification           (parallel_processing.hpp   : 167)   Bulge remover triggered 0 times\n",
      "  0:03:30.337   104M / 13G   INFO   Simplification           (parallel_processing.hpp   : 165)   Running Low coverage edge remover\n",
      "  0:03:30.337   104M / 13G   INFO   Simplification           (parallel_processing.hpp   : 167)   Low coverage edge remover triggered 0 times\n",
      "  0:03:30.337   104M / 13G   INFO   StageManager             (stage.cpp                 : 132)   STAGE == Simplification Cleanup\n",
      "  0:03:30.337   104M / 13G   INFO    General                 (simplification.cpp        : 196)   PROCEDURE == Post simplification\n",
      "  0:03:30.337   104M / 13G   INFO    General                 (graph_simplification.hpp  : 453)   Disconnection of relatively low covered edges disabled\n",
      "  0:03:30.337   104M / 13G   INFO    General                 (graph_simplification.hpp  : 489)   Complex tip clipping disabled\n",
      "  0:03:30.338   104M / 13G   INFO    General                 (graph_simplification.hpp  : 634)   Creating parallel br instance\n",
      "  0:03:30.338   104M / 13G   INFO   Simplification           (parallel_processing.hpp   : 165)   Running Tip clipper\n",
      "  0:03:30.339   104M / 13G   INFO   Simplification           (parallel_processing.hpp   : 167)   Tip clipper triggered 0 times\n",
      "  0:03:30.339   104M / 13G   INFO   Simplification           (parallel_processing.hpp   : 165)   Running Bulge remover\n",
      "  0:03:30.346   104M / 13G   INFO   Simplification           (parallel_processing.hpp   : 167)   Bulge remover triggered 0 times\n",
      "  0:03:30.346   104M / 13G   INFO   Simplification           (parallel_processing.hpp   : 165)   Running Tip clipper\n",
      "  0:03:30.348   104M / 13G   INFO   Simplification           (parallel_processing.hpp   : 167)   Tip clipper triggered 0 times\n",
      "  0:03:30.348   104M / 13G   INFO   Simplification           (parallel_processing.hpp   : 165)   Running Bulge remover\n",
      "  0:03:30.355   104M / 13G   INFO   Simplification           (parallel_processing.hpp   : 167)   Bulge remover triggered 0 times\n",
      "  0:03:30.355   104M / 13G   INFO    General                 (simplification.cpp        : 330)   Disrupting self-conjugate edges\n",
      "  0:03:30.361   104M / 13G   INFO   Simplification           (parallel_processing.hpp   : 165)   Running Removing isolated edges\n",
      "  0:03:30.362   104M / 13G   INFO   Simplification           (parallel_processing.hpp   : 167)   Removing isolated edges triggered 35 times\n",
      "  0:03:30.362   104M / 13G   INFO    General                 (simplification.cpp        : 470)   Counting average coverage\n",
      "  0:03:30.365   104M / 13G   INFO    General                 (simplification.cpp        : 476)   Average coverage = 515.215\n",
      "  0:03:30.365   104M / 13G   INFO   StageManager             (stage.cpp                 : 132)   STAGE == Contig Output\n",
      "  0:03:30.366   104M / 13G   INFO    General                 (contig_output_stage.cpp   :  40)   Writing GFA to /mnt/d/linux/W/MoonNGS/raw/20180813Novogene16species_Human_Methyl/1.rawdata/MNH08365/Tychus/spades//K21/assembly_graph_with_scaffolds.gfa\n",
      "  0:03:30.427   104M / 13G   INFO    General                 (contig_output.hpp         :  22)   Outputting contigs to /mnt/d/linux/W/MoonNGS/raw/20180813Novogene16species_Human_Methyl/1.rawdata/MNH08365/Tychus/spades//K21/before_rr.fasta\n",
      "  0:03:30.512   104M / 13G   INFO    General                 (contig_output_stage.cpp   :  51)   Outputting FastG graph to /mnt/d/linux/W/MoonNGS/raw/20180813Novogene16species_Human_Methyl/1.rawdata/MNH08365/Tychus/spades//K21/assembly_graph.fastg\n",
      "  0:03:30.747   104M / 13G   INFO    General                 (contig_output.hpp         :  22)   Outputting contigs to /mnt/d/linux/W/MoonNGS/raw/20180813Novogene16species_Human_Methyl/1.rawdata/MNH08365/Tychus/spades//K21/simplified_contigs.fasta\n",
      "  0:03:30.836   104M / 13G   INFO    General                 (contig_output.hpp         :  22)   Outputting contigs to /mnt/d/linux/W/MoonNGS/raw/20180813Novogene16species_Human_Methyl/1.rawdata/MNH08365/Tychus/spades//K21/final_contigs.fasta\n",
      "  0:03:30.924   104M / 13G   INFO   StageManager             (stage.cpp                 : 132)   STAGE == Contig Output\n",
      "  0:03:30.924   104M / 13G   INFO    General                 (contig_output_stage.cpp   :  40)   Writing GFA to /mnt/d/linux/W/MoonNGS/raw/20180813Novogene16species_Human_Methyl/1.rawdata/MNH08365/Tychus/spades//K21/assembly_graph_with_scaffolds.gfa\n",
      "  0:03:30.986   104M / 13G   INFO    General                 (contig_output.hpp         :  22)   Outputting contigs to /mnt/d/linux/W/MoonNGS/raw/20180813Novogene16species_Human_Methyl/1.rawdata/MNH08365/Tychus/spades//K21/before_rr.fasta\n",
      "  0:03:31.072   104M / 13G   INFO    General                 (contig_output_stage.cpp   :  51)   Outputting FastG graph to /mnt/d/linux/W/MoonNGS/raw/20180813Novogene16species_Human_Methyl/1.rawdata/MNH08365/Tychus/spades//K21/assembly_graph.fastg\n",
      "  0:03:31.312   104M / 13G   INFO    General                 (contig_output.hpp         :  22)   Outputting contigs to /mnt/d/linux/W/MoonNGS/raw/20180813Novogene16species_Human_Methyl/1.rawdata/MNH08365/Tychus/spades//K21/simplified_contigs.fasta\n",
      "  0:03:31.400   104M / 13G   INFO    General                 (contig_output.hpp         :  22)   Outputting contigs to /mnt/d/linux/W/MoonNGS/raw/20180813Novogene16species_Human_Methyl/1.rawdata/MNH08365/Tychus/spades//K21/final_contigs.fasta\n",
      "  0:03:31.487   104M / 13G   INFO    General                 (launch.hpp                : 149)   SPAdes finished\n",
      "  0:03:31.528    96M / 13G   INFO    General                 (main.cpp                  : 109)   Assembling time: 0 hours 3 minutes 31 seconds\n",
      "Max read length detected as 150\n",
      "Default k-mer sizes were set to [21, 33, 55, 77] because estimated read length (150) is equal to or greater than 150\n",
      "\n",
      "== Running assembler: K33\n",
      "\n",
      "  0:00:00.000     4M / 4M    INFO    General                 (main.cpp                  :  74)   Loaded config from /mnt/d/linux/W/MoonNGS/raw/20180813Novogene16species_Human_Methyl/1.rawdata/MNH08365/Tychus/spades/K33/configs/config.info\n",
      "  0:00:00.000     4M / 4M    INFO    General                 (memory_limit.cpp          :  49)   Memory limit set to 62 Gb\n",
      "  0:00:00.000     4M / 4M    INFO    General                 (main.cpp                  :  87)   Starting SPAdes, built from N/A, git revision N/A\n",
      "  0:00:00.000     4M / 4M    INFO    General                 (main.cpp                  :  88)   Maximum k-mer length: 128\n",
      "  0:00:00.000     4M / 4M    INFO    General                 (main.cpp                  :  89)   Assembling dataset (/mnt/d/linux/W/MoonNGS/raw/20180813Novogene16species_Human_Methyl/1.rawdata/MNH08365/Tychus/spades/dataset.info) with K=33\n",
      "  0:00:00.000     4M / 4M    INFO    General                 (main.cpp                  :  90)   Maximum # of threads to use (adjusted due to OMP capabilities): 24\n",
      "  0:00:00.000     4M / 4M    INFO    General                 (launch.hpp                :  51)   SPAdes started\n",
      "  0:00:00.000     4M / 4M    INFO    General                 (launch.hpp                :  58)   Starting from stage: construction\n",
      "  0:00:00.000     4M / 4M    INFO    General                 (launch.hpp                :  65)   Two-step RR enabled: 0\n",
      "  0:00:00.000     4M / 4M    INFO   StageManager             (stage.cpp                 : 132)   STAGE == de Bruijn graph construction\n",
      "  0:00:00.001     4M / 4M    INFO    General                 (read_converter.hpp        :  59)   Binary reads detected\n",
      "  0:00:00.003     4M / 4M    INFO    General                 (construction.cpp          : 111)   Max read length 150\n",
      "  0:00:00.003     4M / 4M    INFO    General                 (construction.cpp          : 117)   Average read length 146.615\n",
      "  0:00:00.003     4M / 4M    INFO    General                 (stage.cpp                 : 101)   PROCEDURE == k+1-mer counting\n",
      "  0:00:00.003     4M / 4M    INFO    General                 (kmer_index_builder.hpp    : 117)   Splitting kmer instances into 576 files using 24 threads. This might take a while.\n",
      "  0:00:00.003     4M / 4M    INFO    General                 (file_limit.hpp            :  32)   Open file limit set to 1024\n",
      "  0:00:00.003     4M / 4M    INFO    General                 (kmer_splitters.hpp        :  89)   Memory available for splitting buffers: 0.861057 Gb\n",
      "  0:00:00.003     4M / 4M    INFO    General                 (kmer_splitters.hpp        :  97)   Using cell size of 58254\n",
      "  0:00:11.480    13G / 13G   INFO    General                 (kmer_splitters.hpp        : 289)   Processed 13544415 reads\n",
      "  0:00:23.676    13G / 13G   INFO    General                 (kmer_splitters.hpp        : 289)   Processed 27060386 reads\n",
      "  0:00:34.616    13G / 13G   INFO    General                 (kmer_splitters.hpp        : 289)   Processed 39380436 reads\n",
      "  0:00:34.616    13G / 13G   INFO    General                 (kmer_splitters.hpp        : 295)   Adding contigs from previous K\n",
      "  0:00:36.938    96M / 13G   INFO    General                 (kmer_splitters.hpp        : 308)   Used 39380436 reads\n",
      "  0:00:36.938    96M / 13G   INFO    General                 (kmer_index_builder.hpp    : 120)   Starting k-mer counting.\n",
      "  0:00:38.022    96M / 13G   INFO    General                 (kmer_index_builder.hpp    : 127)   K-mer counting done. There are 10045941 kmers in total.\n",
      "  0:00:38.022    96M / 13G   INFO    General                 (kmer_index_builder.hpp    : 133)   Merging temporary buckets.\n",
      "  0:00:38.283    96M / 13G   INFO    General                 (stage.cpp                 : 101)   PROCEDURE == Extension index construction\n",
      "  0:00:38.283    96M / 13G   INFO   K-mer Index Building     (kmer_index_builder.hpp    : 301)   Building kmer index\n",
      "  0:00:38.283    96M / 13G   INFO    General                 (kmer_index_builder.hpp    : 117)   Splitting kmer instances into 384 files using 24 threads. This might take a while.\n",
      "  0:00:38.283    96M / 13G   INFO    General                 (file_limit.hpp            :  32)   Open file limit set to 1024\n",
      "  0:00:38.283    96M / 13G   INFO    General                 (kmer_splitters.hpp        :  89)   Memory available for splitting buffers: 0.859809 Gb\n",
      "  0:00:38.283    96M / 13G   INFO    General                 (kmer_splitters.hpp        :  97)   Using cell size of 87381\n",
      "  0:00:38.821    13G / 13G   INFO    General                 (kmer_splitters.hpp        : 380)   Processed 10045941 kmers\n",
      "  0:00:38.821    13G / 13G   INFO    General                 (kmer_splitters.hpp        : 385)   Used 10045941 kmers.\n",
      "  0:00:38.972    96M / 13G   INFO    General                 (kmer_index_builder.hpp    : 120)   Starting k-mer counting.\n",
      "  0:00:39.134    96M / 13G   INFO    General                 (kmer_index_builder.hpp    : 127)   K-mer counting done. There are 10045512 kmers in total.\n",
      "  0:00:39.134    96M / 13G   INFO    General                 (kmer_index_builder.hpp    : 133)   Merging temporary buckets.\n",
      "  0:00:39.320    96M / 13G   INFO   K-mer Index Building     (kmer_index_builder.hpp    : 314)   Building perfect hash indices\n",
      "  0:00:39.591    96M / 13G   INFO    General                 (kmer_index_builder.hpp    : 150)   Merging final buckets.\n",
      "  0:00:39.798    96M / 13G   INFO   K-mer Index Building     (kmer_index_builder.hpp    : 336)   Index built. Total 4666264 bytes occupied (3.7161 bits per kmer).\n",
      "  0:00:39.806   108M / 13G   INFO   DeBruijnExtensionIndexBu (kmer_extension_index_build:  99)   Building k-mer extensions from k+1-mers\n",
      "  0:00:40.111   108M / 13G   INFO   DeBruijnExtensionIndexBu (kmer_extension_index_build: 103)   Building k-mer extensions from k+1-mers finished.\n",
      "  0:00:40.112   108M / 13G   INFO    General                 (stage.cpp                 : 101)   PROCEDURE == Early tip clipping\n",
      "  0:00:40.112   108M / 13G   INFO    General                 (construction.cpp          : 253)   Early tip clipper length bound set as (RL - K)\n",
      "  0:00:40.112   108M / 13G   INFO   Early tip clipping       (early_simplification.hpp  : 181)   Early tip clipping\n",
      "  0:00:40.555   108M / 13G   INFO   Early tip clipping       (early_simplification.hpp  : 184)   1229890 34-mers were removed by early tip clipper\n",
      "  0:00:40.555   108M / 13G   INFO    General                 (stage.cpp                 : 101)   PROCEDURE == Condensing graph\n",
      "  0:00:40.557   108M / 13G   INFO   UnbranchingPathExtractor (debruijn_graph_constructor: 355)   Extracting unbranching paths\n",
      "  0:00:40.809   116M / 13G   INFO   UnbranchingPathExtractor (debruijn_graph_constructor: 374)   Extracting unbranching paths finished. 139710 sequences extracted\n",
      "  0:00:40.980   116M / 13G   INFO   UnbranchingPathExtractor (debruijn_graph_constructor: 310)   Collecting perfect loops\n",
      "  0:00:41.057   116M / 13G   INFO   UnbranchingPathExtractor (debruijn_graph_constructor: 343)   Collecting perfect loops finished. 274 loops collected\n",
      "  0:00:41.124   116M / 13G   INFO    General                 (stage.cpp                 : 101)   PROCEDURE == Filling coverage indices (PHM)\n",
      "  0:00:41.124   116M / 13G   INFO   K-mer Index Building     (kmer_index_builder.hpp    : 301)   Building kmer index\n",
      "  0:00:41.124   116M / 13G   INFO   K-mer Index Building     (kmer_index_builder.hpp    : 314)   Building perfect hash indices\n",
      "  0:00:41.228   116M / 13G   INFO   K-mer Index Building     (kmer_index_builder.hpp    : 336)   Index built. Total 4670944 bytes occupied (3.71967 bits per kmer).\n",
      "  0:00:41.254   156M / 13G   INFO    General                 (construction.cpp          : 388)   Collecting k-mer coverage information from reads, this takes a while.\n",
      "  0:01:06.797   156M / 13G   INFO    General                 (construction.cpp          : 508)   Filling coverage and flanking coverage from PHM\n",
      "  0:01:07.037   156M / 13G   INFO    General                 (construction.cpp          : 464)   Processed 279893 edges\n",
      "  0:01:07.106   104M / 13G   INFO   StageManager             (stage.cpp                 : 132)   STAGE == EC Threshold Finding\n",
      "  0:01:07.108   104M / 13G   INFO    General                 (kmer_coverage_model.cpp   : 181)   Kmer coverage valley at: 7\n",
      "  0:01:07.109   104M / 13G   INFO    General                 (kmer_coverage_model.cpp   : 201)   K-mer histogram maximum: 12\n",
      "  0:01:07.109   104M / 13G   INFO    General                 (kmer_coverage_model.cpp   : 237)   Estimated median coverage: 13. Coverage mad: 4.4478\n",
      "  0:01:07.109   104M / 13G   INFO    General                 (kmer_coverage_model.cpp   : 259)   Fitting coverage model\n",
      "  0:01:07.127   104M / 13G   INFO    General                 (kmer_coverage_model.cpp   : 295)   ... iteration 2\n",
      "  0:01:07.165   104M / 13G   INFO    General                 (kmer_coverage_model.cpp   : 295)   ... iteration 4\n",
      "  0:01:07.362   104M / 13G   INFO    General                 (kmer_coverage_model.cpp   : 295)   ... iteration 8\n",
      "  0:01:07.603   104M / 13G   INFO    General                 (kmer_coverage_model.cpp   : 309)   Fitted mean coverage: 1. Fitted coverage std. dev: 3.57429e-10\n",
      "  0:01:07.603   104M / 13G   INFO    General                 (kmer_coverage_model.cpp   : 334)   Probability of erroneous kmer at valley: 4.17825e-10\n",
      "  0:01:07.603   104M / 13G   INFO    General                 (kmer_coverage_model.cpp   : 358)   Preliminary threshold calculated as: 7\n",
      "  0:01:07.603   104M / 13G   INFO    General                 (kmer_coverage_model.cpp   : 362)   Threshold adjusted to: 7\n",
      "  0:01:07.603   104M / 13G   INFO    General                 (kmer_coverage_model.cpp   : 375)   Estimated genome size (ignoring repeats): 199380\n",
      "  0:01:07.603   104M / 13G   INFO    General                 (genomic_info_filler.cpp   : 112)   Mean coverage was calculated as 1\n",
      "  0:01:07.603   104M / 13G   INFO    General                 (genomic_info_filler.cpp   : 127)   EC coverage threshold value was calculated as 7\n",
      "  0:01:07.603   104M / 13G   INFO    General                 (genomic_info_filler.cpp   : 128)   Trusted kmer low bound: 0\n",
      "  0:01:07.603   104M / 13G   INFO   StageManager             (stage.cpp                 : 132)   STAGE == Raw Simplification\n",
      "  0:01:07.603   104M / 13G   INFO    General                 (simplification.cpp        : 128)   PROCEDURE == InitialCleaning\n",
      "  0:01:07.603   104M / 13G   INFO    General                 (simplification.cpp        :  68)   Most init cleaning disabled since detected mean 1 was less than activation coverage 10\n",
      "  0:01:07.603   104M / 13G   INFO   Simplification           (parallel_processing.hpp   : 165)   Running Self conjugate edge remover\n",
      "  0:01:07.612   104M / 13G   INFO   Simplification           (parallel_processing.hpp   : 167)   Self conjugate edge remover triggered 0 times\n",
      "  0:01:07.612   104M / 13G   INFO   StageManager             (stage.cpp                 : 132)   STAGE == Simplification\n",
      "  0:01:07.612   104M / 13G   INFO    General                 (simplification.cpp        : 357)   Graph simplification started\n",
      "  0:01:07.612   104M / 13G   INFO    General                 (graph_simplification.hpp  : 634)   Creating parallel br instance\n",
      "  0:01:07.612   104M / 13G   INFO    General                 (simplification.cpp        : 362)   PROCEDURE == Simplification cycle, iteration 1\n",
      "  0:01:07.612   104M / 13G   INFO   Simplification           (parallel_processing.hpp   : 165)   Running Tip clipper\n",
      "  0:01:07.645   100M / 13G   INFO   Simplification           (parallel_processing.hpp   : 167)   Tip clipper triggered 1505 times\n",
      "  0:01:07.645   100M / 13G   INFO   Simplification           (parallel_processing.hpp   : 165)   Running Bulge remover\n",
      "  0:01:08.420   104M / 13G   INFO   Simplification           (parallel_processing.hpp   : 167)   Bulge remover triggered 18491 times\n",
      "  0:01:08.421   104M / 13G   INFO   Simplification           (parallel_processing.hpp   : 165)   Running Low coverage edge remover\n",
      "  0:01:08.436   104M / 13G   INFO   Simplification           (parallel_processing.hpp   : 167)   Low coverage edge remover triggered 19 times\n",
      "  0:01:08.436   104M / 13G   INFO    General                 (simplification.cpp        : 362)   PROCEDURE == Simplification cycle, iteration 2\n",
      "  0:01:08.436   104M / 13G   INFO   Simplification           (parallel_processing.hpp   : 165)   Running Tip clipper\n",
      "  0:01:08.441   104M / 13G   INFO   Simplification           (parallel_processing.hpp   : 167)   Tip clipper triggered 22 times\n",
      "  0:01:08.442   104M / 13G   INFO   Simplification           (parallel_processing.hpp   : 165)   Running Bulge remover\n",
      "  0:01:08.442   104M / 13G   INFO   Simplification           (parallel_processing.hpp   : 167)   Bulge remover triggered 7 times\n",
      "  0:01:08.442   104M / 13G   INFO   Simplification           (parallel_processing.hpp   : 165)   Running Low coverage edge remover\n",
      "  0:01:08.616   104M / 13G   INFO   Simplification           (parallel_processing.hpp   : 167)   Low coverage edge remover triggered 8307 times\n",
      "  0:01:08.616   104M / 13G   INFO    General                 (simplification.cpp        : 362)   PROCEDURE == Simplification cycle, iteration 3\n",
      "  0:01:08.616   104M / 13G   INFO   Simplification           (parallel_processing.hpp   : 165)   Running Tip clipper\n",
      "  0:01:08.619   104M / 13G   INFO   Simplification           (parallel_processing.hpp   : 167)   Tip clipper triggered 71 times\n",
      "  0:01:08.619   104M / 13G   INFO   Simplification           (parallel_processing.hpp   : 165)   Running Bulge remover\n",
      "  0:01:08.738   104M / 13G   INFO   Simplification           (parallel_processing.hpp   : 167)   Bulge remover triggered 1330 times\n",
      "  0:01:08.738   104M / 13G   INFO   Simplification           (parallel_processing.hpp   : 165)   Running Low coverage edge remover\n",
      "  0:01:08.795   104M / 13G   INFO   Simplification           (parallel_processing.hpp   : 167)   Low coverage edge remover triggered 2099 times\n",
      "  0:01:08.795   104M / 13G   INFO    General                 (simplification.cpp        : 362)   PROCEDURE == Simplification cycle, iteration 4\n",
      "  0:01:08.795   104M / 13G   INFO   Simplification           (parallel_processing.hpp   : 165)   Running Tip clipper\n",
      "  0:01:08.796   104M / 13G   INFO   Simplification           (parallel_processing.hpp   : 167)   Tip clipper triggered 6 times\n",
      "  0:01:08.796   104M / 13G   INFO   Simplification           (parallel_processing.hpp   : 165)   Running Bulge remover\n",
      "  0:01:08.850   104M / 13G   INFO   Simplification           (parallel_processing.hpp   : 167)   Bulge remover triggered 608 times\n",
      "  0:01:08.850   104M / 13G   INFO   Simplification           (parallel_processing.hpp   : 165)   Running Low coverage edge remover\n",
      "  0:01:08.861   104M / 13G   INFO   Simplification           (parallel_processing.hpp   : 167)   Low coverage edge remover triggered 526 times\n",
      "  0:01:08.861   104M / 13G   INFO    General                 (simplification.cpp        : 362)   PROCEDURE == Simplification cycle, iteration 5\n",
      "  0:01:08.861   104M / 13G   INFO   Simplification           (parallel_processing.hpp   : 165)   Running Tip clipper\n",
      "  0:01:08.862   104M / 13G   INFO   Simplification           (parallel_processing.hpp   : 167)   Tip clipper triggered 1 times\n",
      "  0:01:08.862   104M / 13G   INFO   Simplification           (parallel_processing.hpp   : 165)   Running Bulge remover\n",
      "  0:01:08.880   104M / 13G   INFO   Simplification           (parallel_processing.hpp   : 167)   Bulge remover triggered 182 times\n",
      "  0:01:08.880   104M / 13G   INFO   Simplification           (parallel_processing.hpp   : 165)   Running Low coverage edge remover\n",
      "  0:01:08.901   104M / 13G   INFO   Simplification           (parallel_processing.hpp   : 167)   Low coverage edge remover triggered 682 times\n",
      "  0:01:08.901   104M / 13G   INFO    General                 (simplification.cpp        : 362)   PROCEDURE == Simplification cycle, iteration 6\n",
      "  0:01:08.901   104M / 13G   INFO   Simplification           (parallel_processing.hpp   : 165)   Running Tip clipper\n",
      "  0:01:08.901   104M / 13G   INFO   Simplification           (parallel_processing.hpp   : 167)   Tip clipper triggered 0 times\n",
      "  0:01:08.901   104M / 13G   INFO   Simplification           (parallel_processing.hpp   : 165)   Running Bulge remover\n",
      "  0:01:08.926   104M / 13G   INFO   Simplification           (parallel_processing.hpp   : 167)   Bulge remover triggered 274 times\n",
      "  0:01:08.926   104M / 13G   INFO   Simplification           (parallel_processing.hpp   : 165)   Running Low coverage edge remover\n",
      "  0:01:08.936   104M / 13G   INFO   Simplification           (parallel_processing.hpp   : 167)   Low coverage edge remover triggered 348 times\n",
      "  0:01:08.936   104M / 13G   INFO    General                 (simplification.cpp        : 362)   PROCEDURE == Simplification cycle, iteration 7\n",
      "  0:01:08.936   104M / 13G   INFO   Simplification           (parallel_processing.hpp   : 165)   Running Tip clipper\n",
      "  0:01:08.936   104M / 13G   INFO   Simplification           (parallel_processing.hpp   : 167)   Tip clipper triggered 0 times\n",
      "  0:01:08.936   104M / 13G   INFO   Simplification           (parallel_processing.hpp   : 165)   Running Bulge remover\n",
      "  0:01:08.950   104M / 13G   INFO   Simplification           (parallel_processing.hpp   : 167)   Bulge remover triggered 157 times\n",
      "  0:01:08.950   104M / 13G   INFO   Simplification           (parallel_processing.hpp   : 165)   Running Low coverage edge remover\n",
      "  0:01:08.954   104M / 13G   INFO   Simplification           (parallel_processing.hpp   : 167)   Low coverage edge remover triggered 190 times\n",
      "  0:01:08.954   104M / 13G   INFO    General                 (simplification.cpp        : 362)   PROCEDURE == Simplification cycle, iteration 8\n",
      "  0:01:08.955   104M / 13G   INFO   Simplification           (parallel_processing.hpp   : 165)   Running Tip clipper\n",
      "  0:01:08.955   104M / 13G   INFO   Simplification           (parallel_processing.hpp   : 167)   Tip clipper triggered 0 times\n",
      "  0:01:08.955   104M / 13G   INFO   Simplification           (parallel_processing.hpp   : 165)   Running Bulge remover\n",
      "  0:01:08.964   104M / 13G   INFO   Simplification           (parallel_processing.hpp   : 167)   Bulge remover triggered 98 times\n",
      "  0:01:08.964   104M / 13G   INFO   Simplification           (parallel_processing.hpp   : 165)   Running Low coverage edge remover\n",
      "  0:01:08.968   104M / 13G   INFO   Simplification           (parallel_processing.hpp   : 167)   Low coverage edge remover triggered 211 times\n",
      "  0:01:08.968   104M / 13G   INFO    General                 (simplification.cpp        : 362)   PROCEDURE == Simplification cycle, iteration 9\n",
      "  0:01:08.968   104M / 13G   INFO   Simplification           (parallel_processing.hpp   : 165)   Running Tip clipper\n",
      "  0:01:08.968   104M / 13G   INFO   Simplification           (parallel_processing.hpp   : 167)   Tip clipper triggered 0 times\n",
      "  0:01:08.968   104M / 13G   INFO   Simplification           (parallel_processing.hpp   : 165)   Running Bulge remover\n",
      "  0:01:08.978   104M / 13G   INFO   Simplification           (parallel_processing.hpp   : 167)   Bulge remover triggered 111 times\n",
      "  0:01:08.978   104M / 13G   INFO   Simplification           (parallel_processing.hpp   : 165)   Running Low coverage edge remover\n",
      "  0:01:08.981   104M / 13G   INFO   Simplification           (parallel_processing.hpp   : 167)   Low coverage edge remover triggered 153 times\n",
      "  0:01:08.981   104M / 13G   INFO    General                 (simplification.cpp        : 362)   PROCEDURE == Simplification cycle, iteration 10\n",
      "  0:01:08.981   104M / 13G   INFO   Simplification           (parallel_processing.hpp   : 165)   Running Tip clipper\n",
      "  0:01:08.981   104M / 13G   INFO   Simplification           (parallel_processing.hpp   : 167)   Tip clipper triggered 0 times\n",
      "  0:01:08.981   104M / 13G   INFO   Simplification           (parallel_processing.hpp   : 165)   Running Bulge remover\n",
      "  0:01:08.989   104M / 13G   INFO   Simplification           (parallel_processing.hpp   : 167)   Bulge remover triggered 104 times\n",
      "  0:01:08.989   104M / 13G   INFO   Simplification           (parallel_processing.hpp   : 165)   Running Low coverage edge remover\n",
      "  0:01:08.992   104M / 13G   INFO   Simplification           (parallel_processing.hpp   : 167)   Low coverage edge remover triggered 148 times\n",
      "  0:01:08.992   104M / 13G   INFO    General                 (simplification.cpp        : 362)   PROCEDURE == Simplification cycle, iteration 11\n",
      "  0:01:08.992   104M / 13G   INFO   Simplification           (parallel_processing.hpp   : 165)   Running Tip clipper\n",
      "  0:01:08.997   104M / 13G   INFO   Simplification           (parallel_processing.hpp   : 167)   Tip clipper triggered 2 times\n",
      "  0:01:08.997   104M / 13G   INFO   Simplification           (parallel_processing.hpp   : 165)   Running Bulge remover\n",
      "  0:01:09.017   104M / 13G   INFO   Simplification           (parallel_processing.hpp   : 167)   Bulge remover triggered 95 times\n",
      "  0:01:09.017   104M / 13G   INFO   Simplification           (parallel_processing.hpp   : 165)   Running Low coverage edge remover\n",
      "  0:01:09.022   104M / 13G   INFO   Simplification           (parallel_processing.hpp   : 167)   Low coverage edge remover triggered 0 times\n",
      "  0:01:09.022   104M / 13G   INFO    General                 (simplification.cpp        : 362)   PROCEDURE == Simplification cycle, iteration 12\n",
      "  0:01:09.022   104M / 13G   INFO   Simplification           (parallel_processing.hpp   : 165)   Running Tip clipper\n",
      "  0:01:09.023   104M / 13G   INFO   Simplification           (parallel_processing.hpp   : 167)   Tip clipper triggered 0 times\n",
      "  0:01:09.023   104M / 13G   INFO   Simplification           (parallel_processing.hpp   : 165)   Running Bulge remover\n",
      "  0:01:09.024   104M / 13G   INFO   Simplification           (parallel_processing.hpp   : 167)   Bulge remover triggered 0 times\n",
      "  0:01:09.024   104M / 13G   INFO   Simplification           (parallel_processing.hpp   : 165)   Running Low coverage edge remover\n",
      "  0:01:09.024   104M / 13G   INFO   Simplification           (parallel_processing.hpp   : 167)   Low coverage edge remover triggered 0 times\n",
      "  0:01:09.024   104M / 13G   INFO   StageManager             (stage.cpp                 : 132)   STAGE == Simplification Cleanup\n",
      "  0:01:09.024   104M / 13G   INFO    General                 (simplification.cpp        : 196)   PROCEDURE == Post simplification\n",
      "  0:01:09.024   104M / 13G   INFO    General                 (graph_simplification.hpp  : 453)   Disconnection of relatively low covered edges disabled\n",
      "  0:01:09.024   104M / 13G   INFO    General                 (graph_simplification.hpp  : 489)   Complex tip clipping disabled\n",
      "  0:01:09.024   104M / 13G   INFO    General                 (graph_simplification.hpp  : 634)   Creating parallel br instance\n",
      "  0:01:09.024   104M / 13G   INFO   Simplification           (parallel_processing.hpp   : 165)   Running Tip clipper\n",
      "  0:01:09.029   104M / 13G   INFO   Simplification           (parallel_processing.hpp   : 167)   Tip clipper triggered 0 times\n",
      "  0:01:09.029   104M / 13G   INFO   Simplification           (parallel_processing.hpp   : 165)   Running Bulge remover\n",
      "  0:01:09.040   104M / 13G   INFO   Simplification           (parallel_processing.hpp   : 167)   Bulge remover triggered 0 times\n",
      "  0:01:09.040   104M / 13G   INFO   Simplification           (parallel_processing.hpp   : 165)   Running Tip clipper\n",
      "  0:01:09.044   104M / 13G   INFO   Simplification           (parallel_processing.hpp   : 167)   Tip clipper triggered 0 times\n",
      "  0:01:09.045   104M / 13G   INFO   Simplification           (parallel_processing.hpp   : 165)   Running Bulge remover\n",
      "  0:01:09.055   104M / 13G   INFO   Simplification           (parallel_processing.hpp   : 167)   Bulge remover triggered 0 times\n",
      "  0:01:09.055   104M / 13G   INFO    General                 (simplification.cpp        : 330)   Disrupting self-conjugate edges\n",
      "  0:01:09.079   104M / 13G   INFO   Simplification           (parallel_processing.hpp   : 165)   Running Removing isolated edges\n",
      "  0:01:09.156   104M / 13G   INFO   Simplification           (parallel_processing.hpp   : 167)   Removing isolated edges triggered 27454 times\n",
      "  0:01:09.156   104M / 13G   INFO    General                 (simplification.cpp        : 470)   Counting average coverage\n",
      "  0:01:09.161   104M / 13G   INFO    General                 (simplification.cpp        : 476)   Average coverage = 474.809\n",
      "  0:01:09.161   104M / 13G   INFO   StageManager             (stage.cpp                 : 132)   STAGE == Contig Output\n",
      "  0:01:09.161   104M / 13G   INFO    General                 (contig_output_stage.cpp   :  40)   Writing GFA to /mnt/d/linux/W/MoonNGS/raw/20180813Novogene16species_Human_Methyl/1.rawdata/MNH08365/Tychus/spades//K33/assembly_graph_with_scaffolds.gfa\n",
      "  0:01:09.224   104M / 13G   INFO    General                 (contig_output.hpp         :  22)   Outputting contigs to /mnt/d/linux/W/MoonNGS/raw/20180813Novogene16species_Human_Methyl/1.rawdata/MNH08365/Tychus/spades//K33/before_rr.fasta\n",
      "  0:01:09.311   104M / 13G   INFO    General                 (contig_output_stage.cpp   :  51)   Outputting FastG graph to /mnt/d/linux/W/MoonNGS/raw/20180813Novogene16species_Human_Methyl/1.rawdata/MNH08365/Tychus/spades//K33/assembly_graph.fastg\n",
      "  0:01:09.555   104M / 13G   INFO    General                 (contig_output.hpp         :  22)   Outputting contigs to /mnt/d/linux/W/MoonNGS/raw/20180813Novogene16species_Human_Methyl/1.rawdata/MNH08365/Tychus/spades//K33/simplified_contigs.fasta\n",
      "  0:01:09.643   104M / 13G   INFO    General                 (contig_output.hpp         :  22)   Outputting contigs to /mnt/d/linux/W/MoonNGS/raw/20180813Novogene16species_Human_Methyl/1.rawdata/MNH08365/Tychus/spades//K33/final_contigs.fasta\n",
      "  0:01:09.730   104M / 13G   INFO   StageManager             (stage.cpp                 : 132)   STAGE == Contig Output\n",
      "  0:01:09.730   104M / 13G   INFO    General                 (contig_output_stage.cpp   :  40)   Writing GFA to /mnt/d/linux/W/MoonNGS/raw/20180813Novogene16species_Human_Methyl/1.rawdata/MNH08365/Tychus/spades//K33/assembly_graph_with_scaffolds.gfa\n",
      "  0:01:09.793   104M / 13G   INFO    General                 (contig_output.hpp         :  22)   Outputting contigs to /mnt/d/linux/W/MoonNGS/raw/20180813Novogene16species_Human_Methyl/1.rawdata/MNH08365/Tychus/spades//K33/before_rr.fasta\n",
      "  0:01:09.880   104M / 13G   INFO    General                 (contig_output_stage.cpp   :  51)   Outputting FastG graph to /mnt/d/linux/W/MoonNGS/raw/20180813Novogene16species_Human_Methyl/1.rawdata/MNH08365/Tychus/spades//K33/assembly_graph.fastg\n",
      "  0:01:10.133   104M / 13G   INFO    General                 (contig_output.hpp         :  22)   Outputting contigs to /mnt/d/linux/W/MoonNGS/raw/20180813Novogene16species_Human_Methyl/1.rawdata/MNH08365/Tychus/spades//K33/simplified_contigs.fasta\n",
      "  0:01:10.226   104M / 13G   INFO    General                 (contig_output.hpp         :  22)   Outputting contigs to /mnt/d/linux/W/MoonNGS/raw/20180813Novogene16species_Human_Methyl/1.rawdata/MNH08365/Tychus/spades//K33/final_contigs.fasta\n",
      "  0:01:10.318   104M / 13G   INFO    General                 (launch.hpp                : 149)   SPAdes finished\n",
      "  0:01:10.356    96M / 13G   INFO    General                 (main.cpp                  : 109)   Assembling time: 0 hours 1 minutes 10 seconds\n",
      "\n",
      "== Running assembler: K55\n",
      "\n",
      "  0:00:00.000     4M / 4M    INFO    General                 (main.cpp                  :  74)   Loaded config from /mnt/d/linux/W/MoonNGS/raw/20180813Novogene16species_Human_Methyl/1.rawdata/MNH08365/Tychus/spades/K55/configs/config.info\n",
      "  0:00:00.000     4M / 4M    INFO    General                 (memory_limit.cpp          :  49)   Memory limit set to 62 Gb\n",
      "  0:00:00.000     4M / 4M    INFO    General                 (main.cpp                  :  87)   Starting SPAdes, built from N/A, git revision N/A\n",
      "  0:00:00.000     4M / 4M    INFO    General                 (main.cpp                  :  88)   Maximum k-mer length: 128\n",
      "  0:00:00.000     4M / 4M    INFO    General                 (main.cpp                  :  89)   Assembling dataset (/mnt/d/linux/W/MoonNGS/raw/20180813Novogene16species_Human_Methyl/1.rawdata/MNH08365/Tychus/spades/dataset.info) with K=55\n",
      "  0:00:00.000     4M / 4M    INFO    General                 (main.cpp                  :  90)   Maximum # of threads to use (adjusted due to OMP capabilities): 24\n",
      "  0:00:00.000     4M / 4M    INFO    General                 (launch.hpp                :  51)   SPAdes started\n",
      "  0:00:00.000     4M / 4M    INFO    General                 (launch.hpp                :  58)   Starting from stage: construction\n",
      "  0:00:00.000     4M / 4M    INFO    General                 (launch.hpp                :  65)   Two-step RR enabled: 0\n",
      "  0:00:00.000     4M / 4M    INFO    General                 (launch.hpp                :  76)   Will need read mapping, kmer mapper will be attached\n",
      "  0:00:00.000     4M / 4M    INFO   StageManager             (stage.cpp                 : 132)   STAGE == de Bruijn graph construction\n",
      "  0:00:00.000     4M / 4M    INFO    General                 (read_converter.hpp        :  59)   Binary reads detected\n",
      "  0:00:00.001     4M / 4M    INFO    General                 (construction.cpp          : 111)   Max read length 150\n",
      "  0:00:00.001     4M / 4M    INFO    General                 (construction.cpp          : 117)   Average read length 146.615\n",
      "  0:00:00.001     4M / 4M    INFO    General                 (stage.cpp                 : 101)   PROCEDURE == k+1-mer counting\n",
      "  0:00:00.001     4M / 4M    INFO    General                 (kmer_index_builder.hpp    : 117)   Splitting kmer instances into 576 files using 24 threads. This might take a while.\n",
      "  0:00:00.002     4M / 4M    INFO    General                 (file_limit.hpp            :  32)   Open file limit set to 1024\n",
      "  0:00:00.002     4M / 4M    INFO    General                 (kmer_splitters.hpp        :  89)   Memory available for splitting buffers: 0.861057 Gb\n",
      "  0:00:00.002     4M / 4M    INFO    General                 (kmer_splitters.hpp        :  97)   Using cell size of 58254\n",
      "  0:00:11.852    13G / 13G   INFO    General                 (kmer_splitters.hpp        : 289)   Processed 16800732 reads\n",
      "  0:00:24.500    13G / 13G   INFO    General                 (kmer_splitters.hpp        : 289)   Processed 33519373 reads\n",
      "  0:00:28.072    13G / 13G   INFO    General                 (kmer_splitters.hpp        : 289)   Processed 39380436 reads\n",
      "  0:00:28.072    13G / 13G   INFO    General                 (kmer_splitters.hpp        : 295)   Adding contigs from previous K\n",
      "  0:00:30.380    96M / 13G   INFO    General                 (kmer_splitters.hpp        : 308)   Used 39380436 reads\n",
      "  0:00:30.380    96M / 13G   INFO    General                 (kmer_index_builder.hpp    : 120)   Starting k-mer counting.\n",
      "  0:00:30.797    96M / 13G   INFO    General                 (kmer_index_builder.hpp    : 127)   K-mer counting done. There are 9850569 kmers in total.\n",
      "  0:00:30.797    96M / 13G   INFO    General                 (kmer_index_builder.hpp    : 133)   Merging temporary buckets.\n",
      "  0:00:31.019    96M / 13G   INFO    General                 (stage.cpp                 : 101)   PROCEDURE == Extension index construction\n",
      "  0:00:31.019    96M / 13G   INFO   K-mer Index Building     (kmer_index_builder.hpp    : 301)   Building kmer index\n",
      "  0:00:31.019    96M / 13G   INFO    General                 (kmer_index_builder.hpp    : 117)   Splitting kmer instances into 384 files using 24 threads. This might take a while.\n",
      "  0:00:31.019    96M / 13G   INFO    General                 (file_limit.hpp            :  32)   Open file limit set to 1024\n",
      "  0:00:31.019    96M / 13G   INFO    General                 (kmer_splitters.hpp        :  89)   Memory available for splitting buffers: 0.859809 Gb\n",
      "  0:00:31.019    96M / 13G   INFO    General                 (kmer_splitters.hpp        :  97)   Using cell size of 87381\n",
      "  0:00:31.581    13G / 13G   INFO    General                 (kmer_splitters.hpp        : 380)   Processed 9850569 kmers\n",
      "  0:00:31.582    13G / 13G   INFO    General                 (kmer_splitters.hpp        : 385)   Used 9850569 kmers.\n",
      "  0:00:31.761    96M / 13G   INFO    General                 (kmer_index_builder.hpp    : 120)   Starting k-mer counting.\n",
      "  0:00:31.916    96M / 13G   INFO    General                 (kmer_index_builder.hpp    : 127)   K-mer counting done. There are 9866601 kmers in total.\n",
      "  0:00:31.916    96M / 13G   INFO    General                 (kmer_index_builder.hpp    : 133)   Merging temporary buckets.\n",
      "  0:00:32.099    96M / 13G   INFO   K-mer Index Building     (kmer_index_builder.hpp    : 314)   Building perfect hash indices\n",
      "  0:00:32.422    96M / 13G   INFO    General                 (kmer_index_builder.hpp    : 150)   Merging final buckets.\n",
      "  0:00:32.673    96M / 13G   INFO   K-mer Index Building     (kmer_index_builder.hpp    : 336)   Index built. Total 4583376 bytes occupied (3.71628 bits per kmer).\n",
      "  0:00:32.684   108M / 13G   INFO   DeBruijnExtensionIndexBu (kmer_extension_index_build:  99)   Building k-mer extensions from k+1-mers\n",
      "  0:00:33.079   108M / 13G   INFO   DeBruijnExtensionIndexBu (kmer_extension_index_build: 103)   Building k-mer extensions from k+1-mers finished.\n",
      "  0:00:33.081   108M / 13G   INFO    General                 (stage.cpp                 : 101)   PROCEDURE == Condensing graph\n",
      "  0:00:33.103   108M / 13G   INFO   UnbranchingPathExtractor (debruijn_graph_constructor: 355)   Extracting unbranching paths\n",
      "  0:00:33.446   124M / 13G   INFO   UnbranchingPathExtractor (debruijn_graph_constructor: 374)   Extracting unbranching paths finished. 410778 sequences extracted\n",
      "  0:00:33.646   124M / 13G   INFO   UnbranchingPathExtractor (debruijn_graph_constructor: 310)   Collecting perfect loops\n",
      "  0:00:33.709   124M / 13G   INFO   UnbranchingPathExtractor (debruijn_graph_constructor: 343)   Collecting perfect loops finished. 24 loops collected\n",
      "  0:00:33.924   220M / 13G   INFO    General                 (stage.cpp                 : 101)   PROCEDURE == Filling coverage indices (PHM)\n",
      "  0:00:33.924   220M / 13G   INFO   K-mer Index Building     (kmer_index_builder.hpp    : 301)   Building kmer index\n",
      "  0:00:33.924   220M / 13G   INFO   K-mer Index Building     (kmer_index_builder.hpp    : 314)   Building perfect hash indices\n",
      "  0:00:34.025   220M / 13G   INFO   K-mer Index Building     (kmer_index_builder.hpp    : 336)   Index built. Total 4580264 bytes occupied (3.7198 bits per kmer).\n",
      "  0:00:34.052   260M / 13G   INFO    General                 (construction.cpp          : 388)   Collecting k-mer coverage information from reads, this takes a while.\n",
      "  0:00:56.591   260M / 13G   INFO    General                 (construction.cpp          : 508)   Filling coverage and flanking coverage from PHM\n",
      "  0:00:56.904   260M / 13G   INFO    General                 (construction.cpp          : 464)   Processed 821575 edges\n",
      "  0:00:56.957   208M / 13G   INFO   StageManager             (stage.cpp                 : 132)   STAGE == EC Threshold Finding\n",
      "  0:00:56.959   208M / 13G   INFO    General                 (kmer_coverage_model.cpp   : 181)   Kmer coverage valley at: 7\n",
      "  0:00:56.959   208M / 13G   INFO    General                 (kmer_coverage_model.cpp   : 201)   K-mer histogram maximum: 10\n",
      "  0:00:56.959   208M / 13G   INFO    General                 (kmer_coverage_model.cpp   : 237)   Estimated median coverage: 11. Coverage mad: 4.4478\n",
      "  0:00:56.959   208M / 13G   INFO    General                 (kmer_coverage_model.cpp   : 259)   Fitting coverage model\n",
      "  0:00:56.973   208M / 13G   INFO    General                 (kmer_coverage_model.cpp   : 295)   ... iteration 2\n",
      "  0:00:57.006   208M / 13G   INFO    General                 (kmer_coverage_model.cpp   : 295)   ... iteration 4\n",
      "  0:00:57.087   208M / 13G   INFO    General                 (kmer_coverage_model.cpp   : 295)   ... iteration 8\n",
      "  0:00:57.225   208M / 13G   INFO    General                 (kmer_coverage_model.cpp   : 309)   Fitted mean coverage: 1. Fitted coverage std. dev: 3.71239e-09\n",
      "  0:00:57.225   208M / 13G   INFO    General                 (kmer_coverage_model.cpp   : 334)   Probability of erroneous kmer at valley: 8.3008e-09\n",
      "  0:00:57.225   208M / 13G   INFO    General                 (kmer_coverage_model.cpp   : 358)   Preliminary threshold calculated as: 7\n",
      "  0:00:57.226   208M / 13G   INFO    General                 (kmer_coverage_model.cpp   : 362)   Threshold adjusted to: 7\n",
      "  0:00:57.226   208M / 13G   INFO    General                 (kmer_coverage_model.cpp   : 375)   Estimated genome size (ignoring repeats): 278744\n",
      "  0:00:57.226   208M / 13G   INFO    General                 (genomic_info_filler.cpp   : 112)   Mean coverage was calculated as 1\n",
      "  0:00:57.226   208M / 13G   INFO    General                 (genomic_info_filler.cpp   : 127)   EC coverage threshold value was calculated as 7\n",
      "  0:00:57.226   208M / 13G   INFO    General                 (genomic_info_filler.cpp   : 128)   Trusted kmer low bound: 0\n",
      "  0:00:57.226   208M / 13G   INFO   StageManager             (stage.cpp                 : 132)   STAGE == Gap Closer\n",
      "  0:00:57.226   208M / 13G   INFO    General                 (graph_pack.hpp            : 101)   Index refill\n",
      "  0:00:57.226   208M / 13G   INFO   K-mer Index Building     (kmer_index_builder.hpp    : 301)   Building kmer index\n",
      "  0:00:57.226   208M / 13G   INFO    General                 (kmer_index_builder.hpp    : 117)   Splitting kmer instances into 384 files using 24 threads. This might take a while.\n",
      "  0:00:57.226   208M / 13G   INFO    General                 (file_limit.hpp            :  32)   Open file limit set to 1024\n",
      "  0:00:57.226   208M / 13G   INFO    General                 (kmer_splitters.hpp        :  89)   Memory available for splitting buffers: 0.85829 Gb\n",
      "  0:00:57.226   208M / 13G   INFO    General                 (kmer_splitters.hpp        :  97)   Using cell size of 87381\n",
      "  0:00:57.755    13G / 13G   INFO    General                 (edge_index_builders.hpp   :  77)   Processed 821575 edges\n",
      "  0:00:57.755    13G / 13G   INFO    General                 (edge_index_builders.hpp   :  82)   Used 821575 sequences.\n",
      "  0:00:57.875   208M / 13G   INFO    General                 (kmer_index_builder.hpp    : 120)   Starting k-mer counting.\n",
      "  0:00:58.032   208M / 13G   INFO    General                 (kmer_index_builder.hpp    : 127)   K-mer counting done. There are 9850569 kmers in total.\n",
      "  0:00:58.032   208M / 13G   INFO    General                 (kmer_index_builder.hpp    : 133)   Merging temporary buckets.\n",
      "  0:00:58.223   208M / 13G   INFO   K-mer Index Building     (kmer_index_builder.hpp    : 314)   Building perfect hash indices\n",
      "  0:00:58.478   208M / 13G   INFO    General                 (kmer_index_builder.hpp    : 150)   Merging final buckets.\n",
      "  0:00:58.690   208M / 13G   INFO   K-mer Index Building     (kmer_index_builder.hpp    : 336)   Index built. Total 4575992 bytes occupied (3.71633 bits per kmer).\n",
      "  0:00:58.824   436M / 13G   INFO    General                 (edge_index_builders.hpp   : 107)   Collecting edge information from graph, this takes a while.\n",
      "  0:00:59.021   436M / 13G   INFO    General                 (edge_index.hpp            :  92)   Index refilled\n",
      "  0:00:59.021   436M / 13G   INFO    General                 (gap_closer.cpp            : 159)   Preparing shift maps\n",
      "  0:00:59.501   468M / 13G   INFO    General                 (gap_closer.cpp            : 119)   Processing paired reads (takes a while)\n",
      "  0:01:06.452   468M / 13G   INFO    General                 (gap_closer.cpp            : 138)   Used 9842709 paired reads\n",
      "  0:01:06.452   468M / 13G   INFO    General                 (gap_closer.cpp            : 140)   Merging paired indices\n",
      "  0:01:06.589   440M / 13G   INFO   GapCloser                (gap_closer.cpp            : 346)   Closing short gaps\n",
      "  0:01:07.268   440M / 13G   INFO   GapCloser                (gap_closer.cpp            : 380)   Closing short gaps complete: filled 99 gaps after checking 2536 candidates\n",
      "  0:01:07.324   436M / 13G   INFO   StageManager             (stage.cpp                 : 132)   STAGE == Raw Simplification\n",
      "  0:01:07.353   208M / 13G   INFO    General                 (simplification.cpp        : 128)   PROCEDURE == InitialCleaning\n",
      "  0:01:07.353   208M / 13G   INFO    General                 (simplification.cpp        :  68)   Most init cleaning disabled since detected mean 1 was less than activation coverage 10\n",
      "  0:01:07.353   208M / 13G   INFO   Simplification           (parallel_processing.hpp   : 165)   Running Self conjugate edge remover\n",
      "  0:01:07.379   208M / 13G   INFO   Simplification           (parallel_processing.hpp   : 167)   Self conjugate edge remover triggered 0 times\n",
      "  0:01:07.379   208M / 13G   INFO   StageManager             (stage.cpp                 : 132)   STAGE == Simplification\n",
      "  0:01:07.379   208M / 13G   INFO    General                 (simplification.cpp        : 357)   Graph simplification started\n",
      "  0:01:07.379   208M / 13G   INFO    General                 (graph_simplification.hpp  : 634)   Creating parallel br instance\n",
      "  0:01:07.379   208M / 13G   INFO    General                 (simplification.cpp        : 362)   PROCEDURE == Simplification cycle, iteration 1\n",
      "  0:01:07.379   208M / 13G   INFO   Simplification           (parallel_processing.hpp   : 165)   Running Tip clipper\n",
      "  0:01:09.060   216M / 13G   INFO   Simplification           (parallel_processing.hpp   : 167)   Tip clipper triggered 156438 times\n",
      "  0:01:09.060   216M / 13G   INFO   Simplification           (parallel_processing.hpp   : 165)   Running Bulge remover\n",
      "  0:01:11.283   272M / 13G   INFO   Simplification           (parallel_processing.hpp   : 167)   Bulge remover triggered 12207 times\n",
      "  0:01:11.283   272M / 13G   INFO   Simplification           (parallel_processing.hpp   : 165)   Running Low coverage edge remover\n",
      "  0:01:11.316   272M / 13G   INFO   Simplification           (parallel_processing.hpp   : 167)   Low coverage edge remover triggered 9 times\n",
      "  0:01:11.316   272M / 13G   INFO    General                 (simplification.cpp        : 362)   PROCEDURE == Simplification cycle, iteration 2\n",
      "  0:01:11.316   272M / 13G   INFO   Simplification           (parallel_processing.hpp   : 165)   Running Tip clipper\n",
      "  0:01:11.320   272M / 13G   INFO   Simplification           (parallel_processing.hpp   : 167)   Tip clipper triggered 21 times\n",
      "  0:01:11.320   272M / 13G   INFO   Simplification           (parallel_processing.hpp   : 165)   Running Bulge remover\n",
      "  0:01:11.321   272M / 13G   INFO   Simplification           (parallel_processing.hpp   : 167)   Bulge remover triggered 2 times\n",
      "  0:01:11.321   272M / 13G   INFO   Simplification           (parallel_processing.hpp   : 165)   Running Low coverage edge remover\n",
      "  0:01:11.440   272M / 13G   INFO   Simplification           (parallel_processing.hpp   : 167)   Low coverage edge remover triggered 4366 times\n",
      "  0:01:11.440   272M / 13G   INFO    General                 (simplification.cpp        : 362)   PROCEDURE == Simplification cycle, iteration 3\n",
      "  0:01:11.440   272M / 13G   INFO   Simplification           (parallel_processing.hpp   : 165)   Running Tip clipper\n",
      "  0:01:11.444   272M / 13G   INFO   Simplification           (parallel_processing.hpp   : 167)   Tip clipper triggered 166 times\n",
      "  0:01:11.444   272M / 13G   INFO   Simplification           (parallel_processing.hpp   : 165)   Running Bulge remover\n",
      "  0:01:11.755   276M / 13G   INFO   Simplification           (parallel_processing.hpp   : 167)   Bulge remover triggered 1220 times\n",
      "  0:01:11.755   276M / 13G   INFO   Simplification           (parallel_processing.hpp   : 165)   Running Low coverage edge remover\n",
      "  0:01:11.799   280M / 13G   INFO   Simplification           (parallel_processing.hpp   : 167)   Low coverage edge remover triggered 1140 times\n",
      "  0:01:11.799   280M / 13G   INFO    General                 (simplification.cpp        : 362)   PROCEDURE == Simplification cycle, iteration 4\n",
      "  0:01:11.800   280M / 13G   INFO   Simplification           (parallel_processing.hpp   : 165)   Running Tip clipper\n",
      "  0:01:11.801   276M / 13G   INFO   Simplification           (parallel_processing.hpp   : 167)   Tip clipper triggered 22 times\n",
      "  0:01:11.801   276M / 13G   INFO   Simplification           (parallel_processing.hpp   : 165)   Running Bulge remover\n",
      "  0:01:11.922   280M / 13G   INFO   Simplification           (parallel_processing.hpp   : 167)   Bulge remover triggered 484 times\n",
      "  0:01:11.922   280M / 13G   INFO   Simplification           (parallel_processing.hpp   : 165)   Running Low coverage edge remover\n",
      "  0:01:11.935   280M / 13G   INFO   Simplification           (parallel_processing.hpp   : 167)   Low coverage edge remover triggered 396 times\n",
      "  0:01:11.935   280M / 13G   INFO    General                 (simplification.cpp        : 362)   PROCEDURE == Simplification cycle, iteration 5\n",
      "  0:01:11.935   280M / 13G   INFO   Simplification           (parallel_processing.hpp   : 165)   Running Tip clipper\n",
      "  0:01:11.935   280M / 13G   INFO   Simplification           (parallel_processing.hpp   : 167)   Tip clipper triggered 4 times\n",
      "  0:01:11.935   280M / 13G   INFO   Simplification           (parallel_processing.hpp   : 165)   Running Bulge remover\n",
      "  0:01:11.989   280M / 13G   INFO   Simplification           (parallel_processing.hpp   : 167)   Bulge remover triggered 210 times\n",
      "  0:01:11.989   280M / 13G   INFO   Simplification           (parallel_processing.hpp   : 165)   Running Low coverage edge remover\n",
      "  0:01:12.004   280M / 13G   INFO   Simplification           (parallel_processing.hpp   : 167)   Low coverage edge remover triggered 396 times\n",
      "  0:01:12.005   280M / 13G   INFO    General                 (simplification.cpp        : 362)   PROCEDURE == Simplification cycle, iteration 6\n",
      "  0:01:12.005   280M / 13G   INFO   Simplification           (parallel_processing.hpp   : 165)   Running Tip clipper\n",
      "  0:01:12.005   280M / 13G   INFO   Simplification           (parallel_processing.hpp   : 167)   Tip clipper triggered 0 times\n",
      "  0:01:12.005   280M / 13G   INFO   Simplification           (parallel_processing.hpp   : 165)   Running Bulge remover\n",
      "  0:01:12.059   280M / 13G   INFO   Simplification           (parallel_processing.hpp   : 167)   Bulge remover triggered 223 times\n",
      "  0:01:12.059   280M / 13G   INFO   Simplification           (parallel_processing.hpp   : 165)   Running Low coverage edge remover\n",
      "  0:01:12.068   280M / 13G   INFO   Simplification           (parallel_processing.hpp   : 167)   Low coverage edge remover triggered 251 times\n",
      "  0:01:12.068   280M / 13G   INFO    General                 (simplification.cpp        : 362)   PROCEDURE == Simplification cycle, iteration 7\n",
      "  0:01:12.068   280M / 13G   INFO   Simplification           (parallel_processing.hpp   : 165)   Running Tip clipper\n",
      "  0:01:12.069   280M / 13G   INFO   Simplification           (parallel_processing.hpp   : 167)   Tip clipper triggered 1 times\n",
      "  0:01:12.069   280M / 13G   INFO   Simplification           (parallel_processing.hpp   : 165)   Running Bulge remover\n",
      "  0:01:12.098   284M / 13G   INFO   Simplification           (parallel_processing.hpp   : 167)   Bulge remover triggered 144 times\n",
      "  0:01:12.098   284M / 13G   INFO   Simplification           (parallel_processing.hpp   : 165)   Running Low coverage edge remover\n",
      "  0:01:12.104   284M / 13G   INFO   Simplification           (parallel_processing.hpp   : 167)   Low coverage edge remover triggered 150 times\n",
      "  0:01:12.104   284M / 13G   INFO    General                 (simplification.cpp        : 362)   PROCEDURE == Simplification cycle, iteration 8\n",
      "  0:01:12.104   284M / 13G   INFO   Simplification           (parallel_processing.hpp   : 165)   Running Tip clipper\n",
      "  0:01:12.104   284M / 13G   INFO   Simplification           (parallel_processing.hpp   : 167)   Tip clipper triggered 0 times\n",
      "  0:01:12.104   284M / 13G   INFO   Simplification           (parallel_processing.hpp   : 165)   Running Bulge remover\n",
      "  0:01:12.124   284M / 13G   INFO   Simplification           (parallel_processing.hpp   : 167)   Bulge remover triggered 102 times\n",
      "  0:01:12.124   284M / 13G   INFO   Simplification           (parallel_processing.hpp   : 165)   Running Low coverage edge remover\n",
      "  0:01:12.129   284M / 13G   INFO   Simplification           (parallel_processing.hpp   : 167)   Low coverage edge remover triggered 129 times\n",
      "  0:01:12.129   284M / 13G   INFO    General                 (simplification.cpp        : 362)   PROCEDURE == Simplification cycle, iteration 9\n",
      "  0:01:12.129   284M / 13G   INFO   Simplification           (parallel_processing.hpp   : 165)   Running Tip clipper\n",
      "  0:01:12.129   284M / 13G   INFO   Simplification           (parallel_processing.hpp   : 167)   Tip clipper triggered 0 times\n",
      "  0:01:12.129   284M / 13G   INFO   Simplification           (parallel_processing.hpp   : 165)   Running Bulge remover\n",
      "  0:01:12.146   284M / 13G   INFO   Simplification           (parallel_processing.hpp   : 167)   Bulge remover triggered 87 times\n",
      "  0:01:12.146   284M / 13G   INFO   Simplification           (parallel_processing.hpp   : 165)   Running Low coverage edge remover\n",
      "  0:01:12.149   284M / 13G   INFO   Simplification           (parallel_processing.hpp   : 167)   Low coverage edge remover triggered 106 times\n",
      "  0:01:12.149   284M / 13G   INFO    General                 (simplification.cpp        : 362)   PROCEDURE == Simplification cycle, iteration 10\n",
      "  0:01:12.149   284M / 13G   INFO   Simplification           (parallel_processing.hpp   : 165)   Running Tip clipper\n",
      "  0:01:12.150   284M / 13G   INFO   Simplification           (parallel_processing.hpp   : 167)   Tip clipper triggered 0 times\n",
      "  0:01:12.150   284M / 13G   INFO   Simplification           (parallel_processing.hpp   : 165)   Running Bulge remover\n",
      "  0:01:12.167   284M / 13G   INFO   Simplification           (parallel_processing.hpp   : 167)   Bulge remover triggered 78 times\n",
      "  0:01:12.167   284M / 13G   INFO   Simplification           (parallel_processing.hpp   : 165)   Running Low coverage edge remover\n",
      "  0:01:12.171   284M / 13G   INFO   Simplification           (parallel_processing.hpp   : 167)   Low coverage edge remover triggered 78 times\n",
      "  0:01:12.171   284M / 13G   INFO    General                 (simplification.cpp        : 362)   PROCEDURE == Simplification cycle, iteration 11\n",
      "  0:01:12.171   284M / 13G   INFO   Simplification           (parallel_processing.hpp   : 165)   Running Tip clipper\n",
      "  0:01:12.191   284M / 13G   INFO   Simplification           (parallel_processing.hpp   : 167)   Tip clipper triggered 1 times\n",
      "  0:01:12.191   284M / 13G   INFO   Simplification           (parallel_processing.hpp   : 165)   Running Bulge remover\n",
      "  0:01:12.234   284M / 13G   INFO   Simplification           (parallel_processing.hpp   : 167)   Bulge remover triggered 59 times\n",
      "  0:01:12.234   284M / 13G   INFO   Simplification           (parallel_processing.hpp   : 165)   Running Low coverage edge remover\n",
      "  0:01:12.252   284M / 13G   INFO   Simplification           (parallel_processing.hpp   : 167)   Low coverage edge remover triggered 0 times\n",
      "  0:01:12.252   284M / 13G   INFO    General                 (simplification.cpp        : 362)   PROCEDURE == Simplification cycle, iteration 12\n",
      "  0:01:12.252   284M / 13G   INFO   Simplification           (parallel_processing.hpp   : 165)   Running Tip clipper\n",
      "  0:01:12.252   284M / 13G   INFO   Simplification           (parallel_processing.hpp   : 167)   Tip clipper triggered 0 times\n",
      "  0:01:12.252   284M / 13G   INFO   Simplification           (parallel_processing.hpp   : 165)   Running Bulge remover\n",
      "  0:01:12.252   284M / 13G   INFO   Simplification           (parallel_processing.hpp   : 167)   Bulge remover triggered 0 times\n",
      "  0:01:12.252   284M / 13G   INFO   Simplification           (parallel_processing.hpp   : 165)   Running Low coverage edge remover\n",
      "  0:01:12.252   284M / 13G   INFO   Simplification           (parallel_processing.hpp   : 167)   Low coverage edge remover triggered 0 times\n",
      "  0:01:12.252   284M / 13G   INFO   StageManager             (stage.cpp                 : 132)   STAGE == Gap Closer\n",
      "  0:01:12.252   284M / 13G   INFO    General                 (graph_pack.hpp            : 101)   Index refill\n",
      "  0:01:12.252   284M / 13G   INFO   K-mer Index Building     (kmer_index_builder.hpp    : 301)   Building kmer index\n",
      "  0:01:12.252   284M / 13G   INFO    General                 (kmer_index_builder.hpp    : 117)   Splitting kmer instances into 384 files using 24 threads. This might take a while.\n",
      "  0:01:12.252   284M / 13G   INFO    General                 (file_limit.hpp            :  32)   Open file limit set to 1024\n",
      "  0:01:12.252   284M / 13G   INFO    General                 (kmer_splitters.hpp        :  89)   Memory available for splitting buffers: 0.857259 Gb\n",
      "  0:01:12.252   284M / 13G   INFO    General                 (kmer_splitters.hpp        :  97)   Using cell size of 87381\n",
      "  0:01:12.566    13G / 13G   INFO    General                 (edge_index_builders.hpp   :  77)   Processed 80359 edges\n",
      "  0:01:12.566    13G / 13G   INFO    General                 (edge_index_builders.hpp   :  82)   Used 80359 sequences.\n",
      "  0:01:12.677   284M / 13G   INFO    General                 (kmer_index_builder.hpp    : 120)   Starting k-mer counting.\n",
      "  0:01:12.800   284M / 13G   INFO    General                 (kmer_index_builder.hpp    : 127)   K-mer counting done. There are 7119572 kmers in total.\n",
      "  0:01:12.800   284M / 13G   INFO    General                 (kmer_index_builder.hpp    : 133)   Merging temporary buckets.\n",
      "  0:01:12.966   284M / 13G   INFO   K-mer Index Building     (kmer_index_builder.hpp    : 314)   Building perfect hash indices\n",
      "  0:01:13.076   284M / 13G   INFO    General                 (kmer_index_builder.hpp    : 150)   Merging final buckets.\n",
      "  0:01:13.244   284M / 13G   INFO   K-mer Index Building     (kmer_index_builder.hpp    : 336)   Index built. Total 3309656 bytes occupied (3.71894 bits per kmer).\n",
      "  0:01:13.371   448M / 13G   INFO    General                 (edge_index_builders.hpp   : 107)   Collecting edge information from graph, this takes a while.\n",
      "  0:01:13.523   448M / 13G   INFO    General                 (edge_index.hpp            :  92)   Index refilled\n",
      "  0:01:13.524   448M / 13G   INFO    General                 (gap_closer.cpp            : 159)   Preparing shift maps\n",
      "  0:01:13.610   456M / 13G   INFO    General                 (gap_closer.cpp            : 119)   Processing paired reads (takes a while)\n",
      "  0:01:17.735   460M / 13G   INFO    General                 (gap_closer.cpp            : 138)   Used 9842709 paired reads\n",
      "  0:01:17.736   460M / 13G   INFO    General                 (gap_closer.cpp            : 140)   Merging paired indices\n",
      "  0:01:17.793   448M / 13G   INFO   GapCloser                (gap_closer.cpp            : 346)   Closing short gaps\n",
      "  0:01:17.865   448M / 13G   INFO   GapCloser                (gap_closer.cpp            : 380)   Closing short gaps complete: filled 4 gaps after checking 495 candidates\n",
      "  0:01:17.882   448M / 13G   INFO   StageManager             (stage.cpp                 : 132)   STAGE == Simplification Cleanup\n",
      "  0:01:17.882   448M / 13G   INFO    General                 (simplification.cpp        : 196)   PROCEDURE == Post simplification\n",
      "  0:01:17.882   448M / 13G   INFO    General                 (graph_simplification.hpp  : 453)   Disconnection of relatively low covered edges disabled\n",
      "  0:01:17.882   448M / 13G   INFO    General                 (graph_simplification.hpp  : 489)   Complex tip clipping disabled\n",
      "  0:01:17.882   448M / 13G   INFO    General                 (graph_simplification.hpp  : 634)   Creating parallel br instance\n",
      "  0:01:17.882   448M / 13G   INFO   Simplification           (parallel_processing.hpp   : 165)   Running Tip clipper\n",
      "  0:01:17.889   448M / 13G   INFO   Simplification           (parallel_processing.hpp   : 167)   Tip clipper triggered 0 times\n",
      "  0:01:17.889   448M / 13G   INFO   Simplification           (parallel_processing.hpp   : 165)   Running Bulge remover\n",
      "  0:01:17.898   448M / 13G   INFO   Simplification           (parallel_processing.hpp   : 167)   Bulge remover triggered 0 times\n",
      "  0:01:17.898   448M / 13G   INFO   Simplification           (parallel_processing.hpp   : 165)   Running Tip clipper\n",
      "  0:01:17.905   448M / 13G   INFO   Simplification           (parallel_processing.hpp   : 167)   Tip clipper triggered 0 times\n",
      "  0:01:17.905   448M / 13G   INFO   Simplification           (parallel_processing.hpp   : 165)   Running Bulge remover\n",
      "  0:01:17.913   448M / 13G   INFO   Simplification           (parallel_processing.hpp   : 167)   Bulge remover triggered 0 times\n",
      "  0:01:17.913   448M / 13G   INFO    General                 (simplification.cpp        : 330)   Disrupting self-conjugate edges\n",
      "  0:01:17.948   448M / 13G   INFO   Simplification           (parallel_processing.hpp   : 165)   Running Removing isolated edges\n",
      "  0:01:19.509   448M / 13G   INFO   Simplification           (parallel_processing.hpp   : 167)   Removing isolated edges triggered 31575 times\n",
      "  0:01:19.509   448M / 13G   INFO    General                 (simplification.cpp        : 470)   Counting average coverage\n",
      "  0:01:19.517   448M / 13G   INFO    General                 (simplification.cpp        : 476)   Average coverage = 404.928\n",
      "  0:01:19.517   448M / 13G   INFO   StageManager             (stage.cpp                 : 132)   STAGE == Contig Output\n",
      "  0:01:19.517   448M / 13G   INFO    General                 (contig_output_stage.cpp   :  40)   Writing GFA to /mnt/d/linux/W/MoonNGS/raw/20180813Novogene16species_Human_Methyl/1.rawdata/MNH08365/Tychus/spades//K55/assembly_graph_with_scaffolds.gfa\n",
      "  0:01:19.577   448M / 13G   INFO    General                 (contig_output.hpp         :  22)   Outputting contigs to /mnt/d/linux/W/MoonNGS/raw/20180813Novogene16species_Human_Methyl/1.rawdata/MNH08365/Tychus/spades//K55/before_rr.fasta\n",
      "  0:01:19.654   448M / 13G   INFO    General                 (contig_output_stage.cpp   :  51)   Outputting FastG graph to /mnt/d/linux/W/MoonNGS/raw/20180813Novogene16species_Human_Methyl/1.rawdata/MNH08365/Tychus/spades//K55/assembly_graph.fastg\n",
      "  0:01:19.877   448M / 13G   INFO    General                 (contig_output.hpp         :  22)   Outputting contigs to /mnt/d/linux/W/MoonNGS/raw/20180813Novogene16species_Human_Methyl/1.rawdata/MNH08365/Tychus/spades//K55/simplified_contigs.fasta\n",
      "  0:01:19.954   448M / 13G   INFO    General                 (contig_output.hpp         :  22)   Outputting contigs to /mnt/d/linux/W/MoonNGS/raw/20180813Novogene16species_Human_Methyl/1.rawdata/MNH08365/Tychus/spades//K55/final_contigs.fasta\n",
      "  0:01:20.032   448M / 13G   INFO   StageManager             (stage.cpp                 : 132)   STAGE == Contig Output\n",
      "  0:01:20.032   448M / 13G   INFO    General                 (contig_output_stage.cpp   :  40)   Writing GFA to /mnt/d/linux/W/MoonNGS/raw/20180813Novogene16species_Human_Methyl/1.rawdata/MNH08365/Tychus/spades//K55/assembly_graph_with_scaffolds.gfa\n",
      "  0:01:20.089   448M / 13G   INFO    General                 (contig_output.hpp         :  22)   Outputting contigs to /mnt/d/linux/W/MoonNGS/raw/20180813Novogene16species_Human_Methyl/1.rawdata/MNH08365/Tychus/spades//K55/before_rr.fasta\n",
      "  0:01:20.167   448M / 13G   INFO    General                 (contig_output_stage.cpp   :  51)   Outputting FastG graph to /mnt/d/linux/W/MoonNGS/raw/20180813Novogene16species_Human_Methyl/1.rawdata/MNH08365/Tychus/spades//K55/assembly_graph.fastg\n",
      "  0:01:20.390   448M / 13G   INFO    General                 (contig_output.hpp         :  22)   Outputting contigs to /mnt/d/linux/W/MoonNGS/raw/20180813Novogene16species_Human_Methyl/1.rawdata/MNH08365/Tychus/spades//K55/simplified_contigs.fasta\n",
      "  0:01:20.469   448M / 13G   INFO    General                 (contig_output.hpp         :  22)   Outputting contigs to /mnt/d/linux/W/MoonNGS/raw/20180813Novogene16species_Human_Methyl/1.rawdata/MNH08365/Tychus/spades//K55/final_contigs.fasta\n",
      "  0:01:20.549   448M / 13G   INFO    General                 (launch.hpp                : 149)   SPAdes finished\n",
      "  0:01:21.002    96M / 13G   INFO    General                 (main.cpp                  : 109)   Assembling time: 0 hours 1 minutes 21 seconds\n",
      "\n",
      "== Running assembler: K77\n",
      "\n",
      "  0:00:00.000     4M / 4M    INFO    General                 (main.cpp                  :  74)   Loaded config from /mnt/d/linux/W/MoonNGS/raw/20180813Novogene16species_Human_Methyl/1.rawdata/MNH08365/Tychus/spades/K77/configs/config.info\n",
      "  0:00:00.000     4M / 4M    INFO    General                 (memory_limit.cpp          :  49)   Memory limit set to 62 Gb\n",
      "  0:00:00.000     4M / 4M    INFO    General                 (main.cpp                  :  87)   Starting SPAdes, built from N/A, git revision N/A\n",
      "  0:00:00.000     4M / 4M    INFO    General                 (main.cpp                  :  88)   Maximum k-mer length: 128\n",
      "  0:00:00.000     4M / 4M    INFO    General                 (main.cpp                  :  89)   Assembling dataset (/mnt/d/linux/W/MoonNGS/raw/20180813Novogene16species_Human_Methyl/1.rawdata/MNH08365/Tychus/spades/dataset.info) with K=77\n",
      "  0:00:00.000     4M / 4M    INFO    General                 (main.cpp                  :  90)   Maximum # of threads to use (adjusted due to OMP capabilities): 24\n",
      "  0:00:00.000     4M / 4M    INFO    General                 (launch.hpp                :  51)   SPAdes started\n",
      "  0:00:00.000     4M / 4M    INFO    General                 (launch.hpp                :  58)   Starting from stage: construction\n",
      "  0:00:00.000     4M / 4M    INFO    General                 (launch.hpp                :  65)   Two-step RR enabled: 0\n",
      "  0:00:00.000     4M / 4M    INFO    General                 (launch.hpp                :  76)   Will need read mapping, kmer mapper will be attached\n",
      "  0:00:00.000     4M / 4M    INFO   StageManager             (stage.cpp                 : 132)   STAGE == de Bruijn graph construction\n",
      "  0:00:00.001     4M / 4M    INFO    General                 (read_converter.hpp        :  59)   Binary reads detected\n",
      "  0:00:00.003     4M / 4M    INFO    General                 (construction.cpp          : 111)   Max read length 150\n",
      "  0:00:00.003     4M / 4M    INFO    General                 (construction.cpp          : 117)   Average read length 146.615\n",
      "  0:00:00.003     4M / 4M    INFO    General                 (stage.cpp                 : 101)   PROCEDURE == k+1-mer counting\n",
      "  0:00:00.003     4M / 4M    INFO    General                 (kmer_index_builder.hpp    : 117)   Splitting kmer instances into 576 files using 24 threads. This might take a while.\n",
      "  0:00:00.003     4M / 4M    INFO    General                 (file_limit.hpp            :  32)   Open file limit set to 1024\n",
      "  0:00:00.004     4M / 4M    INFO    General                 (kmer_splitters.hpp        :  89)   Memory available for splitting buffers: 0.861057 Gb\n",
      "  0:00:00.004     4M / 4M    INFO    General                 (kmer_splitters.hpp        :  97)   Using cell size of 38836\n",
      "  0:00:09.964    13G / 13G   INFO    General                 (kmer_splitters.hpp        : 289)   Processed 14664038 reads\n",
      "  0:00:18.484    13G / 13G   INFO    General                 (kmer_splitters.hpp        : 289)   Processed 29286456 reads\n",
      "  0:00:25.545    13G / 13G   INFO    General                 (kmer_splitters.hpp        : 289)   Processed 39380436 reads\n",
      "  0:00:25.545    13G / 13G   INFO    General                 (kmer_splitters.hpp        : 295)   Adding contigs from previous K\n",
      "  0:00:28.018    96M / 13G   INFO    General                 (kmer_splitters.hpp        : 308)   Used 39380436 reads\n",
      "  0:00:28.018    96M / 13G   INFO    General                 (kmer_index_builder.hpp    : 120)   Starting k-mer counting.\n",
      "  0:00:28.660    96M / 13G   INFO    General                 (kmer_index_builder.hpp    : 127)   K-mer counting done. There are 9339318 kmers in total.\n",
      "  0:00:28.660    96M / 13G   INFO    General                 (kmer_index_builder.hpp    : 133)   Merging temporary buckets.\n",
      "  0:00:29.003    96M / 13G   INFO    General                 (stage.cpp                 : 101)   PROCEDURE == Extension index construction\n",
      "  0:00:29.003    96M / 13G   INFO   K-mer Index Building     (kmer_index_builder.hpp    : 301)   Building kmer index\n",
      "  0:00:29.003    96M / 13G   INFO    General                 (kmer_index_builder.hpp    : 117)   Splitting kmer instances into 384 files using 24 threads. This might take a while.\n",
      "  0:00:29.003    96M / 13G   INFO    General                 (file_limit.hpp            :  32)   Open file limit set to 1024\n",
      "  0:00:29.003    96M / 13G   INFO    General                 (kmer_splitters.hpp        :  89)   Memory available for splitting buffers: 0.859809 Gb\n",
      "  0:00:29.003    96M / 13G   INFO    General                 (kmer_splitters.hpp        :  97)   Using cell size of 58254\n",
      "  0:00:29.821    13G / 13G   INFO    General                 (kmer_splitters.hpp        : 380)   Processed 9339318 kmers\n",
      "  0:00:29.821    13G / 13G   INFO    General                 (kmer_splitters.hpp        : 385)   Used 9339318 kmers.\n",
      "  0:00:29.999    96M / 13G   INFO    General                 (kmer_index_builder.hpp    : 120)   Starting k-mer counting.\n",
      "  0:00:30.245    96M / 13G   INFO    General                 (kmer_index_builder.hpp    : 127)   K-mer counting done. There are 9368229 kmers in total.\n",
      "  0:00:30.245    96M / 13G   INFO    General                 (kmer_index_builder.hpp    : 133)   Merging temporary buckets.\n",
      "  0:00:30.498    96M / 13G   INFO   K-mer Index Building     (kmer_index_builder.hpp    : 314)   Building perfect hash indices\n",
      "  0:00:30.881    96M / 13G   INFO    General                 (kmer_index_builder.hpp    : 150)   Merging final buckets.\n",
      "  0:00:31.277    96M / 13G   INFO   K-mer Index Building     (kmer_index_builder.hpp    : 336)   Index built. Total 4352200 bytes occupied (3.71656 bits per kmer).\n",
      "  0:00:31.287   108M / 13G   INFO   DeBruijnExtensionIndexBu (kmer_extension_index_build:  99)   Building k-mer extensions from k+1-mers\n",
      "  0:00:31.591   108M / 13G   INFO   DeBruijnExtensionIndexBu (kmer_extension_index_build: 103)   Building k-mer extensions from k+1-mers finished.\n",
      "  0:00:31.592   108M / 13G   INFO    General                 (stage.cpp                 : 101)   PROCEDURE == Condensing graph\n",
      "  0:00:31.612   108M / 13G   INFO   UnbranchingPathExtractor (debruijn_graph_constructor: 355)   Extracting unbranching paths\n",
      "  0:00:31.954   120M / 13G   INFO   UnbranchingPathExtractor (debruijn_graph_constructor: 374)   Extracting unbranching paths finished. 385718 sequences extracted\n",
      "  0:00:32.206   120M / 13G   INFO   UnbranchingPathExtractor (debruijn_graph_constructor: 310)   Collecting perfect loops\n",
      "  0:00:32.289   120M / 13G   INFO   UnbranchingPathExtractor (debruijn_graph_constructor: 343)   Collecting perfect loops finished. 20 loops collected\n",
      "  0:00:32.481   220M / 13G   INFO    General                 (stage.cpp                 : 101)   PROCEDURE == Filling coverage indices (PHM)\n",
      "  0:00:32.481   220M / 13G   INFO   K-mer Index Building     (kmer_index_builder.hpp    : 301)   Building kmer index\n",
      "  0:00:32.481   220M / 13G   INFO   K-mer Index Building     (kmer_index_builder.hpp    : 314)   Building perfect hash indices\n",
      "  0:00:32.595   220M / 13G   INFO   K-mer Index Building     (kmer_index_builder.hpp    : 336)   Index built. Total 4342856 bytes occupied (3.72006 bits per kmer).\n",
      "  0:00:32.619   256M / 13G   INFO    General                 (construction.cpp          : 388)   Collecting k-mer coverage information from reads, this takes a while.\n",
      "  0:00:49.558   256M / 13G   INFO    General                 (construction.cpp          : 508)   Filling coverage and flanking coverage from PHM\n",
      "  0:00:49.914   256M / 13G   INFO    General                 (construction.cpp          : 464)   Processed 771470 edges\n",
      "  0:00:49.986   208M / 13G   INFO   StageManager             (stage.cpp                 : 132)   STAGE == EC Threshold Finding\n",
      "  0:00:49.987   208M / 13G   INFO    General                 (kmer_coverage_model.cpp   : 181)   Kmer coverage valley at: 41\n",
      "  0:00:49.987   208M / 13G   INFO    General                 (kmer_coverage_model.cpp   : 201)   K-mer histogram maximum: 419\n",
      "  0:00:49.987   208M / 13G   INFO    General                 (kmer_coverage_model.cpp   : 237)   Estimated median coverage: 424. Coverage mad: 59.304\n",
      "  0:00:49.987   208M / 13G   INFO    General                 (kmer_coverage_model.cpp   : 259)   Fitting coverage model\n",
      "  0:00:50.408   208M / 13G   INFO    General                 (kmer_coverage_model.cpp   : 295)   ... iteration 2\n",
      "  0:00:51.490   208M / 13G   INFO    General                 (kmer_coverage_model.cpp   : 295)   ... iteration 4\n",
      "  0:00:55.188   208M / 13G   INFO    General                 (kmer_coverage_model.cpp   : 295)   ... iteration 8\n",
      "  0:01:03.042   208M / 13G   INFO    General                 (kmer_coverage_model.cpp   : 295)   ... iteration 16\n",
      "  0:01:04.020   208M / 13G   INFO    General                 (kmer_coverage_model.cpp   : 309)   Fitted mean coverage: 425.609. Fitted coverage std. dev: 57.3924\n",
      "  0:01:04.035   208M / 13G   INFO    General                 (kmer_coverage_model.cpp   : 334)   Probability of erroneous kmer at valley: 1\n",
      "  0:01:04.035   208M / 13G   INFO    General                 (kmer_coverage_model.cpp   : 358)   Preliminary threshold calculated as: 282\n",
      "  0:01:04.035   208M / 13G   INFO    General                 (kmer_coverage_model.cpp   : 362)   Threshold adjusted to: 233\n",
      "  0:01:04.035   208M / 13G   INFO    General                 (kmer_coverage_model.cpp   : 375)   Estimated genome size (ignoring repeats): 3025866\n",
      "  0:01:04.035   208M / 13G   INFO    General                 (genomic_info_filler.cpp   : 112)   Mean coverage was calculated as 425.609\n",
      "  0:01:04.035   208M / 13G   INFO    General                 (genomic_info_filler.cpp   : 127)   EC coverage threshold value was calculated as 233\n",
      "  0:01:04.035   208M / 13G   INFO    General                 (genomic_info_filler.cpp   : 128)   Trusted kmer low bound: 0\n",
      "  0:01:04.035   208M / 13G   INFO   StageManager             (stage.cpp                 : 132)   STAGE == Gap Closer\n",
      "  0:01:04.035   208M / 13G   INFO    General                 (graph_pack.hpp            : 101)   Index refill\n",
      "  0:01:04.035   208M / 13G   INFO   K-mer Index Building     (kmer_index_builder.hpp    : 301)   Building kmer index\n",
      "  0:01:04.035   208M / 13G   INFO    General                 (kmer_index_builder.hpp    : 117)   Splitting kmer instances into 384 files using 24 threads. This might take a while.\n",
      "  0:01:04.035   208M / 13G   INFO    General                 (file_limit.hpp            :  32)   Open file limit set to 1024\n",
      "  0:01:04.035   208M / 13G   INFO    General                 (kmer_splitters.hpp        :  89)   Memory available for splitting buffers: 0.85829 Gb\n",
      "  0:01:04.035   208M / 13G   INFO    General                 (kmer_splitters.hpp        :  97)   Using cell size of 58254\n",
      "  0:01:04.648    13G / 13G   INFO    General                 (edge_index_builders.hpp   :  77)   Processed 771470 edges\n",
      "  0:01:04.648    13G / 13G   INFO    General                 (edge_index_builders.hpp   :  82)   Used 771470 sequences.\n",
      "  0:01:04.794   204M / 13G   INFO    General                 (kmer_index_builder.hpp    : 120)   Starting k-mer counting.\n",
      "  0:01:05.012   204M / 13G   INFO    General                 (kmer_index_builder.hpp    : 127)   K-mer counting done. There are 9339318 kmers in total.\n",
      "  0:01:05.012   204M / 13G   INFO    General                 (kmer_index_builder.hpp    : 133)   Merging temporary buckets.\n",
      "  0:01:05.290   204M / 13G   INFO   K-mer Index Building     (kmer_index_builder.hpp    : 314)   Building perfect hash indices\n",
      "  0:01:05.627   208M / 13G   INFO    General                 (kmer_index_builder.hpp    : 150)   Merging final buckets.\n",
      "  0:01:05.925   208M / 13G   INFO   K-mer Index Building     (kmer_index_builder.hpp    : 336)   Index built. Total 4338832 bytes occupied (3.71662 bits per kmer).\n",
      "  0:01:06.052   424M / 13G   INFO    General                 (edge_index_builders.hpp   : 107)   Collecting edge information from graph, this takes a while.\n",
      "  0:01:06.257   424M / 13G   INFO    General                 (edge_index.hpp            :  92)   Index refilled\n",
      "  0:01:06.257   424M / 13G   INFO    General                 (gap_closer.cpp            : 159)   Preparing shift maps\n",
      "  0:01:06.722   456M / 13G   INFO    General                 (gap_closer.cpp            : 119)   Processing paired reads (takes a while)\n",
      "  0:01:12.187   456M / 13G   INFO    General                 (gap_closer.cpp            : 138)   Used 9842709 paired reads\n",
      "  0:01:12.187   456M / 13G   INFO    General                 (gap_closer.cpp            : 140)   Merging paired indices\n",
      "  0:01:12.311   428M / 13G   INFO   GapCloser                (gap_closer.cpp            : 346)   Closing short gaps\n",
      "  0:01:12.945   428M / 13G   INFO   GapCloser                (gap_closer.cpp            : 380)   Closing short gaps complete: filled 70 gaps after checking 1976 candidates\n",
      "  0:01:12.995   424M / 13G   INFO   StageManager             (stage.cpp                 : 132)   STAGE == Raw Simplification\n",
      "  0:01:13.021   208M / 13G   INFO    General                 (simplification.cpp        : 128)   PROCEDURE == InitialCleaning\n",
      "  0:01:13.021   208M / 13G   INFO    General                 (graph_simplification.hpp  : 662)   Flanking coverage based disconnection disabled\n",
      "  0:01:13.021   208M / 13G   INFO   Simplification           (parallel_processing.hpp   : 165)   Running Self conjugate edge remover\n",
      "  0:01:13.045   208M / 13G   INFO   Simplification           (parallel_processing.hpp   : 167)   Self conjugate edge remover triggered 0 times\n",
      "  0:01:13.046   208M / 13G   INFO   Simplification           (parallel_processing.hpp   : 165)   Running Initial tip clipper\n",
      "  0:01:14.960   216M / 13G   INFO   Simplification           (parallel_processing.hpp   : 167)   Initial tip clipper triggered 158097 times\n",
      "  0:01:14.960   216M / 13G   INFO   Simplification           (parallel_processing.hpp   : 165)   Running Initial ec remover\n",
      "  0:01:15.084   212M / 13G   INFO   Simplification           (parallel_processing.hpp   : 167)   Initial ec remover triggered 3603 times\n",
      "  0:01:15.084   212M / 13G   INFO   Simplification           (parallel_processing.hpp   : 165)   Running Initial isolated edge remover\n",
      "  0:01:15.227   212M / 13G   INFO   Simplification           (parallel_processing.hpp   : 167)   Initial isolated edge remover triggered 34869 times\n",
      "  0:01:15.231   208M / 13G   INFO   StageManager             (stage.cpp                 : 132)   STAGE == Simplification\n",
      "  0:01:15.231   208M / 13G   INFO    General                 (simplification.cpp        : 357)   Graph simplification started\n",
      "  0:01:15.231   208M / 13G   INFO    General                 (graph_simplification.hpp  : 634)   Creating parallel br instance\n",
      "  0:01:15.231   208M / 13G   INFO    General                 (simplification.cpp        : 362)   PROCEDURE == Simplification cycle, iteration 1\n",
      "  0:01:15.231   208M / 13G   INFO   Simplification           (parallel_processing.hpp   : 165)   Running Tip clipper\n",
      "  0:01:15.238   208M / 13G   INFO   Simplification           (parallel_processing.hpp   : 167)   Tip clipper triggered 250 times\n",
      "  0:01:15.238   208M / 13G   INFO   Simplification           (parallel_processing.hpp   : 165)   Running Bulge remover\n",
      "  0:01:16.725   264M / 13G   INFO   Simplification           (parallel_processing.hpp   : 167)   Bulge remover triggered 6023 times\n",
      "  0:01:16.725   264M / 13G   INFO   Simplification           (parallel_processing.hpp   : 165)   Running Low coverage edge remover\n",
      "  0:01:16.810   212M / 13G   INFO   Simplification           (parallel_processing.hpp   : 167)   Low coverage edge remover triggered 1739 times\n",
      "  0:01:16.810   212M / 13G   INFO    General                 (simplification.cpp        : 362)   PROCEDURE == Simplification cycle, iteration 2\n",
      "  0:01:16.810   212M / 13G   INFO   Simplification           (parallel_processing.hpp   : 165)   Running Tip clipper\n",
      "  0:01:16.813   212M / 13G   INFO   Simplification           (parallel_processing.hpp   : 167)   Tip clipper triggered 23 times\n",
      "  0:01:16.813   212M / 13G   INFO   Simplification           (parallel_processing.hpp   : 165)   Running Bulge remover\n",
      "  0:01:16.859   208M / 13G   INFO   Simplification           (parallel_processing.hpp   : 167)   Bulge remover triggered 116 times\n",
      "  0:01:16.859   208M / 13G   INFO   Simplification           (parallel_processing.hpp   : 165)   Running Low coverage edge remover\n",
      "  0:01:16.859   208M / 13G   INFO   Simplification           (parallel_processing.hpp   : 167)   Low coverage edge remover triggered 1 times\n",
      "  0:01:16.859   208M / 13G   INFO    General                 (simplification.cpp        : 362)   PROCEDURE == Simplification cycle, iteration 3\n",
      "  0:01:16.859   208M / 13G   INFO   Simplification           (parallel_processing.hpp   : 165)   Running Tip clipper\n",
      "  0:01:16.859   208M / 13G   INFO   Simplification           (parallel_processing.hpp   : 167)   Tip clipper triggered 0 times\n",
      "  0:01:16.859   208M / 13G   INFO   Simplification           (parallel_processing.hpp   : 165)   Running Bulge remover\n",
      "  0:01:16.859   208M / 13G   INFO   Simplification           (parallel_processing.hpp   : 167)   Bulge remover triggered 0 times\n",
      "  0:01:16.859   208M / 13G   INFO   Simplification           (parallel_processing.hpp   : 165)   Running Low coverage edge remover\n",
      "  0:01:16.859   208M / 13G   INFO   Simplification           (parallel_processing.hpp   : 167)   Low coverage edge remover triggered 0 times\n",
      "  0:01:16.859   208M / 13G   INFO    General                 (simplification.cpp        : 362)   PROCEDURE == Simplification cycle, iteration 4\n",
      "  0:01:16.859   208M / 13G   INFO   Simplification           (parallel_processing.hpp   : 165)   Running Tip clipper\n",
      "  0:01:16.859   208M / 13G   INFO   Simplification           (parallel_processing.hpp   : 167)   Tip clipper triggered 0 times\n",
      "  0:01:16.859   208M / 13G   INFO   Simplification           (parallel_processing.hpp   : 165)   Running Bulge remover\n",
      "  0:01:16.859   208M / 13G   INFO   Simplification           (parallel_processing.hpp   : 167)   Bulge remover triggered 0 times\n",
      "  0:01:16.859   208M / 13G   INFO   Simplification           (parallel_processing.hpp   : 165)   Running Low coverage edge remover\n",
      "  0:01:16.859   208M / 13G   INFO   Simplification           (parallel_processing.hpp   : 167)   Low coverage edge remover triggered 0 times\n",
      "  0:01:16.859   208M / 13G   INFO    General                 (simplification.cpp        : 362)   PROCEDURE == Simplification cycle, iteration 5\n",
      "  0:01:16.859   208M / 13G   INFO   Simplification           (parallel_processing.hpp   : 165)   Running Tip clipper\n",
      "  0:01:16.859   208M / 13G   INFO   Simplification           (parallel_processing.hpp   : 167)   Tip clipper triggered 0 times\n",
      "  0:01:16.859   208M / 13G   INFO   Simplification           (parallel_processing.hpp   : 165)   Running Bulge remover\n",
      "  0:01:16.859   208M / 13G   INFO   Simplification           (parallel_processing.hpp   : 167)   Bulge remover triggered 0 times\n",
      "  0:01:16.859   208M / 13G   INFO   Simplification           (parallel_processing.hpp   : 165)   Running Low coverage edge remover\n",
      "  0:01:16.859   208M / 13G   INFO   Simplification           (parallel_processing.hpp   : 167)   Low coverage edge remover triggered 1 times\n",
      "  0:01:16.859   208M / 13G   INFO    General                 (simplification.cpp        : 362)   PROCEDURE == Simplification cycle, iteration 6\n",
      "  0:01:16.859   208M / 13G   INFO   Simplification           (parallel_processing.hpp   : 165)   Running Tip clipper\n",
      "  0:01:16.859   208M / 13G   INFO   Simplification           (parallel_processing.hpp   : 167)   Tip clipper triggered 0 times\n",
      "  0:01:16.859   208M / 13G   INFO   Simplification           (parallel_processing.hpp   : 165)   Running Bulge remover\n",
      "  0:01:16.859   208M / 13G   INFO   Simplification           (parallel_processing.hpp   : 167)   Bulge remover triggered 0 times\n",
      "  0:01:16.859   208M / 13G   INFO   Simplification           (parallel_processing.hpp   : 165)   Running Low coverage edge remover\n",
      "  0:01:16.860   208M / 13G   INFO   Simplification           (parallel_processing.hpp   : 167)   Low coverage edge remover triggered 0 times\n",
      "  0:01:16.860   208M / 13G   INFO    General                 (simplification.cpp        : 362)   PROCEDURE == Simplification cycle, iteration 7\n",
      "  0:01:16.860   208M / 13G   INFO   Simplification           (parallel_processing.hpp   : 165)   Running Tip clipper\n",
      "  0:01:16.860   208M / 13G   INFO   Simplification           (parallel_processing.hpp   : 167)   Tip clipper triggered 0 times\n",
      "  0:01:16.860   208M / 13G   INFO   Simplification           (parallel_processing.hpp   : 165)   Running Bulge remover\n",
      "  0:01:16.860   208M / 13G   INFO   Simplification           (parallel_processing.hpp   : 167)   Bulge remover triggered 0 times\n",
      "  0:01:16.860   208M / 13G   INFO   Simplification           (parallel_processing.hpp   : 165)   Running Low coverage edge remover\n",
      "  0:01:16.860   208M / 13G   INFO   Simplification           (parallel_processing.hpp   : 167)   Low coverage edge remover triggered 0 times\n",
      "  0:01:16.860   208M / 13G   INFO    General                 (simplification.cpp        : 362)   PROCEDURE == Simplification cycle, iteration 8\n",
      "  0:01:16.860   208M / 13G   INFO   Simplification           (parallel_processing.hpp   : 165)   Running Tip clipper\n",
      "  0:01:16.860   208M / 13G   INFO   Simplification           (parallel_processing.hpp   : 167)   Tip clipper triggered 0 times\n",
      "  0:01:16.860   208M / 13G   INFO   Simplification           (parallel_processing.hpp   : 165)   Running Bulge remover\n",
      "  0:01:16.860   208M / 13G   INFO   Simplification           (parallel_processing.hpp   : 167)   Bulge remover triggered 0 times\n",
      "  0:01:16.860   208M / 13G   INFO   Simplification           (parallel_processing.hpp   : 165)   Running Low coverage edge remover\n",
      "  0:01:16.860   208M / 13G   INFO   Simplification           (parallel_processing.hpp   : 167)   Low coverage edge remover triggered 0 times\n",
      "  0:01:16.860   208M / 13G   INFO    General                 (simplification.cpp        : 362)   PROCEDURE == Simplification cycle, iteration 9\n",
      "  0:01:16.860   208M / 13G   INFO   Simplification           (parallel_processing.hpp   : 165)   Running Tip clipper\n",
      "  0:01:16.860   208M / 13G   INFO   Simplification           (parallel_processing.hpp   : 167)   Tip clipper triggered 0 times\n",
      "  0:01:16.860   208M / 13G   INFO   Simplification           (parallel_processing.hpp   : 165)   Running Bulge remover\n",
      "  0:01:16.860   208M / 13G   INFO   Simplification           (parallel_processing.hpp   : 167)   Bulge remover triggered 0 times\n",
      "  0:01:16.860   208M / 13G   INFO   Simplification           (parallel_processing.hpp   : 165)   Running Low coverage edge remover\n",
      "  0:01:16.860   208M / 13G   INFO   Simplification           (parallel_processing.hpp   : 167)   Low coverage edge remover triggered 1 times\n",
      "  0:01:16.860   208M / 13G   INFO    General                 (simplification.cpp        : 362)   PROCEDURE == Simplification cycle, iteration 10\n",
      "  0:01:16.860   208M / 13G   INFO   Simplification           (parallel_processing.hpp   : 165)   Running Tip clipper\n",
      "  0:01:16.860   208M / 13G   INFO   Simplification           (parallel_processing.hpp   : 167)   Tip clipper triggered 0 times\n",
      "  0:01:16.860   208M / 13G   INFO   Simplification           (parallel_processing.hpp   : 165)   Running Bulge remover\n",
      "  0:01:16.860   208M / 13G   INFO   Simplification           (parallel_processing.hpp   : 167)   Bulge remover triggered 0 times\n",
      "  0:01:16.860   208M / 13G   INFO   Simplification           (parallel_processing.hpp   : 165)   Running Low coverage edge remover\n",
      "  0:01:16.860   208M / 13G   INFO   Simplification           (parallel_processing.hpp   : 167)   Low coverage edge remover triggered 0 times\n",
      "  0:01:16.860   208M / 13G   INFO    General                 (simplification.cpp        : 362)   PROCEDURE == Simplification cycle, iteration 11\n",
      "  0:01:16.860   208M / 13G   INFO   Simplification           (parallel_processing.hpp   : 165)   Running Tip clipper\n",
      "  0:01:16.861   208M / 13G   INFO   Simplification           (parallel_processing.hpp   : 167)   Tip clipper triggered 0 times\n",
      "  0:01:16.861   208M / 13G   INFO   Simplification           (parallel_processing.hpp   : 165)   Running Bulge remover\n",
      "  0:01:16.863   208M / 13G   INFO   Simplification           (parallel_processing.hpp   : 167)   Bulge remover triggered 0 times\n",
      "  0:01:16.863   208M / 13G   INFO   Simplification           (parallel_processing.hpp   : 165)   Running Low coverage edge remover\n",
      "  0:01:16.869   204M / 13G   INFO   Simplification           (parallel_processing.hpp   : 167)   Low coverage edge remover triggered 0 times\n",
      "  0:01:16.869   204M / 13G   INFO   StageManager             (stage.cpp                 : 132)   STAGE == Gap Closer\n",
      "  0:01:16.869   204M / 13G   INFO    General                 (graph_pack.hpp            : 101)   Index refill\n",
      "  0:01:16.870   204M / 13G   INFO   K-mer Index Building     (kmer_index_builder.hpp    : 301)   Building kmer index\n",
      "  0:01:16.870   204M / 13G   INFO    General                 (kmer_index_builder.hpp    : 117)   Splitting kmer instances into 384 files using 24 threads. This might take a while.\n",
      "  0:01:16.870   200M / 13G   INFO    General                 (file_limit.hpp            :  32)   Open file limit set to 1024\n",
      "  0:01:16.870   200M / 13G   INFO    General                 (kmer_splitters.hpp        :  89)   Memory available for splitting buffers: 0.858398 Gb\n",
      "  0:01:16.870   200M / 13G   INFO    General                 (kmer_splitters.hpp        :  97)   Using cell size of 58254\n",
      "  0:01:17.319    13G / 13G   INFO    General                 (edge_index_builders.hpp   :  77)   Processed 12022 edges\n",
      "  0:01:17.319    13G / 13G   INFO    General                 (edge_index_builders.hpp   :  82)   Used 12022 sequences.\n",
      "  0:01:17.430   180M / 13G   INFO    General                 (kmer_index_builder.hpp    : 120)   Starting k-mer counting.\n",
      "  0:01:17.578   180M / 13G   INFO    General                 (kmer_index_builder.hpp    : 127)   K-mer counting done. There are 4113559 kmers in total.\n",
      "  0:01:17.579   180M / 13G   INFO    General                 (kmer_index_builder.hpp    : 133)   Merging temporary buckets.\n",
      "  0:01:17.699   180M / 13G   INFO   K-mer Index Building     (kmer_index_builder.hpp    : 314)   Building perfect hash indices\n",
      "  0:01:17.878   184M / 13G   INFO    General                 (kmer_index_builder.hpp    : 150)   Merging final buckets.\n",
      "  0:01:18.039   184M / 13G   INFO   K-mer Index Building     (kmer_index_builder.hpp    : 336)   Index built. Total 1915600 bytes occupied (3.72544 bits per kmer).\n",
      "  0:01:18.167   280M / 13G   INFO    General                 (edge_index_builders.hpp   : 107)   Collecting edge information from graph, this takes a while.\n",
      "  0:01:18.261   276M / 13G   INFO    General                 (edge_index.hpp            :  92)   Index refilled\n",
      "  0:01:18.262   276M / 13G   INFO    General                 (gap_closer.cpp            : 159)   Preparing shift maps\n",
      "  0:01:18.274   280M / 13G   INFO    General                 (gap_closer.cpp            : 119)   Processing paired reads (takes a while)\n",
      "  0:01:21.951   284M / 13G   INFO    General                 (gap_closer.cpp            : 138)   Used 9842709 paired reads\n",
      "  0:01:21.951   284M / 13G   INFO    General                 (gap_closer.cpp            : 140)   Merging paired indices\n",
      "  0:01:21.978   276M / 13G   INFO   GapCloser                (gap_closer.cpp            : 346)   Closing short gaps\n",
      "  0:01:21.996   276M / 13G   INFO   GapCloser                (gap_closer.cpp            : 380)   Closing short gaps complete: filled 0 gaps after checking 10 candidates\n",
      "  0:01:22.003   276M / 13G   INFO   StageManager             (stage.cpp                 : 132)   STAGE == Simplification Cleanup\n",
      "  0:01:22.004   276M / 13G   INFO    General                 (simplification.cpp        : 196)   PROCEDURE == Post simplification\n",
      "  0:01:22.004   276M / 13G   INFO    General                 (graph_simplification.hpp  : 453)   Disconnection of relatively low covered edges disabled\n",
      "  0:01:22.004   276M / 13G   INFO    General                 (graph_simplification.hpp  : 489)   Complex tip clipping disabled\n",
      "  0:01:22.004   276M / 13G   INFO    General                 (graph_simplification.hpp  : 634)   Creating parallel br instance\n",
      "  0:01:22.004   276M / 13G   INFO   Simplification           (parallel_processing.hpp   : 165)   Running Tip clipper\n",
      "  0:01:22.005   276M / 13G   INFO   Simplification           (parallel_processing.hpp   : 167)   Tip clipper triggered 0 times\n",
      "  0:01:22.005   276M / 13G   INFO   Simplification           (parallel_processing.hpp   : 165)   Running Bulge remover\n",
      "  0:01:22.006   276M / 13G   INFO   Simplification           (parallel_processing.hpp   : 167)   Bulge remover triggered 0 times\n",
      "  0:01:22.006   276M / 13G   INFO   Simplification           (parallel_processing.hpp   : 165)   Running Tip clipper\n",
      "  0:01:22.007   276M / 13G   INFO   Simplification           (parallel_processing.hpp   : 167)   Tip clipper triggered 0 times\n",
      "  0:01:22.007   276M / 13G   INFO   Simplification           (parallel_processing.hpp   : 165)   Running Bulge remover\n",
      "  0:01:22.008   276M / 13G   INFO   Simplification           (parallel_processing.hpp   : 167)   Bulge remover triggered 0 times\n",
      "  0:01:22.008   276M / 13G   INFO    General                 (simplification.cpp        : 330)   Disrupting self-conjugate edges\n",
      "  0:01:22.012   276M / 13G   INFO   Simplification           (parallel_processing.hpp   : 165)   Running Removing isolated edges\n",
      "  0:01:22.013   276M / 13G   INFO   Simplification           (parallel_processing.hpp   : 167)   Removing isolated edges triggered 0 times\n",
      "  0:01:22.013   276M / 13G   INFO    General                 (simplification.cpp        : 470)   Counting average coverage\n",
      "  0:01:22.015   276M / 13G   INFO    General                 (simplification.cpp        : 476)   Average coverage = 332.88\n",
      "  0:01:22.015   276M / 13G   INFO   StageManager             (stage.cpp                 : 132)   STAGE == Mismatch Correction\n",
      "  0:01:22.015   276M / 13G   INFO    General                 (graph_pack.hpp            : 109)   Normalizing k-mer map. Total 963478 kmers to process\n",
      "  0:01:22.669   276M / 13G   INFO    General                 (graph_pack.hpp            : 111)   Normalizing done\n",
      "  0:01:37.368   336M / 13G   INFO    General                 (mismatch_shall_not_pass.hp: 189)   Finished collecting potential mismatches positions\n",
      "  0:01:37.582   280M / 13G   INFO    General                 (mismatch_shall_not_pass.hp: 290)   All edges processed\n",
      "  0:01:37.585   280M / 13G   INFO    General                 (mismatch_correction.cpp   :  27)   Corrected 0 nucleotides\n",
      "  0:01:37.585   276M / 13G   INFO   StageManager             (stage.cpp                 : 132)   STAGE == Contig Output\n",
      "  0:01:37.585   276M / 13G   INFO    General                 (contig_output_stage.cpp   :  40)   Writing GFA to /mnt/d/linux/W/MoonNGS/raw/20180813Novogene16species_Human_Methyl/1.rawdata/MNH08365/Tychus/spades//K77/assembly_graph_with_scaffolds.gfa\n",
      "  0:01:37.654   276M / 13G   INFO    General                 (contig_output.hpp         :  22)   Outputting contigs to /mnt/d/linux/W/MoonNGS/raw/20180813Novogene16species_Human_Methyl/1.rawdata/MNH08365/Tychus/spades//K77/before_rr.fasta\n",
      "  0:01:37.743   276M / 13G   INFO    General                 (contig_output_stage.cpp   :  51)   Outputting FastG graph to /mnt/d/linux/W/MoonNGS/raw/20180813Novogene16species_Human_Methyl/1.rawdata/MNH08365/Tychus/spades//K77/assembly_graph.fastg\n",
      "  0:01:37.997   276M / 13G   INFO    General                 (contig_output.hpp         :  22)   Outputting contigs to /mnt/d/linux/W/MoonNGS/raw/20180813Novogene16species_Human_Methyl/1.rawdata/MNH08365/Tychus/spades//K77/simplified_contigs.fasta\n",
      "  0:01:38.064   276M / 13G   INFO    General                 (contig_output.hpp         :  22)   Outputting contigs to /mnt/d/linux/W/MoonNGS/raw/20180813Novogene16species_Human_Methyl/1.rawdata/MNH08365/Tychus/spades//K77/intermediate_contigs.fasta\n",
      "  0:01:38.130   276M / 13G   INFO   StageManager             (stage.cpp                 : 132)   STAGE == Paired Information Counting\n",
      "  0:01:38.147   276M / 13G   INFO    General                 (graph_pack.hpp            : 109)   Normalizing k-mer map. Total 963478 kmers to process\n",
      "  0:01:38.147   276M / 13G   INFO    General                 (graph_pack.hpp            : 111)   Normalizing done\n",
      "  0:01:38.149   276M / 13G   INFO    General                 (pair_info_count.cpp       : 320)   Min edge length for estimation: 41844\n",
      "  0:01:38.149   276M / 13G   INFO    General                 (pair_info_count.cpp       : 331)   Estimating insert size for library #0\n",
      "  0:01:38.149   276M / 13G   INFO    General                 (pair_info_count.cpp       : 190)   Estimating insert size (takes a while)\n",
      "  0:01:38.432   676M / 13G   INFO    General                 (pair_info_count.cpp       :  39)   Selecting usual mapper\n",
      "  0:01:41.704   676M / 13G   INFO    General                 (sequence_mapper_notifier.h:  80)   Processed 200000 reads\n",
      "  0:01:41.739   680M / 13G   INFO    General                 (sequence_mapper_notifier.h:  80)   Processed 400000 reads\n",
      "  0:01:41.789   676M / 13G   INFO    General                 (sequence_mapper_notifier.h:  80)   Processed 600000 reads\n",
      "  0:01:41.816   676M / 13G   INFO    General                 (sequence_mapper_notifier.h:  80)   Processed 800000 reads\n",
      "  0:01:41.863   676M / 13G   INFO    General                 (sequence_mapper_notifier.h:  80)   Processed 1000000 reads\n",
      "  0:01:41.921   676M / 13G   INFO    General                 (sequence_mapper_notifier.h:  80)   Processed 1200000 reads\n",
      "  0:01:42.126   676M / 13G   INFO    General                 (sequence_mapper_notifier.h:  80)   Processed 2200000 reads\n",
      "  0:01:42.651   676M / 13G   INFO    General                 (sequence_mapper_notifier.h:  80)   Processed 4200000 reads\n",
      "  0:01:45.223   676M / 13G   INFO    General                 (sequence_mapper_notifier.h:  80)   Processed 8551694 reads\n",
      "  0:01:46.228   676M / 13G   INFO    General                 (sequence_mapper_notifier.h:  98)   Total 9842709 reads processed\n",
      "  0:01:47.709   676M / 13G   INFO    General                 (pair_info_count.cpp       : 209)   Edge pairs: 67108864 (rough upper limit)\n",
      "  0:01:47.709   676M / 13G   INFO    General                 (pair_info_count.cpp       : 213)   5988163 paired reads (60.8386% of all) aligned to long edges\n",
      "  0:01:47.753   276M / 13G   INFO    General                 (pair_info_count.cpp       : 354)     Insert size = 328.079, deviation = 63.1472, left quantile = 249, right quantile = 408, read length = 150\n",
      "  0:01:47.861   468M / 13G   INFO    General                 (pair_info_count.cpp       : 371)   Filtering data for library #0\n",
      "  0:01:47.862   468M / 13G   INFO    General                 (pair_info_count.cpp       :  39)   Selecting usual mapper\n",
      "  0:01:50.941   468M / 13G   INFO    General                 (sequence_mapper_notifier.h:  80)   Processed 200000 reads\n",
      "  0:01:51.011   472M / 13G   INFO    General                 (sequence_mapper_notifier.h:  80)   Processed 400000 reads\n",
      "  0:01:51.021   472M / 13G   INFO    General                 (sequence_mapper_notifier.h:  80)   Processed 600000 reads\n",
      "  0:01:51.025   472M / 13G   INFO    General                 (sequence_mapper_notifier.h:  80)   Processed 800000 reads\n",
      "  0:01:51.042   468M / 13G   INFO    General                 (sequence_mapper_notifier.h:  80)   Processed 1000000 reads\n",
      "  0:01:51.043   468M / 13G   INFO    General                 (sequence_mapper_notifier.h:  80)   Processed 1200000 reads\n",
      "  0:01:51.066   468M / 13G   INFO    General                 (sequence_mapper_notifier.h:  80)   Processed 2200000 reads\n",
      "  0:01:51.112   468M / 13G   INFO    General                 (sequence_mapper_notifier.h:  80)   Processed 4200000 reads\n",
      "  0:01:54.395   468M / 13G   INFO    General                 (sequence_mapper_notifier.h:  80)   Processed 8400000 reads\n",
      "  0:01:54.569   468M / 13G   INFO    General                 (sequence_mapper_notifier.h:  98)   Total 9842709 reads processed\n",
      "  0:01:54.569   468M / 13G   INFO    General                 (pair_info_count.cpp       : 383)   Mapping library #0\n",
      "  0:01:54.569   468M / 13G   INFO    General                 (pair_info_count.cpp       : 385)   Mapping paired reads (takes a while)\n",
      "  0:01:54.569   468M / 13G   INFO    General                 (pair_info_count.cpp       : 289)   Left insert size quantile 249, right insert size quantile 408, filtering threshold 2, rounding threshold 0\n",
      "  0:01:54.581   480M / 13G   INFO    General                 (pair_info_count.cpp       :  39)   Selecting usual mapper\n",
      "  0:01:57.289   492M / 13G   INFO    General                 (sequence_mapper_notifier.h:  80)   Processed 200000 reads\n",
      "  0:01:57.300   492M / 13G   INFO    General                 (sequence_mapper_notifier.h:  80)   Processed 400000 reads\n",
      "  0:01:57.311   492M / 13G   INFO    General                 (sequence_mapper_notifier.h:  80)   Processed 600000 reads\n",
      "  0:01:57.315   492M / 13G   INFO    General                 (sequence_mapper_notifier.h:  80)   Processed 800000 reads\n",
      "  0:01:57.316   492M / 13G   INFO    General                 (sequence_mapper_notifier.h:  80)   Processed 1000000 reads\n",
      "  0:01:57.319   492M / 13G   INFO    General                 (sequence_mapper_notifier.h:  80)   Processed 1200000 reads\n",
      "  0:01:57.330   492M / 13G   INFO    General                 (sequence_mapper_notifier.h:  80)   Processed 2200000 reads\n",
      "  0:01:57.369   492M / 13G   INFO    General                 (sequence_mapper_notifier.h:  80)   Processed 4200000 reads\n",
      "  0:02:00.146   500M / 13G   INFO    General                 (sequence_mapper_notifier.h:  80)   Processed 8400000 reads\n",
      "  0:02:00.356   500M / 13G   INFO    General                 (sequence_mapper_notifier.h:  98)   Total 9842709 reads processed\n",
      "  0:02:00.391   296M / 13G   INFO   StageManager             (stage.cpp                 : 132)   STAGE == Distance Estimation\n",
      "  0:02:00.391   296M / 13G   INFO    General                 (distance_estimation.cpp   : 173)   Processing library #0\n",
      "  0:02:00.391   296M / 13G   INFO    General                 (distance_estimation.cpp   : 149)   Weight Filter Done\n",
      "  0:02:00.391   296M / 13G   INFO   DistanceEstimator        (distance_estimation.hpp   : 116)   Using SIMPLE distance estimator\n",
      "  0:02:00.409   300M / 13G   INFO    General                 (distance_estimation.cpp   :  34)   Filtering info\n",
      "  0:02:00.409   300M / 13G   INFO    General                 (pair_info_filters.hpp     : 242)   Start filtering; index size: 17400\n",
      "  0:02:00.421   300M / 13G   INFO    General                 (pair_info_filters.hpp     : 263)   Done filtering\n",
      "  0:02:00.421   300M / 13G   INFO    General                 (distance_estimation.cpp   : 156)   Refining clustered pair information\n",
      "  0:02:00.423   300M / 13G   INFO    General                 (distance_estimation.cpp   : 158)   The refining of clustered pair information has been finished\n",
      "  0:02:00.423   300M / 13G   INFO    General                 (distance_estimation.cpp   : 160)   Improving paired information\n",
      "  0:02:00.461   328M / 13G   INFO   PairInfoImprover         (pair_info_improver.hpp    : 103)   Paired info stats: missing = 101; contradictional = 18\n",
      "  0:02:00.507   328M / 13G   INFO   PairInfoImprover         (pair_info_improver.hpp    : 103)   Paired info stats: missing = 0; contradictional = 0\n",
      "  0:02:00.507   328M / 13G   INFO    General                 (distance_estimation.cpp   : 103)   Filling scaffolding index\n",
      "  0:02:00.507   328M / 13G   INFO   DistanceEstimator        (distance_estimation.hpp   : 116)   Using SMOOTHING distance estimator\n",
      "  0:02:00.525   360M / 13G   INFO    General                 (distance_estimation.cpp   :  34)   Filtering info\n",
      "  0:02:00.525   360M / 13G   INFO    General                 (pair_info_filters.hpp     : 242)   Start filtering; index size: 12946\n",
      "  0:02:00.530   360M / 13G   INFO    General                 (pair_info_filters.hpp     : 263)   Done filtering\n",
      "  0:02:00.530   360M / 13G   INFO    General                 (distance_estimation.cpp   : 182)   Clearing raw paired index\n",
      "  0:02:00.532   360M / 13G   INFO   StageManager             (stage.cpp                 : 132)   STAGE == Repeat Resolving\n",
      "  0:02:00.533   360M / 13G   INFO    General                 (repeat_resolving.cpp      :  69)   Using Path-Extend repeat resolving\n",
      "  0:02:00.533   360M / 13G   INFO    General                 (launcher.cpp              : 477)   ExSPAnder repeat resolving tool started\n",
      "  0:02:00.557   384M / 13G   INFO    General                 (launcher.cpp              : 392)   Creating main extenders, unique edge length = 2000\n",
      "  0:02:00.558   384M / 13G   INFO    General                 (extenders_logic.cpp       : 275)   Estimated coverage of library #0 is 332.88\n",
      "  0:02:00.559   384M / 13G   INFO    General                 (extenders_logic.cpp       : 275)   Estimated coverage of library #0 is 332.88\n",
      "  0:02:00.565   384M / 13G   INFO    General                 (extenders_logic.cpp       : 472)   Using 1 paired-end library\n",
      "  0:02:00.565   384M / 13G   INFO    General                 (extenders_logic.cpp       : 473)   Using 1 paired-end scaffolding library\n",
      "  0:02:00.565   384M / 13G   INFO    General                 (extenders_logic.cpp       : 474)   Using 0 single read libraries\n",
      "  0:02:00.565   384M / 13G   INFO    General                 (launcher.cpp              : 420)   Total number of extenders is 3\n",
      "  0:02:00.565   384M / 13G   INFO    General                 (path_extender.hpp         : 885)   Processed 0 paths from 5912 (0%)\n",
      "  0:02:00.649   388M / 13G   INFO    General                 (path_extender.hpp         : 883)   Processed 128 paths from 5912 (2%)\n",
      "  0:02:00.657   388M / 13G   INFO    General                 (path_extender.hpp         : 883)   Processed 256 paths from 5912 (4%)\n",
      "  0:02:00.671   392M / 13G   INFO    General                 (path_extender.hpp         : 883)   Processed 512 paths from 5912 (8%)\n",
      "  0:02:00.675   392M / 13G   INFO    General                 (path_extender.hpp         : 885)   Processed 592 paths from 5912 (10%)\n",
      "  0:02:00.690   396M / 13G   INFO    General                 (path_extender.hpp         : 883)   Processed 1024 paths from 5912 (17%)\n",
      "  0:02:00.696   396M / 13G   INFO    General                 (path_extender.hpp         : 885)   Processed 1184 paths from 5912 (20%)\n",
      "  0:02:00.716   400M / 13G   INFO    General                 (path_extender.hpp         : 885)   Processed 1776 paths from 5912 (30%)\n",
      "  0:02:00.726   404M / 13G   INFO    General                 (path_extender.hpp         : 883)   Processed 2048 paths from 5912 (34%)\n",
      "  0:02:00.737   404M / 13G   INFO    General                 (path_extender.hpp         : 885)   Processed 2368 paths from 5912 (40%)\n",
      "  0:02:00.757   412M / 13G   INFO    General                 (path_extender.hpp         : 885)   Processed 2960 paths from 5912 (50%)\n",
      "  0:02:00.778   416M / 13G   INFO    General                 (path_extender.hpp         : 885)   Processed 3552 paths from 5912 (60%)\n",
      "  0:02:00.797   420M / 13G   INFO    General                 (path_extender.hpp         : 883)   Processed 4096 paths from 5912 (69%)\n",
      "  0:02:00.799   420M / 13G   INFO    General                 (path_extender.hpp         : 885)   Processed 4144 paths from 5912 (70%)\n",
      "  0:02:00.819   424M / 13G   INFO    General                 (path_extender.hpp         : 885)   Processed 4736 paths from 5912 (80%)\n",
      "  0:02:00.839   432M / 13G   INFO    General                 (path_extender.hpp         : 885)   Processed 5328 paths from 5912 (90%)\n",
      "  0:02:00.861   436M / 13G   INFO    General                 (launcher.cpp              : 234)   Finalizing paths\n",
      "  0:02:00.861   436M / 13G   INFO    General                 (launcher.cpp              : 236)   Deduplicating paths\n",
      "  0:02:00.873   432M / 13G   INFO    General                 (launcher.cpp              : 240)   Paths deduplicated\n",
      "  0:02:00.873   432M / 13G   INFO   PEResolver               (pe_resolver.hpp           : 295)   Removing overlaps\n",
      "  0:02:00.873   432M / 13G   INFO   PEResolver               (pe_resolver.hpp           : 298)   Sorting paths\n",
      "  0:02:00.875   432M / 13G   INFO   PEResolver               (pe_resolver.hpp           : 305)   Marking overlaps\n",
      "  0:02:00.875   432M / 13G   INFO   OverlapRemover           (pe_resolver.hpp           : 130)   Marking start/end overlaps\n",
      "  0:02:00.881   432M / 13G   INFO   OverlapRemover           (pe_resolver.hpp           : 133)   Marking remaining overlaps\n",
      "  0:02:00.888   432M / 13G   INFO   PEResolver               (pe_resolver.hpp           : 308)   Splitting paths\n",
      "  0:02:00.889   432M / 13G   INFO   PEResolver               (pe_resolver.hpp           : 313)   Deduplicating paths\n",
      "  0:02:00.892   432M / 13G   INFO   PEResolver               (pe_resolver.hpp           : 315)   Overlaps removed\n",
      "  0:02:00.899   432M / 13G   INFO    General                 (launcher.cpp              : 257)   Paths finalized\n",
      "  0:02:00.899   432M / 13G   INFO    General                 (launcher.cpp              : 427)   Closing gaps in paths\n",
      "  0:02:00.917   436M / 13G   INFO    General                 (launcher.cpp              : 455)   Gap closing completed\n",
      "  0:02:00.923   436M / 13G   INFO    General                 (launcher.cpp              : 286)   Traversing tandem repeats\n",
      "  0:02:00.950   436M / 13G   INFO    General                 (launcher.cpp              : 296)   Traversed 3 loops\n",
      "  0:02:00.950   436M / 13G   INFO    General                 (launcher.cpp              : 234)   Finalizing paths\n",
      "  0:02:00.950   436M / 13G   INFO    General                 (launcher.cpp              : 236)   Deduplicating paths\n",
      "  0:02:00.952   436M / 13G   INFO    General                 (launcher.cpp              : 240)   Paths deduplicated\n",
      "  0:02:00.952   436M / 13G   INFO   PEResolver               (pe_resolver.hpp           : 295)   Removing overlaps\n",
      "  0:02:00.952   436M / 13G   INFO   PEResolver               (pe_resolver.hpp           : 298)   Sorting paths\n",
      "  0:02:00.953   436M / 13G   INFO   PEResolver               (pe_resolver.hpp           : 305)   Marking overlaps\n",
      "  0:02:00.953   436M / 13G   INFO   OverlapRemover           (pe_resolver.hpp           : 130)   Marking start/end overlaps\n",
      "  0:02:00.958   436M / 13G   INFO   OverlapRemover           (pe_resolver.hpp           : 133)   Marking remaining overlaps\n",
      "  0:02:00.963   436M / 13G   INFO   PEResolver               (pe_resolver.hpp           : 308)   Splitting paths\n",
      "  0:02:00.963   436M / 13G   INFO   PEResolver               (pe_resolver.hpp           : 313)   Deduplicating paths\n",
      "  0:02:00.965   436M / 13G   INFO   PEResolver               (pe_resolver.hpp           : 315)   Overlaps removed\n",
      "  0:02:00.971   436M / 13G   INFO    General                 (launcher.cpp              : 257)   Paths finalized\n",
      "  0:02:00.992   436M / 13G   INFO    General                 (launcher.cpp              : 529)   ExSPAnder repeat resolving tool finished\n",
      "  0:02:01.014   408M / 13G   INFO   StageManager             (stage.cpp                 : 132)   STAGE == Contig Output\n",
      "  0:02:01.014   408M / 13G   INFO    General                 (contig_output_stage.cpp   :  40)   Writing GFA to /mnt/d/linux/W/MoonNGS/raw/20180813Novogene16species_Human_Methyl/1.rawdata/MNH08365/Tychus/spades//K77/assembly_graph_with_scaffolds.gfa\n",
      "  0:02:01.068   408M / 13G   INFO    General                 (contig_output.hpp         :  22)   Outputting contigs to /mnt/d/linux/W/MoonNGS/raw/20180813Novogene16species_Human_Methyl/1.rawdata/MNH08365/Tychus/spades//K77/before_rr.fasta\n",
      "  0:02:01.136   408M / 13G   INFO    General                 (contig_output_stage.cpp   :  51)   Outputting FastG graph to /mnt/d/linux/W/MoonNGS/raw/20180813Novogene16species_Human_Methyl/1.rawdata/MNH08365/Tychus/spades//K77/assembly_graph.fastg\n",
      "  0:02:01.357   408M / 13G   INFO    General                 (contig_output_stage.cpp   :  20)   Outputting FastG paths to /mnt/d/linux/W/MoonNGS/raw/20180813Novogene16species_Human_Methyl/1.rawdata/MNH08365/Tychus/spades//K77/final_contigs.paths\n",
      "  0:02:01.455   408M / 13G   INFO    General                 (contig_output_stage.cpp   :  20)   Outputting FastG paths to /mnt/d/linux/W/MoonNGS/raw/20180813Novogene16species_Human_Methyl/1.rawdata/MNH08365/Tychus/spades//K77/scaffolds.paths\n",
      "  0:02:01.551   408M / 13G   INFO    General                 (launch.hpp                : 149)   SPAdes finished\n",
      "  0:02:01.891    96M / 13G   INFO    General                 (main.cpp                  : 109)   Assembling time: 0 hours 2 minutes 1 seconds\n",
      "\n",
      "===== Assembling finished. Used k-mer sizes: 21, 33, 55, 77 \n",
      "\n",
      " * Corrected reads are in /mnt/d/linux/W/MoonNGS/raw/20180813Novogene16species_Human_Methyl/1.rawdata/MNH08365/Tychus/spades/corrected/\n",
      " * Assembled contigs are in /mnt/d/linux/W/MoonNGS/raw/20180813Novogene16species_Human_Methyl/1.rawdata/MNH08365/Tychus/spades/contigs.fasta\n",
      " * Assembled scaffolds are in /mnt/d/linux/W/MoonNGS/raw/20180813Novogene16species_Human_Methyl/1.rawdata/MNH08365/Tychus/spades/scaffolds.fasta\n",
      " * Assembly graph is in /mnt/d/linux/W/MoonNGS/raw/20180813Novogene16species_Human_Methyl/1.rawdata/MNH08365/Tychus/spades/assembly_graph.fastg\n",
      " * Assembly graph in GFA format is in /mnt/d/linux/W/MoonNGS/raw/20180813Novogene16species_Human_Methyl/1.rawdata/MNH08365/Tychus/spades/assembly_graph_with_scaffolds.gfa\n",
      " * Paths in the assembly graph corresponding to the contigs are in /mnt/d/linux/W/MoonNGS/raw/20180813Novogene16species_Human_Methyl/1.rawdata/MNH08365/Tychus/spades/contigs.paths\n",
      " * Paths in the assembly graph corresponding to the scaffolds are in /mnt/d/linux/W/MoonNGS/raw/20180813Novogene16species_Human_Methyl/1.rawdata/MNH08365/Tychus/spades/scaffolds.paths\n",
      "\n",
      "======= SPAdes pipeline finished.\n",
      "\n",
      "SPAdes log can be found here: /mnt/d/linux/W/MoonNGS/raw/20180813Novogene16species_Human_Methyl/1.rawdata/MNH08365/Tychus/spades/spades.log\n",
      "\n",
      "Thank you for using SPAdes!\n",
      "\n"
     ]
    }
   ],
   "source": [
    "result = result.decode('utf-8')\n",
    "print(result)\n",
    "os.chdir('..')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## run IDBA-UD"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "if not os.path.exists('IDBA-UD'):\n",
    "    os.mkdir('IDBA-UD')\n",
    "os.chdir('IDBA-UD')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/home/p/anaconda/anaconda3_5.2.0/bin/fq2fa --merge --filter /mnt/d/linux/W/MoonNGS/raw/20180813Novogene16species_Human_Methyl/1.rawdata/MNH08365/Tychus/Trimmomatic/read1.pair.fq /mnt/d/linux/W/MoonNGS/raw/20180813Novogene16species_Human_Methyl/1.rawdata/MNH08365/Tychus/Trimmomatic/read2.pair.fq idba-paired-contigs.fa \n",
      "\n"
     ]
    }
   ],
   "source": [
    "command_line = '''{FQ2FA} --merge --filter {read1} {read2} idba-paired-contigs.fa '''.format( \\\n",
    "               FQ2FA = FQ2FA, bestkmer=bestkmer, outfolder = outfolder, thread = thread, \\\n",
    "               read1 = os.path.join(outfolder,'Trimmomatic','read1.pair.fq'), read2 = os.path.join(outfolder,'Trimmomatic','read2.pair.fq'))\n",
    "print(command_line)\n",
    "result = subprocess.check_output(command_line, shell=True)\n",
    "result = result.decode('utf-8')\n",
    "print(result)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/home/p/anaconda/anaconda3_5.2.0/bin/idba_ud -r idba-paired-contigs.fa --num_threads 24 -o . \n",
      "number of threads 24\n",
      "reads 19664504\n",
      "long reads 0\n",
      "extra reads 0\n",
      "read_length 150\n",
      "kmer 20\n",
      "kmers 19062498 19607055\n",
      "merge bubble 1157\n",
      "contigs: 42009 n50: 484 max: 5656 mean: 102 total length: 4296420 n80: 39\n",
      "aligned 12876007 reads\n",
      "confirmed bases: 2292413 correct reads: 10615469 bases: 2384891\n",
      "distance mean 321.22 sd 60.0233\n",
      "seed contigs 4003 local contigs 84018\n",
      "kmer 40\n",
      "kmers 15609138 15859200\n",
      "merge bubble 3070\n",
      "contigs: 6916 n50: 17629 max: 118926 mean: 504 total length: 3490265 n80: 3298\n",
      "aligned 18304412 reads\n",
      "confirmed bases: 2944690 correct reads: 17798623 bases: 1672150\n",
      "distance mean 327.924 sd 60.7581\n",
      "seed contigs 710 local contigs 13832\n",
      "kmer 60\n",
      "kmers 10003157 9872370\n",
      "merge bubble 309\n",
      "contigs: 2512 n50: 46939 max: 157761 mean: 1311 total length: 3295257 n80: 12857\n",
      "aligned 18877725 reads\n",
      "confirmed bases: 2994720 correct reads: 18409104 bases: 155576\n",
      "distance mean 328.138 sd 60.8018\n",
      "seed contigs 499 local contigs 5024\n",
      "kmer 80\n",
      "kmers 5328852 5327523\n",
      "merge bubble 11\n",
      "contigs: 1108 n50: 91088 max: 278317 mean: 2879 total length: 3190742 n80: 23970\n",
      "aligned 19121338 reads\n",
      "confirmed bases: 3013056 correct reads: 18584813 bases: 48226\n",
      "distance mean 328.184 sd 60.81\n",
      "seed contigs 372 local contigs 2216\n",
      "kmer 100\n",
      "kmers 4074696 4074530\n",
      "merge bubble 22\n",
      "contigs: 610 n50: 94757 max: 341997 mean: 5157 total length: 3146227 n80: 27309\n",
      "reads 19664504\n",
      "aligned 19184941 reads\n",
      "distance mean 328.225 sd 60.8083\n",
      "expected coverage 6.27307\n",
      "edgs 131\n",
      "contigs: 590 n50: 125076 max: 434655 mean: 5329 total length: 3144210 n80: 33077\n",
      "\n"
     ]
    }
   ],
   "source": [
    "command_line = '''{IDBA_UD} -r idba-paired-contigs.fa --num_threads {thread} -o . '''.format( \\\n",
    "               IDBA_UD = IDBA_UD, bestkmer=bestkmer, outfolder = outfolder, thread = thread, \\\n",
    "               read1 = os.path.join(outfolder,'Trimmomatic','read1.pair.fq'), read2 = os.path.join(outfolder,'Trimmomatic','read2.pair.fq'))\n",
    "print(command_line)\n",
    "result = subprocess.check_output(command_line, shell=True)\n",
    "result = result.decode('utf-8')\n",
    "print(result)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "os.chdir('..')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## run soapdenovo"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "if not os.path.exists('soapdenovo'):\n",
    "    os.mkdir('soapdenovo')\n",
    "os.chdir('soapdenovo')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [],
   "source": [
    "fw=open('myconfig','w')\n",
    "fw.write('max_rd_len=150\\n[LIB]\\navg_ins=350\\nreverse_seq=0\\nasm_flags=3\\nq1={read1}\\nq2={read2}\\n'.format(\\\n",
    "        read1 = os.path.join(outfolder,'Trimmomatic','read1.pair.fq'), read2 = os.path.join(outfolder,'Trimmomatic','read2.pair.fq')))\n",
    "fw.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/home/p/anaconda/anaconda3_5.2.0/bin/SOAPdenovo-127mer all -s myconfig -K 61 -R -o soapdenovo -d 10 -p 24  1>ass.log 2>ass.err \n",
      "\n"
     ]
    }
   ],
   "source": [
    "command_line = '''{SOAPDENOVO} all -s myconfig -K {bestkmer} -R -o soapdenovo -d 10 -p 24  1>ass.log 2>ass.err '''.format( \\\n",
    "               SOAPDENOVO = SOAPDENOVO, bestkmer=bestkmer, outfolder = outfolder, thread = thread, \\\n",
    "               read1 = os.path.join(outfolder,'Trimmomatic','read1.pair.fq'), read2 = os.path.join(outfolder,'Trimmomatic','read2.pair.fq'))\n",
    "print(command_line)\n",
    "result = subprocess.check_output(command_line, shell=True)\n",
    "result = result.decode('utf-8')\n",
    "print(result)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "os.chdir('..')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## run CISA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": [
    "if not os.path.exists('CISA'):\n",
    "    os.mkdir('CISA')\n",
    "os.chdir('CISA')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'soapdenovo.fa'"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "shutil.copy('../abyss/abyss-contigs.fa','abyss.fa')\n",
    "shutil.copy('../velvet/contigs.fa','velvet.fa')\n",
    "shutil.copy('../spades/contigs.fasta','spades.fa')\n",
    "shutil.copy('../IDBA-UD/contig.fa','IDBA-UD.fa')\n",
    "shutil.copy('../soapdenovo/soapdenovo.contig','soapdenovo.fa')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [],
   "source": [
    "ctgs = ['soapdenovo.fa', 'velvet.fa', 'abyss.fa', 'spades.fa', 'IDBA-UD.fa']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [],
   "source": [
    "fw=open('ToMerge.config','w')\n",
    "fw.write('count=5\\n')\n",
    "for i in ctgs:\n",
    "    fw.write('data={fastafile},title={fastaname}\\n'.format(fastafile=i,fastaname=i.split('.')[0]))\n",
    "fw.write('min_length=400\\nMaster_file=Merged.fa\\nGap=1\\n')\n",
    "fw.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/home/p/CISA/CISA1.3/Merge.py ToMerge.config \n",
      "soapdenovo.fa.p.fa\n",
      "Number of contigs: 166\n",
      "Length of the longest contig: 157765\n",
      "whole:2986723\n",
      "N50: 49127\n",
      "velvet.fa.p.fa\n",
      "Number of contigs: 133\n",
      "Length of the longest contig: 291605\n",
      "whole:3014292\n",
      "N50: 75219\n",
      "abyss.fa.p.fa\n",
      "Number of contigs: 93\n",
      "Length of the longest contig: 278946\n",
      "whole:3108220\n",
      "N50: 93105\n",
      "spades.fa.p.fa\n",
      "Number of contigs: 316\n",
      "Length of the longest contig: 434654\n",
      "whole:3135457\n",
      "N50: 102757\n",
      "IDBA-UD.fa.p.fa\n",
      "Number of contigs: 200\n",
      "Length of the longest contig: 341997\n",
      "whole:3067591\n",
      "N50: 94757\n",
      "\n"
     ]
    }
   ],
   "source": [
    "command_line = '''{MERGE} ToMerge.config '''.format(MERGE = MERGE)\n",
    "print(command_line)\n",
    "result = subprocess.check_output(command_line, shell=True)\n",
    "result = result.decode('utf-8')\n",
    "print(result)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "max genome length is 3135457\n"
     ]
    }
   ],
   "source": [
    "lis = open('Merge_info').readlines()\n",
    "genomelens = [int(e.split()[-1]) for e in lis if 'Whole Genome' in e]\n",
    "genomelen = max(genomelens)\n",
    "print('max genome length is', genomelen)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [],
   "source": [
    "fw=open('CISA.config','w')\n",
    "fw.write('''\n",
    "genome={genomelen}\n",
    "infile=Merged.fa\n",
    "outfile=Final.fa\n",
    "nucmer={NCUMER}\n",
    "R2_Gap=0.95\n",
    "CISA={CISAPATH}\n",
    "makeblastdb={MAKEBLASTDB}\n",
    "blastn={BLASTN}\n",
    "'''.format(genomelen=genomelen,NCUMER=NCUMER,CISAPATH=CISAPATH,MAKEBLASTDB=MAKEBLASTDB,BLASTN=BLASTN))\n",
    "fw.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "python /home/p/CISA/CISA1.3/CISA.py CISA.config \n"
     ]
    }
   ],
   "source": [
    "command_line = '''python {CISA} CISA.config '''.format(CISA = CISA)\n",
    "print(command_line)\n",
    "result = subprocess.check_output(command_line, shell=True)\n",
    "result = result.decode('utf-8')\n",
    "#print(result)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [],
   "source": [
    "os.chdir('..')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## run prokka"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/mnt/d/linux/W/MoonNGS/raw/20180813Novogene16species_Human_Methyl/1.rawdata/MNH08365/Tychus'"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "os.getcwd()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [],
   "source": [
    "if not os.path.exists('prokka'):\n",
    "    os.mkdir('prokka')\n",
    "os.chdir('prokka')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [],
   "source": [
    "ls_contigs = list(SeqIO.parse(open('../CISA/Final.fa'), 'fasta'))\n",
    "outfile = '{bacteriaStrain}.genome.fa'.format(bacteriaStrain=bacteriaStrain)\n",
    "fout = open(outfile,'w')\n",
    "for _n, _e in enumerate(ls_contigs):\n",
    "    fout.write('>'+bacteriaStrain+'.g'+str(_n+1)+'\\n'+str(_e.seq)+'\\n')\n",
    "fout.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/home/p/prokka/prokka-master/bin/prokka --kingdom Bacteria --cpus 24  MNH08365.genome.fa --outdir .                 --prefix MNH08365 --force --metagenome --locustag MNH08365\n",
      "\n"
     ]
    }
   ],
   "source": [
    "command_line = '''{PROKKA} --kingdom Bacteria --cpus {thread}  {bacteriaStrain}.genome.fa --outdir . \\\n",
    "                --prefix {bacteriaStrain} --force --metagenome --locustag {bacteriaStrain}'''.format(\\\n",
    "                PROKKA = PROKKA, thread=thread, bacteriaStrain=bacteriaStrain)\n",
    "print(command_line)\n",
    "result = subprocess.check_output(command_line, shell=True)\n",
    "result = result.decode('utf-8')\n",
    "print(result)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## run quast"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [],
   "source": [
    "os.chdir('..')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/mnt/d/linux/W/MoonNGS/raw/20180813Novogene16species_Human_Methyl/1.rawdata/MNH08365/Tychus'"
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "os.getcwd()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "python /home/p/quast/quast-4.6.3/quast.py -t 24 -o quast                 ./prokka/MNH08365.genome.fa ./CISA/abyss.fa.p.fa ./CISA/velvet.fa.p.fa ./CISA/spades.fa.p.fa                 ./CISA/IDBA-UD.fa.p.fa ./CISA/soapdenovo.fa.p.fa\n",
      "/home/p/quast/quast-4.6.3/quast.py -t 24 -o quast ./prokka/MNH08365.genome.fa ./CISA/abyss.fa.p.fa ./CISA/velvet.fa.p.fa ./CISA/spades.fa.p.fa ./CISA/IDBA-UD.fa.p.fa ./CISA/soapdenovo.fa.p.fa\n",
      "\n",
      "Version: 4.6.3\n",
      "\n",
      "System information:\n",
      "  OS: Linux-3.10.0-862.6.3.el7.x86_64-x86_64-with-centos-7.5.1804-Core (linux_64)\n",
      "  Python version: 3.6.4\n",
      "  CPUs number: 24\n",
      "\n",
      "Started: 2018-08-20 23:07:36\n",
      "\n",
      "Logging to /mnt/d/linux/W/MoonNGS/raw/20180813Novogene16species_Human_Methyl/1.rawdata/MNH08365/Tychus/quast/quast.log\n",
      "\n",
      "CWD: /mnt/d/linux/W/MoonNGS/raw/20180813Novogene16species_Human_Methyl/1.rawdata/MNH08365/Tychus\n",
      "Main parameters: \n",
      "  Threads: 24, minimum contig length: 500, ambiguity: one, threshold for extensive misassembly size: 1000\n",
      "\n",
      "Contigs:\n",
      "  Pre-processing...\n",
      "  1  ./prokka/MNH08365.genome.fa ==> MNH08365.genome\n",
      "  2  ./CISA/abyss.fa.p.fa ==> abyss.fa.p\n",
      "  3  ./CISA/velvet.fa.p.fa ==> velvet.fa.p\n",
      "  4  ./CISA/spades.fa.p.fa ==> spades.fa.p\n",
      "  5  ./CISA/IDBA-UD.fa.p.fa ==> IDBA-UD.fa.p\n",
      "  6  ./CISA/soapdenovo.fa.p.fa ==> soapdenovo.fa.p\n",
      "\n",
      "2018-08-20 23:07:37\n",
      "Running Basic statistics processor...\n",
      "  Contig files: \n",
      "    1  MNH08365.genome\n",
      "    2  abyss.fa.p\n",
      "    3  velvet.fa.p\n",
      "    4  spades.fa.p\n",
      "    5  IDBA-UD.fa.p\n",
      "    6  soapdenovo.fa.p\n",
      "  Calculating N50 and L50...\n",
      "    1  MNH08365.genome, N50 = 105921, L50 = 9, Total length = 3157694, GC % = 46.19, # N's per 100 kbp =  0.00\n",
      "    2  abyss.fa.p, N50 = 93105, L50 = 11, Total length = 3107797, GC % = 46.27, # N's per 100 kbp =  0.00\n",
      "    3  velvet.fa.p, N50 = 75219, L50 = 12, Total length = 3011058, GC % = 46.25, # N's per 100 kbp =  0.00\n",
      "    4  spades.fa.p, N50 = 102757, L50 = 9, Total length = 3078612, GC % = 46.15, # N's per 100 kbp =  0.00\n",
      "    5  IDBA-UD.fa.p, N50 = 94757, L50 = 10, Total length = 3054657, GC % = 46.17, # N's per 100 kbp =  0.00\n",
      "    6  soapdenovo.fa.p, N50 = 49127, L50 = 17, Total length = 2983546, GC % = 46.24, # N's per 100 kbp =  0.00\n",
      "  Drawing Nx plot...\n",
      "    saved to /mnt/d/linux/W/MoonNGS/raw/20180813Novogene16species_Human_Methyl/1.rawdata/MNH08365/Tychus/quast/basic_stats/Nx_plot.pdf\n",
      "  Drawing cumulative plot...\n",
      "    saved to /mnt/d/linux/W/MoonNGS/raw/20180813Novogene16species_Human_Methyl/1.rawdata/MNH08365/Tychus/quast/basic_stats/cumulative_plot.pdf\n",
      "  Drawing GC content plot...\n",
      "    saved to /mnt/d/linux/W/MoonNGS/raw/20180813Novogene16species_Human_Methyl/1.rawdata/MNH08365/Tychus/quast/basic_stats/GC_content_plot.pdf\n",
      "  Drawing MNH08365.genome GC content plot...\n",
      "    saved to /mnt/d/linux/W/MoonNGS/raw/20180813Novogene16species_Human_Methyl/1.rawdata/MNH08365/Tychus/quast/basic_stats/MNH08365.genome_GC_content_plot.pdf\n",
      "  Drawing abyss.fa.p GC content plot...\n",
      "    saved to /mnt/d/linux/W/MoonNGS/raw/20180813Novogene16species_Human_Methyl/1.rawdata/MNH08365/Tychus/quast/basic_stats/abyss.fa.p_GC_content_plot.pdf\n",
      "  Drawing velvet.fa.p GC content plot...\n",
      "    saved to /mnt/d/linux/W/MoonNGS/raw/20180813Novogene16species_Human_Methyl/1.rawdata/MNH08365/Tychus/quast/basic_stats/velvet.fa.p_GC_content_plot.pdf\n",
      "  Drawing spades.fa.p GC content plot...\n",
      "    saved to /mnt/d/linux/W/MoonNGS/raw/20180813Novogene16species_Human_Methyl/1.rawdata/MNH08365/Tychus/quast/basic_stats/spades.fa.p_GC_content_plot.pdf\n",
      "  Drawing IDBA-UD.fa.p GC content plot...\n",
      "    saved to /mnt/d/linux/W/MoonNGS/raw/20180813Novogene16species_Human_Methyl/1.rawdata/MNH08365/Tychus/quast/basic_stats/IDBA-UD.fa.p_GC_content_plot.pdf\n",
      "  Drawing soapdenovo.fa.p GC content plot...\n",
      "    saved to /mnt/d/linux/W/MoonNGS/raw/20180813Novogene16species_Human_Methyl/1.rawdata/MNH08365/Tychus/quast/basic_stats/soapdenovo.fa.p_GC_content_plot.pdf\n",
      "Done.\n",
      "\n",
      "NOTICE: Genes are not predicted by default. Use --gene-finding option to enable it.\n",
      "\n",
      "2018-08-20 23:07:40\n",
      "Creating large visual summaries...\n",
      "This may take a while: press Ctrl-C to skip this step..\n",
      "  1 of 2: Creating Icarus viewers...\n",
      "  2 of 2: Creating PDF with all tables and plots...\n",
      "Done\n",
      "\n",
      "2018-08-20 23:07:41\n",
      "RESULTS:\n",
      "  Text versions of total report are saved to /mnt/d/linux/W/MoonNGS/raw/20180813Novogene16species_Human_Methyl/1.rawdata/MNH08365/Tychus/quast/report.txt, report.tsv, and report.tex\n",
      "  Text versions of transposed total report are saved to /mnt/d/linux/W/MoonNGS/raw/20180813Novogene16species_Human_Methyl/1.rawdata/MNH08365/Tychus/quast/transposed_report.txt, transposed_report.tsv, and transposed_report.tex\n",
      "  HTML version (interactive tables and plots) saved to /mnt/d/linux/W/MoonNGS/raw/20180813Novogene16species_Human_Methyl/1.rawdata/MNH08365/Tychus/quast/report.html\n",
      "  PDF version (tables and plots) is saved to /mnt/d/linux/W/MoonNGS/raw/20180813Novogene16species_Human_Methyl/1.rawdata/MNH08365/Tychus/quast/report.pdf\n",
      "  Icarus (contig browser) is saved to /mnt/d/linux/W/MoonNGS/raw/20180813Novogene16species_Human_Methyl/1.rawdata/MNH08365/Tychus/quast/icarus.html\n",
      "  Log saved to /mnt/d/linux/W/MoonNGS/raw/20180813Novogene16species_Human_Methyl/1.rawdata/MNH08365/Tychus/quast/quast.log\n",
      "\n",
      "Finished: 2018-08-20 23:07:41\n",
      "Elapsed time: 0:00:05.077674\n",
      "NOTICEs: 1; WARNINGs: 0; non-fatal ERRORs: 0\n",
      "\n",
      "Thank you for using QUAST!\n",
      "\n"
     ]
    }
   ],
   "source": [
    "command_line = '''python {QUAST} -t {thread} -o quast \\\n",
    "                ./prokka/{bacteriaStrain}.genome.fa ./CISA/abyss.fa.p.fa ./CISA/velvet.fa.p.fa ./CISA/spades.fa.p.fa \\\n",
    "                ./CISA/IDBA-UD.fa.p.fa ./CISA/soapdenovo.fa.p.fa'''.format(\\\n",
    "                QUAST = QUAST, thread=thread, bacteriaStrain=bacteriaStrain)\n",
    "print(command_line)\n",
    "result = subprocess.check_output(command_line, shell=True)\n",
    "result = result.decode('utf-8')\n",
    "print(result)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## clean files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [],
   "source": [
    "if cleanOrNot:\n",
    "    os.system('rm -rf {outfolder}/abyss'.format(outfolder=outfolder))\n",
    "    os.system('rm -rf {outfolder}/IDBA-UD'.format(outfolder=outfolder))\n",
    "    os.system('rm -rf {outfolder}/kmergenie'.format(outfolder=outfolder))\n",
    "    os.system('rm -rf {outfolder}/soapdenovo'.format(outfolder=outfolder))\n",
    "    os.system('rm -rf {outfolder}/spades'.format(outfolder=outfolder))\n",
    "    os.system('rm -rf {outfolder}/Trimmomatic'.format(outfolder=outfolder))\n",
    "    os.system('rm -rf {outfolder}/velvet'.format(outfolder=outfolder))\n",
    "    os.system('rm -rf {outfolder}/CISA'.format(outfolder=outfolder))\n",
    "    os.system('rm -rf {outfolder}/readlist.txt'.format(outfolder=outfolder))\n",
    "    "
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.4"
  },
  "latex_envs": {
   "LaTeX_envs_menu_present": true,
   "autoclose": false,
   "autocomplete": true,
   "bibliofile": "biblio.bib",
   "cite_by": "apalike",
   "current_citInitial": 1,
   "eqLabelWithNumbers": true,
   "eqNumInitial": 1,
   "hotkeys": {
    "equation": "Ctrl-E",
    "itemize": "Ctrl-I"
   },
   "labels_anchors": false,
   "latex_user_defs": false,
   "report_style_numbering": false,
   "user_envs_cfg": false
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {
    "height": "140px",
    "width": "178px"
   },
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": true,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {
    "height": "calc(100% - 180px)",
    "left": "10px",
    "top": "150px",
    "width": "287px"
   },
   "toc_section_display": true,
   "toc_window_display": true
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
